// @prisma/cli/dist/lib/project/interactive-setup.js
import { usageError } from "../../shell/errors.js";
import { selectPrompt, textPrompt } from "../../shell/prompt.js";
import { inferTargetName, sortProjects } from "./resolution.js";
import { toProjectSummary, validateProjectSetupNameText } from "./setup.js";
//#region src/lib/project/interactive-setup.ts
async function promptForProjectSetupChoice(options) {
	const sortedProjects = sortProjects(options.projects);
	const projectNames = sortedProjects.map((project) => project.name);
	const duplicateNames = new Set(projectNames.filter((name, index) => projectNames.indexOf(name) !== index));
	const choice = await selectPrompt({
		input: options.context.runtime.stdin,
		output: options.context.runtime.stderr,
		message: "Which Project should this directory use?",
		choices: [
			...sortedProjects.map((project) => ({
				label: duplicateNames.has(project.name) ? `${project.name} (${project.id})` : project.name,
				value: {
					kind: "project",
					project
				}
			})),
			{
				label: "Create a new Project",
				value: { kind: "create" }
			},
			{
				label: "Cancel",
				value: { kind: "cancel" }
			}
		]
	});
	if (choice.kind === "cancel") throw usageError("Project setup canceled", options.cancel.why, options.cancel.fix, options.cancel.nextSteps, "project");
	if (choice.kind === "project") return {
		project: toProjectSummary(choice.project),
		action: "linked",
		targetName: choice.project.name,
		targetNameSource: "prompt"
	};
	const suggestedName = await inferTargetName(options.context.runtime.cwd, options.context.runtime.signal);
	const rawName = await textPrompt({
		input: options.context.runtime.stdin,
		output: options.context.runtime.stderr,
		message: "Project name",
		placeholder: suggestedName.name,
		validate: (value) => validateProjectSetupNameText(value, suggestedName.name)
	});
	const projectName = rawName.trim() || suggestedName.name;
	return {
		project: toProjectSummary(await options.createProject(projectName)),
		action: "created",
		targetName: projectName,
		targetNameSource: rawName.trim() ? "prompt" : suggestedName.source
	};
}
//#endregion
export { promptForProjectSetupChoice };
