// @prisma/cli/dist/lib/project/local-pin.js
import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import path from "node:path";
import { Result, TaggedError, UnhandledException } from "better-result";
//#region src/lib/project/local-pin.ts
const LOCAL_RESOLUTION_PIN_RELATIVE_PATH = ".prisma/local.json";
var LocalResolutionPinInvalidJsonError = class extends TaggedError("LocalResolutionPinInvalidJsonError")() {
	constructor(cause) {
		super({
			message: `${LOCAL_RESOLUTION_PIN_RELATIVE_PATH} contains invalid JSON.`,
			cause,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinInvalidShapeError = class extends TaggedError("LocalResolutionPinInvalidShapeError")() {
	constructor() {
		super({
			message: `${LOCAL_RESOLUTION_PIN_RELATIVE_PATH} must contain workspaceId and projectId string fields only.`,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinReadAbortedError = class extends TaggedError("LocalResolutionPinReadAbortedError")() {
	constructor(cause) {
		super({
			message: `Reading ${LOCAL_RESOLUTION_PIN_RELATIVE_PATH} was aborted.`,
			cause,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinSerializationError = class extends TaggedError("LocalResolutionPinSerializationError")() {
	constructor(cause) {
		super({
			message: `Could not serialize ${LOCAL_RESOLUTION_PIN_RELATIVE_PATH}.`,
			cause,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinWriteAbortedError = class extends TaggedError("LocalResolutionPinWriteAbortedError")() {
	constructor(cause) {
		super({
			message: `Writing ${LOCAL_RESOLUTION_PIN_RELATIVE_PATH} was aborted.`,
			cause,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinWriteFailedError = class extends TaggedError("LocalResolutionPinWriteFailedError")() {
	constructor(operation, cause) {
		super({
			message: `Could not write ${LOCAL_RESOLUTION_PIN_RELATIVE_PATH}.`,
			cause,
			operation,
			pinPath: LOCAL_RESOLUTION_PIN_RELATIVE_PATH
		});
	}
};
var LocalResolutionPinGitignoreUpdateAbortedError = class extends TaggedError("LocalResolutionPinGitignoreUpdateAbortedError")() {
	constructor(cause) {
		super({
			message: "Updating .gitignore for the local Project binding was aborted.",
			cause,
			gitignorePath: ".gitignore"
		});
	}
};
var LocalResolutionPinGitignoreUpdateFailedError = class extends TaggedError("LocalResolutionPinGitignoreUpdateFailedError")() {
	constructor(operation, cause) {
		super({
			message: "Could not update .gitignore for the local Project binding.",
			cause,
			operation,
			gitignorePath: ".gitignore"
		});
	}
};
async function readLocalResolutionPin(cwd, signal) {
	return Result.gen(async function* () {
		yield* ensureLocalResolutionPinReadNotAborted(signal);
		const file = yield* Result.await(readLocalResolutionPinFile(cwd, signal));
		if (file.kind === "missing") return Result.ok({ kind: "missing" });
		const parsed = yield* parseLocalResolutionPin(file.raw);
		if (!isLocalResolutionPin(parsed)) return Result.err(new LocalResolutionPinInvalidShapeError());
		return Result.ok({
			kind: "present",
			pin: parsed
		});
	});
}
function ensureLocalResolutionPinReadNotAborted(signal) {
	return Result.try({
		try: () => signal?.throwIfAborted(),
		catch: (cause) => new LocalResolutionPinReadAbortedError(cause)
	});
}
async function readLocalResolutionPinFile(cwd, signal) {
	const readResult = await Result.tryPromise({
		try: () => readFile(path.join(cwd, LOCAL_RESOLUTION_PIN_RELATIVE_PATH), {
			encoding: "utf8",
			signal
		}),
		catch: (cause) => signal?.aborted ? new LocalResolutionPinReadAbortedError(cause) : new UnhandledException({ cause })
	});
	if (readResult.isErr()) {
		if (readResult.error instanceof UnhandledException && readResult.error.cause.code === "ENOENT") return Result.ok({ kind: "missing" });
		return Result.err(readResult.error);
	}
	return Result.ok({
		kind: "present",
		raw: readResult.value
	});
}
function parseLocalResolutionPin(raw) {
	return Result.try({
		try: () => JSON.parse(raw),
		catch: (cause) => cause instanceof SyntaxError ? new LocalResolutionPinInvalidJsonError(cause) : new UnhandledException({ cause })
	});
}
async function writeLocalResolutionPin(cwd, pin, signal) {
	return Result.gen(async function* () {
		const prismaDir = path.join(cwd, ".prisma");
		yield* ensureLocalResolutionPinWriteNotAborted(signal);
		yield* Result.await(writeLocalResolutionPinBoundary(() => mkdir(prismaDir, { recursive: true }), "create-directory", signal));
		const pinPath = path.join(cwd, LOCAL_RESOLUTION_PIN_RELATIVE_PATH);
		const tmpPath = path.join(prismaDir, `local.${process.pid}.${Date.now()}.tmp`);
		const serialized = yield* serializeLocalResolutionPin(pin);
		yield* Result.await(writeLocalResolutionPinBoundary(() => writeFile(tmpPath, serialized, {
			encoding: "utf8",
			signal
		}), "write-temp-file", signal));
		yield* ensureLocalResolutionPinWriteNotAborted(signal);
		yield* Result.await(writeLocalResolutionPinBoundary(() => rename(tmpPath, pinPath), "rename-temp-file", signal));
		return Result.ok(void 0);
	});
}
async function ensureLocalResolutionPinGitignore(cwd, signal) {
	const gitignorePath = path.join(cwd, ".gitignore");
	let existing = null;
	const notAborted = ensureLocalResolutionPinGitignoreUpdateNotAborted(signal);
	if (notAborted.isErr()) return Result.err(notAborted.error);
	const existingResult = await Result.tryPromise({
		try: () => readFile(gitignorePath, {
			encoding: "utf8",
			signal
		}),
		catch: (cause) => signal?.aborted ? new LocalResolutionPinGitignoreUpdateAbortedError(cause) : new LocalResolutionPinGitignoreUpdateFailedError("read", cause)
	});
	if (existingResult.isErr()) if (existingResult.error instanceof LocalResolutionPinGitignoreUpdateFailedError && existingResult.error.cause.code === "ENOENT") existing = null;
	else return Result.err(existingResult.error);
	else existing = existingResult.value;
	if (existing === null) return writeLocalResolutionPinGitignore(gitignorePath, ".prisma/\n", signal);
	if (existing.split(/\r?\n/).map((line) => line.trim()).some((line) => line === ".prisma/" || line === ".prisma/local.json")) return Result.ok(void 0);
	return writeLocalResolutionPinGitignore(gitignorePath, existing.endsWith("\n") ? `${existing}.prisma/\n` : `${existing}\n.prisma/\n`, signal);
}
function ensureLocalResolutionPinWriteNotAborted(signal) {
	return Result.try({
		try: () => signal?.throwIfAborted(),
		catch: (cause) => new LocalResolutionPinWriteAbortedError(cause)
	});
}
function serializeLocalResolutionPin(pin) {
	return Result.try({
		try: () => `${JSON.stringify(pin, null, 2)}\n`,
		catch: (cause) => new LocalResolutionPinSerializationError(cause)
	});
}
function writeLocalResolutionPinBoundary(run, operation, signal) {
	return Result.tryPromise({
		try: async () => {
			await run();
		},
		catch: (cause) => signal?.aborted ? new LocalResolutionPinWriteAbortedError(cause) : new LocalResolutionPinWriteFailedError(operation, cause)
	});
}
function ensureLocalResolutionPinGitignoreUpdateNotAborted(signal) {
	return Result.try({
		try: () => signal?.throwIfAborted(),
		catch: (cause) => new LocalResolutionPinGitignoreUpdateAbortedError(cause)
	});
}
function writeLocalResolutionPinGitignore(gitignorePath, contents, signal) {
	return Result.tryPromise({
		try: () => writeFile(gitignorePath, contents, {
			encoding: "utf8",
			signal
		}),
		catch: (cause) => signal?.aborted ? new LocalResolutionPinGitignoreUpdateAbortedError(cause) : new LocalResolutionPinGitignoreUpdateFailedError("write", cause)
	});
}
function isLocalResolutionPin(value) {
	if (!value || typeof value !== "object") return false;
	const keys = Object.keys(value);
	if (keys.length !== 2 || !keys.includes("workspaceId") || !keys.includes("projectId")) return false;
	const candidate = value;
	return typeof candidate.workspaceId === "string" && candidate.workspaceId.trim().length > 0 && typeof candidate.projectId === "string" && candidate.projectId.trim().length > 0;
}
//#endregion
export { LOCAL_RESOLUTION_PIN_RELATIVE_PATH, ensureLocalResolutionPinGitignore, readLocalResolutionPin, writeLocalResolutionPin };
