// @prisma/cli/dist/lib/project/setup.js
import { CliError, usageError } from "../../shell/errors.js";
import { shortenHomePath } from "../fs/home-path.js";
import { LOCAL_RESOLUTION_PIN_RELATIVE_PATH, ensureLocalResolutionPinGitignore, writeLocalResolutionPin } from "./local-pin.js";
import "../../shell/command-arguments.js";
import { projectAmbiguousError, projectNotFoundError } from "./resolution.js";
import path from "node:path";
import { Result, matchError } from "better-result";
//#region src/lib/project/setup.ts
function isValidProjectSetupName(projectName) {
	return projectName.trim().length > 0;
}
function validateProjectSetupNameText(value, fallback) {
	if ((value?.trim() || fallback).trim().length > 0) return;
	return "Enter a Project name.";
}
function resolveProjectForSetup(projectRef, projects, workspace) {
	const matches = projects.filter((project) => project.id === projectRef || project.name === projectRef);
	if (matches.length === 1) return matches[0];
	if (matches.length > 1) throw projectAmbiguousError(projectRef, matches);
	throw projectNotFoundError(projectRef, workspace);
}
async function bindProjectToDirectory(context, workspace, project, action, directory = context.runtime.cwd) {
	return Result.gen(async function* () {
		yield* Result.await(writeLocalResolutionPin(directory, {
			workspaceId: workspace.id,
			projectId: project.id
		}, context.runtime.signal));
		yield* Result.await(ensureLocalResolutionPinGitignore(directory, context.runtime.signal));
		return Result.ok({
			workspace,
			project,
			directory: formatSetupDirectory(directory, context),
			localPin: {
				path: LOCAL_RESOLUTION_PIN_RELATIVE_PATH,
				written: true
			},
			action
		});
	});
}
function projectDirectoryBindingErrorToCliError(error) {
	return matchError(error, {
		LocalResolutionPinSerializationError: (error) => {
			throw error;
		},
		LocalResolutionPinWriteAbortedError: (error) => {
			throw error;
		},
		LocalResolutionPinWriteFailedError: (error) => localStateWriteFailedError(error, {
			why: `The CLI could not write ${LOCAL_RESOLUTION_PIN_RELATIVE_PATH}.`,
			meta: {
				pinPath: error.pinPath,
				operation: error.operation
			}
		}),
		LocalResolutionPinGitignoreUpdateAbortedError: (error) => {
			throw error;
		},
		LocalResolutionPinGitignoreUpdateFailedError: (error) => localStateWriteFailedError(error, {
			why: "The CLI could not update .gitignore to keep local Project binding state out of git.",
			meta: {
				gitignorePath: error.gitignorePath,
				operation: error.operation
			}
		})
	});
}
function localStateWriteFailedError(error, options) {
	return new CliError({
		code: "LOCAL_STATE_WRITE_FAILED",
		domain: "project",
		summary: "Could not save local Project binding",
		why: options.why,
		fix: "Check that this directory is writable and that .prisma/local.json and .gitignore are not blocked by directories or permissions, then retry.",
		debug: formatDebugDetails(error.cause),
		meta: options.meta,
		exitCode: 1,
		nextSteps: ["prisma-cli project link <id-or-name>", "prisma-cli app deploy --project <id-or-name>"]
	});
}
function toProjectSummary(project) {
	return {
		id: project.id,
		name: project.name,
		...project.url ? { url: project.url } : {}
	};
}
function projectSetupNameRequiredError(command) {
	return usageError("Project create requires a name", "The project name must be a non-empty value.", "Pass a Project name explicitly.", [`prisma-cli ${command} my-app`], "project");
}
function projectCreateFailedError(error, projectName, workspace, options) {
	const status = extractHttpStatus(error);
	if (status === 401 || status === 403) return new CliError({
		code: "PROJECT_CREATE_FAILED",
		domain: "project",
		summary: `Could not create Project "${projectName}"`,
		why: `The platform rejected the Project create in workspace "${workspace.name}" (HTTP ${status}).`,
		fix: options.permissionFix,
		debug: formatDebugDetails(error),
		exitCode: 1,
		nextSteps: options.nextSteps
	});
	return new CliError({
		code: "PROJECT_CREATE_FAILED",
		domain: "project",
		summary: `Could not create Project "${projectName}"`,
		why: error instanceof Error ? error.message : String(error),
		fix: options.fallbackFix,
		debug: formatDebugDetails(error),
		exitCode: 1,
		nextSteps: options.nextSteps
	});
}
function formatSetupDirectory(directory, context) {
	if (path.resolve(directory) !== path.resolve(context.runtime.cwd)) return shortenHomePath(directory, context.runtime.env);
	const basename = directory.split(/[\\/]/).filter(Boolean).pop();
	return basename ? `./${basename}` : ".";
}
function extractHttpStatus(error) {
	if (!error || typeof error !== "object") return null;
	const candidate = error;
	if (typeof candidate.statusCode === "number") return candidate.statusCode;
	if (typeof candidate.status === "number") return candidate.status;
	if (typeof candidate.message === "string") {
		const match = /\(HTTP (\d{3})\)/.exec(candidate.message);
		if (match) return Number.parseInt(match[1], 10);
	}
	return null;
}
function formatDebugDetails(error) {
	if (error instanceof Error) return error.stack ?? error.message;
	return typeof error === "string" ? error : null;
}
//#endregion
export { bindProjectToDirectory, isValidProjectSetupName, projectCreateFailedError, projectDirectoryBindingErrorToCliError, projectSetupNameRequiredError, resolveProjectForSetup, toProjectSummary, validateProjectSetupNameText };
