// @prisma/cli/dist/commands/database/index.js
import { attachCommandDescriptor } from "../../shell/command-meta.js";
import { addCompactGlobalFlags, addGlobalFlags } from "../../shell/global-flags.js";
import { configureRuntimeCommand } from "../../shell/runtime.js";
import { runCommand } from "../../shell/command-runner.js";
import { runDatabaseConnectionCreate, runDatabaseConnectionList, runDatabaseConnectionRemove, runDatabaseCreate, runDatabaseList, runDatabaseRemove, runDatabaseShow } from "../../controllers/database.js";
import { renderDatabaseConnectionCreate, renderDatabaseConnectionCreateStdout, renderDatabaseConnectionList, renderDatabaseConnectionRemove, renderDatabaseCreate, renderDatabaseCreateStdout, renderDatabaseList, renderDatabaseRemove, renderDatabaseShow, serializeDatabaseConnectionCreate, serializeDatabaseConnectionList, serializeDatabaseConnectionRemove, serializeDatabaseCreate, serializeDatabaseList, serializeDatabaseRemove, serializeDatabaseShow } from "../../presenters/database.js";
import { Command, Option } from "commander";
//#region src/commands/database/index.ts
function createDatabaseCommand(runtime) {
	const database = attachCommandDescriptor(configureRuntimeCommand(new Command("database"), runtime), "database");
	addCompactGlobalFlags(database);
	database.addCommand(createDatabaseListCommand(runtime));
	database.addCommand(createDatabaseShowCommand(runtime));
	database.addCommand(createDatabaseCreateCommand(runtime));
	database.addCommand(createDatabaseRemoveCommand(runtime));
	database.addCommand(createDatabaseConnectionCommand(runtime));
	return database;
}
function addProjectAndBranchOptions(command) {
	return command.addOption(new Option("--project <id-or-name>", "Project id or name")).addOption(new Option("--branch <git-name>", "Branch git name"));
}
function createDatabaseListCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("list"), runtime), "database.list");
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		await runCommand(runtime, "database.list", options, (context) => runDatabaseList(context, {
			projectRef,
			branchName
		}), {
			renderHuman: (context, descriptor, result) => renderDatabaseList(context, descriptor, result),
			renderJson: (result) => serializeDatabaseList(result)
		});
	});
	return command;
}
function createDatabaseShowCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("show"), runtime), "database.show");
	command.argument("<database>", "Database id or name");
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (databaseRef, options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		await runCommand(runtime, "database.show", options, (context) => runDatabaseShow(context, databaseRef, {
			projectRef,
			branchName
		}), {
			renderHuman: (context, descriptor, result) => renderDatabaseShow(context, descriptor, result),
			renderJson: (result) => serializeDatabaseShow(result)
		});
	});
	return command;
}
function createDatabaseCreateCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("create"), runtime), "database.create");
	command.argument("<name>", "Database name").addOption(new Option("--region <region>", "Prisma Postgres region id"));
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (name, options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		const region = options.region;
		await runCommand(runtime, "database.create", options, (context) => runDatabaseCreate(context, name, {
			projectRef,
			branchName,
			region
		}), {
			renderStdout: (context, descriptor, result) => renderDatabaseCreateStdout(context, descriptor, result),
			renderHuman: (context, descriptor, result) => renderDatabaseCreate(context, descriptor, result),
			renderJson: (result) => serializeDatabaseCreate(result)
		});
	});
	return command;
}
function createDatabaseRemoveCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("remove"), runtime), "database.remove");
	command.argument("<database>", "Database id or name").addOption(new Option("--confirm <database-id>", "Exact database id required to remove"));
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (databaseRef, options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		const confirm = options.confirm;
		await runCommand(runtime, "database.remove", options, (context) => runDatabaseRemove(context, databaseRef, {
			projectRef,
			branchName,
			confirm
		}), {
			renderHuman: (context, descriptor, result) => renderDatabaseRemove(context, descriptor, result),
			renderJson: (result) => serializeDatabaseRemove(result)
		});
	});
	return command;
}
function createDatabaseConnectionCommand(runtime) {
	const connection = attachCommandDescriptor(configureRuntimeCommand(new Command("connection"), runtime), "database.connection");
	addCompactGlobalFlags(connection);
	connection.addCommand(createDatabaseConnectionListCommand(runtime));
	connection.addCommand(createDatabaseConnectionCreateCommand(runtime));
	connection.addCommand(createDatabaseConnectionRemoveCommand(runtime));
	return connection;
}
function createDatabaseConnectionListCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("list"), runtime), "database.connection.list");
	command.argument("<database>", "Database id or name");
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (databaseRef, options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		await runCommand(runtime, "database.connection.list", options, (context) => runDatabaseConnectionList(context, databaseRef, {
			projectRef,
			branchName
		}), {
			renderHuman: (context, descriptor, result) => renderDatabaseConnectionList(context, descriptor, result),
			renderJson: (result) => serializeDatabaseConnectionList(result)
		});
	});
	return command;
}
function createDatabaseConnectionCreateCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("create"), runtime), "database.connection.create");
	command.argument("<database>", "Database id or name").addOption(new Option("--name <name>", "Connection name"));
	addProjectAndBranchOptions(command);
	addGlobalFlags(command);
	command.action(async (databaseRef, options) => {
		const projectRef = options.project;
		const branchName = options.branch;
		const name = options.name;
		await runCommand(runtime, "database.connection.create", options, (context) => runDatabaseConnectionCreate(context, databaseRef, {
			projectRef,
			branchName,
			name
		}), {
			renderStdout: (context, descriptor, result) => renderDatabaseConnectionCreateStdout(context, descriptor, result),
			renderHuman: (context, descriptor, result) => renderDatabaseConnectionCreate(context, descriptor, result),
			renderJson: (result) => serializeDatabaseConnectionCreate(result)
		});
	});
	return command;
}
function createDatabaseConnectionRemoveCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("remove"), runtime), "database.connection.remove");
	command.argument("<connection>", "Connection id").addOption(new Option("--confirm <connection-id>", "Exact connection id required to remove"));
	addGlobalFlags(command);
	command.action(async (connectionRef, options) => {
		const confirm = options.confirm;
		await runCommand(runtime, "database.connection.remove", options, (context) => runDatabaseConnectionRemove(context, connectionRef, { confirm }), {
			renderHuman: (context, descriptor, result) => renderDatabaseConnectionRemove(context, descriptor, result),
			renderJson: (result) => serializeDatabaseConnectionRemove(result)
		});
	});
	return command;
}
//#endregion
export { createDatabaseCommand };
