// @prisma/cli/dist/adapters/git.js
import { execFile } from "node:child_process";
import { promisify } from "node:util";
//#region src/adapters/git.ts
const execFileAsync = promisify(execFile);
async function readGitOriginRemote(cwd, signal) {
	try {
		const { stdout } = await execFileAsync("git", [
			"config",
			"--get",
			"remote.origin.url"
		], {
			cwd,
			timeout: 5e3,
			signal
		});
		const remote = stdout.trim();
		return remote.length > 0 ? remote : null;
	} catch (error) {
		if (signal?.aborted || isAbortError(error)) throw error;
		return null;
	}
}
function isAbortError(error) {
	return error instanceof Error && error.name === "AbortError";
}
function parseGitHubRepositoryUrl(value) {
	const input = value.trim();
	const shorthand = input.match(/^git@github\.com:([^/\s]+)\/([^/\s]+?)(?:\.git)?$/);
	if (shorthand) return toGitHubRepositoryReference(shorthand[1], shorthand[2]);
	let parsed;
	try {
		parsed = new URL(input);
	} catch {
		return null;
	}
	if (parsed.hostname !== "github.com") return null;
	if (parsed.protocol !== "https:" && parsed.protocol !== "http:" && parsed.protocol !== "ssh:") return null;
	const parts = parsed.pathname.split("/").filter(Boolean);
	if (parts.length !== 2) return null;
	const [owner, rawName] = parts;
	return toGitHubRepositoryReference(owner, rawName.endsWith(".git") ? rawName.slice(0, -4) : rawName);
}
function toGitHubRepositoryReference(owner, name) {
	if (!owner || !name || owner.includes("/") || name.includes("/")) return null;
	return {
		provider: "github",
		owner,
		name,
		fullName: `${owner}/${name}`,
		url: `https://github.com/${owner}/${name}`
	};
}
//#endregion
export { parseGitHubRepositoryUrl, readGitOriginRemote };
