// @prisma/cli/dist/adapters/local-state.js
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
//#region src/adapters/local-state.ts
const DEFAULT_STATE = {
	auth: null,
	project: {
		rememberedByWorkspace: {},
		lastResolved: null,
		repositoryConnectionsByProject: {}
	},
	branch: { active: "preview" },
	app: {
		selectedByProject: {},
		knownLiveDeploymentByProject: {}
	}
};
const DEFAULT_STATE_FILE_NAME = "state.json";
function resolveLocalStateFilePath(stateDir) {
	return path.join(stateDir, DEFAULT_STATE_FILE_NAME);
}
var LocalStateStore = class {
	stateFilePath;
	constructor(stateDir, signal) {
		this.signal = signal;
		this.stateFilePath = resolveLocalStateFilePath(stateDir);
	}
	async read() {
		this.signal?.throwIfAborted();
		try {
			const raw = await readFile(this.stateFilePath, {
				encoding: "utf8",
				signal: this.signal
			});
			const parsed = JSON.parse(raw);
			return {
				auth: parsed.auth ?? structuredClone(DEFAULT_STATE.auth),
				project: {
					rememberedByWorkspace: parsed.project?.rememberedByWorkspace ?? {},
					lastResolved: parsed.project?.lastResolved ?? null,
					repositoryConnectionsByProject: parsed.project?.repositoryConnectionsByProject ?? {}
				},
				branch: { active: parsed.branch?.active ?? DEFAULT_STATE.branch.active },
				app: {
					selectedByProject: parsed.app?.selectedByProject ?? {},
					knownLiveDeploymentByProject: parsed.app?.knownLiveDeploymentByProject ?? {}
				}
			};
		} catch (error) {
			if (error.code === "ENOENT") return structuredClone(DEFAULT_STATE);
			throw error;
		}
	}
	async write(state) {
		this.signal?.throwIfAborted();
		await mkdir(path.dirname(this.stateFilePath), { recursive: true });
		this.signal?.throwIfAborted();
		await writeFile(this.stateFilePath, `${JSON.stringify(state, null, 2)}\n`, { encoding: "utf8" });
		this.signal?.throwIfAborted();
	}
	async setAuthSession(session) {
		const state = await this.read();
		state.auth = session;
		await this.write(state);
		return state;
	}
	async clearAuthSession() {
		const state = await this.read();
		state.auth = null;
		await this.write(state);
		return state;
	}
	async setActiveBranch(active) {
		const state = await this.read();
		state.branch.active = active;
		await this.write(state);
		return state;
	}
	async readRememberedProject(workspaceId) {
		return (await this.read()).project.rememberedByWorkspace[workspaceId] ?? null;
	}
	async readLastResolvedProject() {
		return (await this.read()).project.lastResolved;
	}
	async setRememberedProject(project) {
		const state = await this.read();
		state.project.rememberedByWorkspace[project.workspaceId] = project;
		state.project.lastResolved = project;
		await this.write(state);
		return state;
	}
	async readRepositoryConnection(projectId) {
		return (await this.read()).project.repositoryConnectionsByProject[projectId] ?? null;
	}
	async setRepositoryConnection(projectId, connection) {
		const state = await this.read();
		state.project.repositoryConnectionsByProject[projectId] = connection;
		await this.write(state);
		return state;
	}
	async clearRepositoryConnection(projectId) {
		const state = await this.read();
		delete state.project.repositoryConnectionsByProject[projectId];
		await this.write(state);
		return state;
	}
	async readSelectedApp(projectId) {
		return (await this.read()).app.selectedByProject[projectId] ?? null;
	}
	async setSelectedApp(projectId, app) {
		const state = await this.read();
		state.app.selectedByProject[projectId] = app;
		await this.write(state);
		return state;
	}
	async clearSelectedApp(projectId, appId) {
		const state = await this.read();
		const selectedApp = state.app.selectedByProject[projectId];
		if (!selectedApp || selectedApp.id !== appId) return state;
		delete state.app.selectedByProject[projectId];
		await this.write(state);
		return state;
	}
	async readKnownLiveDeployment(projectId, appId) {
		return (await this.read()).app.knownLiveDeploymentByProject[projectId]?.[appId] ?? null;
	}
	async setKnownLiveDeployment(projectId, appId, deploymentId) {
		const state = await this.read();
		state.app.knownLiveDeploymentByProject[projectId] ??= {};
		state.app.knownLiveDeploymentByProject[projectId][appId] = deploymentId;
		await this.write(state);
		return state;
	}
	async clearKnownLiveDeployment(projectId, appId) {
		const state = await this.read();
		const projectDeployments = state.app.knownLiveDeploymentByProject[projectId];
		if (!projectDeployments || !(appId in projectDeployments)) return state;
		delete projectDeployments[appId];
		if (Object.keys(projectDeployments).length === 0) delete state.app.knownLiveDeploymentByProject[projectId];
		await this.write(state);
		return state;
	}
};
//#endregion
export { LocalStateStore, resolveLocalStateFilePath };
