// @prisma/cli/dist/adapters/mock-api.js
import { readFile } from "node:fs/promises";
//#region src/adapters/mock-api.ts
var MockApi = class MockApi {
	data;
	constructor(data) {
		this.data = data;
	}
	static async load(fixturePath, signal) {
		signal?.throwIfAborted();
		const raw = await readFile(fixturePath, {
			encoding: "utf8",
			signal
		});
		return new MockApi(JSON.parse(raw));
	}
	listProviders() {
		return this.data.providers;
	}
	getProvider(providerId) {
		return this.data.providers.find((provider) => provider.id === providerId);
	}
	listUsersForProvider(providerId) {
		return this.data.users.filter((user) => user.providerIds.includes(providerId));
	}
	getUser(userId) {
		return this.data.users.find((user) => user.id === userId);
	}
	getUserForProvider(providerId, userId) {
		return this.listUsersForProvider(providerId).find((user) => user.id === userId);
	}
	listUserWorkspaces(userId) {
		const workspaceIds = this.data.memberships.filter((membership) => membership.userId === userId).map((membership) => membership.workspaceId);
		return this.data.workspaces.filter((workspace) => workspaceIds.includes(workspace.id));
	}
	getWorkspace(workspaceId) {
		return this.data.workspaces.find((workspace) => workspace.id === workspaceId);
	}
	getUserWorkspace(userId, workspaceId) {
		return this.listUserWorkspaces(userId).find((workspace) => workspace.id === workspaceId);
	}
	listProjectsForWorkspace(workspaceId) {
		return this.data.projects.filter((project) => project.workspaceId === workspaceId);
	}
	getProject(projectId) {
		return this.data.projects.find((project) => project.id === projectId);
	}
	getProjectForWorkspace(workspaceId, projectId) {
		return this.listProjectsForWorkspace(workspaceId).find((project) => project.id === projectId);
	}
	listBranchesForProject(projectId) {
		return this.data.branches.filter((branch) => branch.projectId === projectId);
	}
	getBranchForProject(projectId, name) {
		return this.listBranchesForProject(projectId).find((branch) => branch.name === name);
	}
	getDeployment(deploymentId) {
		return this.data.deployments.find((deployment) => deployment.id === deploymentId);
	}
	listDatabasesForProject(projectId, branchName) {
		return (this.data.databases ?? []).filter((database) => database.projectId === projectId && (!branchName || database.branchName === branchName));
	}
	getDatabase(databaseId) {
		return (this.data.databases ?? []).find((database) => database.id === databaseId);
	}
	createDatabase(input) {
		this.data.databases ??= [];
		this.data.databaseConnections ??= [];
		const database = {
			id: `db_${this.data.databases.length + 1e3}`,
			projectId: input.projectId,
			branchId: input.branchName ? this.getBranchForProject(input.projectId, input.branchName)?.id ?? null : null,
			branchName: input.branchName ?? null,
			name: input.name,
			region: input.region ?? null,
			status: "ready",
			isDefault: false,
			createdAt: "2026-06-09T00:00:00.000Z"
		};
		const connectionString = `postgresql://${database.id}.example.prisma.io/postgres`;
		const connection = {
			id: `conn_${this.data.databaseConnections.length + 1e3}`,
			databaseId: database.id,
			name: "primary",
			createdAt: "2026-06-09T00:00:00.000Z",
			connectionString
		};
		this.data.databases.push(database);
		this.data.databaseConnections.push(connection);
		return {
			database,
			connection,
			connectionString
		};
	}
	removeDatabase(databaseId) {
		this.data.databases ??= [];
		this.data.databaseConnections ??= [];
		const database = this.getDatabase(databaseId);
		if (!database) return;
		this.data.databases = this.data.databases.filter((candidate) => candidate.id !== databaseId);
		this.data.databaseConnections = this.data.databaseConnections.filter((connection) => connection.databaseId !== databaseId);
		return database;
	}
	listDatabaseConnections(databaseId) {
		return (this.data.databaseConnections ?? []).filter((connection) => connection.databaseId === databaseId);
	}
	getDatabaseConnection(connectionId) {
		return (this.data.databaseConnections ?? []).find((connection) => connection.id === connectionId);
	}
	createDatabaseConnection(input) {
		if (!this.getDatabase(input.databaseId)) return;
		this.data.databaseConnections ??= [];
		const connectionString = `postgresql://${input.databaseId}-${this.data.databaseConnections.length + 1}.example.prisma.io/postgres`;
		const connection = {
			id: `conn_${this.data.databaseConnections.length + 1e3}`,
			databaseId: input.databaseId,
			name: input.name,
			createdAt: "2026-06-09T00:00:00.000Z",
			connectionString
		};
		this.data.databaseConnections.push(connection);
		return {
			connection,
			connectionString
		};
	}
	removeDatabaseConnection(connectionId) {
		this.data.databaseConnections ??= [];
		const connection = this.getDatabaseConnection(connectionId);
		if (!connection) return;
		this.data.databaseConnections = this.data.databaseConnections.filter((candidate) => candidate.id !== connectionId);
		return connection;
	}
};
//#endregion
export { MockApi };
