// @prisma/cli/dist/commands/auth/index.js
import { attachCommandDescriptor } from "../../shell/command-meta.js";
import { addCompactGlobalFlags, addGlobalFlags } from "../../shell/global-flags.js";
import { configureRuntimeCommand } from "../../shell/runtime.js";
import { runAuthLogin, runAuthLogout, runAuthWhoAmI, runAuthWorkspaceList, runAuthWorkspaceLogout, runAuthWorkspaceUse } from "../../controllers/auth.js";
import { runCommand } from "../../shell/command-runner.js";
import { renderAuthSuccess, renderAuthWorkspaceList, renderAuthWorkspaceLogout, renderAuthWorkspaceUse, serializeAuthWorkspaceList, serializeAuthWorkspaceLogout, serializeAuthWorkspaceUse } from "../../presenters/auth.js";
import { Command, Option } from "commander";
//#region src/commands/auth/index.ts
function createAuthCommand(runtime) {
	const auth = attachCommandDescriptor(configureRuntimeCommand(new Command("auth"), runtime), "auth");
	addCompactGlobalFlags(auth);
	auth.addCommand(createAuthLoginCommand(runtime));
	auth.addCommand(createAuthLogoutCommand(runtime));
	auth.addCommand(createAuthWhoAmICommand(runtime));
	auth.addCommand(createAuthWorkspaceCommand(runtime));
	return auth;
}
function createAuthLoginCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("login"), runtime), "auth.login");
	command.addOption(new Option("--provider <provider>").hideHelp()).addOption(new Option("--user <id>").hideHelp()).addOption(new Option("--workspace <id>").hideHelp());
	addGlobalFlags(command);
	command.action(async (options) => {
		await runCommand(runtime, "auth.login", options, (context) => runAuthLogin(context, options), { renderHuman: (context, descriptor, result) => renderAuthSuccess(context, descriptor, "auth.login", result) });
	});
	return command;
}
function createAuthLogoutCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("logout"), runtime), "auth.logout");
	command.addOption(new Option("--workspace <id-or-name>", "Remove one stored OAuth workspace session"));
	addGlobalFlags(command);
	command.action(async (options) => {
		const logoutOptions = options;
		if (logoutOptions.workspace?.trim()) {
			await runCommand(runtime, "auth.workspace.logout", options, (context) => runAuthWorkspaceLogout(context, logoutOptions.workspace), {
				renderHuman: (context, descriptor, result) => renderAuthWorkspaceLogout(context, descriptor, result),
				renderJson: (result) => serializeAuthWorkspaceLogout(result)
			});
			return;
		}
		await runCommand(runtime, "auth.logout", options, (context) => runAuthLogout(context), { renderHuman: (context, descriptor, result) => renderAuthSuccess(context, descriptor, "auth.logout", result) });
	});
	return command;
}
function createAuthWhoAmICommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("whoami"), runtime), "auth.whoami");
	addGlobalFlags(command);
	command.action(async (options) => {
		await runCommand(runtime, "auth.whoami", options, (context) => runAuthWhoAmI(context), { renderHuman: (context, descriptor, result) => renderAuthSuccess(context, descriptor, "auth.whoami", result) });
	});
	return command;
}
function createAuthWorkspaceCommand(runtime) {
	const workspace = attachCommandDescriptor(configureRuntimeCommand(new Command("workspace"), runtime), "auth.workspace");
	addCompactGlobalFlags(workspace);
	workspace.addCommand(createAuthWorkspaceListCommand(runtime));
	workspace.addCommand(createAuthWorkspaceUseCommand(runtime));
	workspace.addCommand(createAuthWorkspaceLogoutCommand(runtime));
	return workspace;
}
function createAuthWorkspaceListCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("list"), runtime), "auth.workspace.list");
	addGlobalFlags(command);
	command.action(async (options) => {
		await runCommand(runtime, "auth.workspace.list", options, (context) => runAuthWorkspaceList(context), {
			renderHuman: (context, descriptor, result) => renderAuthWorkspaceList(context, descriptor, result),
			renderJson: (result) => serializeAuthWorkspaceList(result)
		});
	});
	return command;
}
function createAuthWorkspaceLogoutCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("logout"), runtime), "auth.workspace.logout");
	command.argument("<id-or-name>", "Workspace id or exact name");
	addGlobalFlags(command);
	command.action(async (workspaceRef, options) => {
		await runCommand(runtime, "auth.workspace.logout", options, (context) => runAuthWorkspaceLogout(context, typeof workspaceRef === "string" ? workspaceRef : void 0), {
			renderHuman: (context, descriptor, result) => renderAuthWorkspaceLogout(context, descriptor, result),
			renderJson: (result) => serializeAuthWorkspaceLogout(result)
		});
	});
	return command;
}
function createAuthWorkspaceUseCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("use"), runtime), "auth.workspace.use");
	command.argument("[id-or-name]", "Workspace id or exact name");
	addGlobalFlags(command);
	command.action(async (workspaceRef, options) => {
		await runCommand(runtime, "auth.workspace.use", options, (context) => runAuthWorkspaceUse(context, typeof workspaceRef === "string" ? workspaceRef : void 0), {
			renderHuman: (context, descriptor, result) => renderAuthWorkspaceUse(context, descriptor, result),
			renderJson: (result) => serializeAuthWorkspaceUse(result)
		});
	});
	return command;
}
//#endregion
export { createAuthCommand };
