// @prisma/cli/dist/commands/env.js
import { attachCommandDescriptor } from "../shell/command-meta.js";
import { addGlobalFlags } from "../shell/global-flags.js";
import { configureRuntimeCommand } from "../shell/runtime.js";
import { runCommand } from "../shell/command-runner.js";
import { runEnvAdd, runEnvList, runEnvRemove, runEnvUpdate } from "../controllers/app-env.js";
import { renderEnvAdd, renderEnvList, renderEnvRm, renderEnvUpdate, serializeEnvAdd, serializeEnvList, serializeEnvRm, serializeEnvUpdate } from "../presenters/app-env.js";
import { Command, Option } from "commander";
//#region src/commands/env.ts
function createEnvCommand(runtime) {
	const env = attachCommandDescriptor(configureRuntimeCommand(new Command("env"), runtime), "project.env");
	env.description("Manage environment variables for the active project");
	env.addCommand(createEnvAddCommand(runtime));
	env.addCommand(createEnvUpdateCommand(runtime));
	env.addCommand(createEnvListCommand(runtime));
	env.addCommand(createEnvRemoveCommand(runtime));
	return env;
}
function createEnvAddCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("add"), runtime), "project.env.add");
	command.argument("[assignment]", "Variable assignment as KEY=VALUE or KEY from the current environment").addOption(new Option("--file <path>", "Read KEY=VALUE assignments from a dotenv file")).addOption(new Option("--role <role>", "Project template scope (production or preview)").choices(["production", "preview"])).addOption(new Option("--branch <git-name>", "Preview branch override scope")).addOption(new Option("--project <id-or-name>", "Project id or name"));
	addGlobalFlags(command);
	command.action(async (assignment, options) => {
		const roleName = options.role;
		const branchName = options.branch;
		const projectRef = options.project;
		const filePath = options.file;
		await runCommand(runtime, "project.env.add", options, (context) => runEnvAdd(context, assignment, {
			roleName,
			branchName,
			projectRef,
			filePath
		}), {
			renderHuman: (context, descriptor, result) => renderEnvAdd(context, descriptor, result),
			renderJson: (result) => serializeEnvAdd(result)
		});
	});
	return command;
}
function createEnvUpdateCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("update"), runtime), "project.env.update");
	command.argument("[assignment]", "Variable assignment as KEY=VALUE or KEY from the current environment").addOption(new Option("--file <path>", "Read KEY=VALUE assignments from a dotenv file")).addOption(new Option("--role <role>", "Project template scope (production or preview)").choices(["production", "preview"])).addOption(new Option("--branch <git-name>", "Preview branch override scope")).addOption(new Option("--project <id-or-name>", "Project id or name"));
	addGlobalFlags(command);
	command.action(async (assignment, options) => {
		const roleName = options.role;
		const branchName = options.branch;
		const projectRef = options.project;
		const filePath = options.file;
		await runCommand(runtime, "project.env.update", options, (context) => runEnvUpdate(context, assignment, {
			roleName,
			branchName,
			projectRef,
			filePath
		}), {
			renderHuman: (context, descriptor, result) => renderEnvUpdate(context, descriptor, result),
			renderJson: (result) => serializeEnvUpdate(result)
		});
	});
	return command;
}
function createEnvListCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("list"), runtime), "project.env.list");
	command.addOption(new Option("--role <role>", "Project template scope").choices(["production", "preview"])).addOption(new Option("--branch <git-name>", "Preview branch resolved scope")).addOption(new Option("--project <id-or-name>", "Project id or name"));
	addGlobalFlags(command);
	command.action(async (options) => {
		const roleName = options.role;
		const branchName = options.branch;
		const projectRef = options.project;
		await runCommand(runtime, "project.env.list", options, (context) => runEnvList(context, {
			roleName,
			branchName,
			projectRef
		}), {
			renderHuman: (context, descriptor, result) => renderEnvList(context, descriptor, result),
			renderJson: (result) => serializeEnvList(result)
		});
	});
	return command;
}
function createEnvRemoveCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("remove"), runtime), "project.env.remove");
	command.alias("rm").argument("<key>", "Variable key to remove").addOption(new Option("--role <role>", "Project template scope (production or preview)").choices(["production", "preview"])).addOption(new Option("--branch <git-name>", "Preview branch override scope")).addOption(new Option("--project <id-or-name>", "Project id or name"));
	addGlobalFlags(command);
	command.action(async (key, options) => {
		const roleName = options.role;
		const branchName = options.branch;
		const projectRef = options.project;
		await runCommand(runtime, "project.env.remove", options, (context) => runEnvRemove(context, key, {
			roleName,
			branchName,
			projectRef
		}), {
			renderHuman: (context, descriptor, result) => renderEnvRm(context, descriptor, result),
			renderJson: (result) => serializeEnvRm(result)
		});
	});
	return command;
}
//#endregion
export { createEnvCommand };
