// @prisma/cli/dist/lib/app/branch-database-api.js
//#region src/lib/app/branch-database-api.ts
async function createBranchDatabase(client, options) {
	const result = await client.POST("/v1/databases", {
		body: {
			projectId: options.projectId,
			branchId: options.branchId,
			name: options.branchName,
			source: { type: "empty" }
		},
		signal: options.signal
	});
	if (result.error || !result.data) throw apiCallError(`Failed to create database for branch "${options.branchName}"`, result.response, result.error);
	return normalizeBranchDatabaseRecord(result.data.data);
}
async function listEnvironmentVariables(client, options) {
	const variables = [];
	let cursor;
	while (true) {
		const result = await client.GET("/v1/environment-variables", {
			params: { query: {
				projectId: options.projectId,
				class: options.className,
				key: options.key,
				branchId: options.branchId,
				cursor
			} },
			signal: options.signal
		});
		if (result.error || !result.data) throw apiCallError("Failed to list environment variables", result.response, result.error);
		variables.push(...result.data.data);
		if (!result.data.pagination.hasMore || !result.data.pagination.nextCursor) break;
		cursor = result.data.pagination.nextCursor;
	}
	return variables.map((variable) => normalizeEnvironmentVariable(variable));
}
async function createEnvironmentVariable(client, options) {
	const result = await client.POST("/v1/environment-variables", {
		body: {
			projectId: options.projectId,
			class: options.className,
			key: options.key,
			value: options.value,
			...options.branchId ? { branchId: options.branchId } : {}
		},
		signal: options.signal
	});
	if (result.error || !result.data) throw apiCallError(`Failed to add ${options.key}`, result.response, result.error);
	return normalizeEnvironmentVariable(result.data.data);
}
async function deleteBranchDatabase(client, options) {
	const result = await client.DELETE("/v1/databases/{databaseId}", {
		params: { path: { databaseId: options.databaseId } },
		signal: options.signal
	});
	if (result.error) throw apiCallError("Failed to delete branch database", result.response, result.error);
}
async function updateEnvironmentVariable(client, options) {
	const result = await client.PATCH("/v1/environment-variables/{envVarId}", {
		params: { path: { envVarId: options.envVarId } },
		body: { value: options.value },
		signal: options.signal
	});
	if (result.error || !result.data) throw apiCallError("Failed to update environment variable", result.response, result.error);
	return normalizeEnvironmentVariable(result.data.data);
}
async function deleteEnvironmentVariable(client, options) {
	const result = await client.DELETE("/v1/environment-variables/{envVarId}", {
		params: { path: { envVarId: options.envVarId } },
		signal: options.signal
	});
	if (result.error) throw apiCallError("Failed to delete environment variable", result.response, result.error);
}
function normalizeEnvironmentVariable(variable) {
	return {
		id: variable.id,
		key: variable.key,
		branchId: variable.branchId,
		className: variable.class,
		isManagedBySystem: variable.isManagedBySystem
	};
}
function normalizeBranchDatabaseRecord(database) {
	const connection = database.connections?.[0];
	const databaseUrl = connection?.endpoints?.pooled?.connectionString;
	const directUrl = connection?.endpoints?.direct?.connectionString ?? null;
	if (!databaseUrl) throw new Error("Created database did not return a pooled connection string.");
	return {
		id: database.id,
		name: database.name,
		branchId: database.branchId,
		databaseUrl,
		directUrl
	};
}
function apiCallError(summary, response, error) {
	if (response.status === 404) return /* @__PURE__ */ new Error("Resource Not Found");
	const message = error.error?.message ?? `Management API returned HTTP ${response.status}.`;
	const hint = error.error?.hint ? ` ${error.error.hint}` : "";
	return /* @__PURE__ */ new Error(`${summary}: ${message}${hint}`);
}
//#endregion
export { createBranchDatabase, createEnvironmentVariable, deleteBranchDatabase, deleteEnvironmentVariable, listEnvironmentVariables, updateEnvironmentVariable };
