// @prisma/cli/dist/lib/app/build.js
import "./build-settings.js";
import "node:fs/promises";
import "node:path";
import { normalizeArtifactSymlinks, resolveBuildStrategy } from "@prisma/compute-sdk";
//#region src/lib/app/build.ts
const APP_BUILD_TYPES = [
	"auto",
	"bun",
	"nextjs",
	"nuxt",
	"astro",
	"nestjs",
	"tanstack-start"
];
const RESOLVED_APP_BUILD_TYPES = APP_BUILD_TYPES.filter((buildType) => buildType !== "auto");
var AppBuildStrategy = class {
	#appPath;
	#entrypoint;
	#buildType;
	#signal;
	#buildSettings;
	constructor(options) {
		this.#appPath = options.appPath;
		this.#entrypoint = options.entrypoint;
		this.#buildType = options.buildType ?? "auto";
		this.#signal = options.signal;
		this.#buildSettings = options.buildSettings;
	}
	async canBuild(signal = this.#signal) {
		const { strategy } = await resolveAppBuildStrategy({
			appPath: this.#appPath,
			entrypoint: this.#entrypoint,
			buildType: this.#buildType,
			signal,
			buildSettings: this.#buildSettings
		});
		return strategy.canBuild(signal);
	}
	async execute(signal = this.#signal) {
		const { artifact } = await executeAppBuild({
			appPath: this.#appPath,
			entrypoint: this.#entrypoint,
			buildType: this.#buildType,
			signal,
			buildSettings: this.#buildSettings
		});
		return artifact;
	}
};
async function executeAppBuild(options) {
	const { strategy, buildType } = await resolveAppBuildStrategy({
		appPath: options.appPath,
		entrypoint: options.entrypoint,
		buildType: options.buildType ?? "auto",
		signal: options.signal,
		buildSettings: options.buildSettings
	});
	const artifact = await strategy.execute(options.signal);
	try {
		await normalizeArtifactSymlinks(artifact.directory, options.appPath, options.signal);
		return {
			artifact,
			buildType
		};
	} catch (error) {
		await artifact.cleanup?.().catch(() => void 0);
		throw error;
	}
}
async function resolveAppBuildStrategy(options) {
	return resolveBuildStrategy({
		appPath: options.appPath,
		buildType: options.buildType,
		entrypoint: options.entrypoint,
		buildSettings: options.buildSettings,
		signal: options.signal
	});
}
//#endregion
export { APP_BUILD_TYPES, AppBuildStrategy, RESOLVED_APP_BUILD_TYPES, executeAppBuild };
