// @prisma/cli/dist/lib/app/build-settings.js
import { readFile } from "node:fs/promises";
import path from "node:path";
import { resolveBuildSettings, resolveConfiguredBuildSettings } from "@prisma/compute-sdk";
//#region src/lib/app/build-settings.ts
/** Legacy build-settings file: no longer read or written, only detected for migration. */
const PRISMA_APP_CONFIG_FILENAME = "prisma.app.json";
/**
* Detects a leftover `prisma.app.json`. The file is no longer used: one that
* matches the effective settings is reported for deletion, one with custom
* values must be migrated to the compute config so builds never silently
* change.
*/
async function detectLegacyBuildSettings(options) {
	const configPath = path.join(options.appPath, PRISMA_APP_CONFIG_FILENAME);
	let content;
	try {
		options.signal?.throwIfAborted();
		content = await readFile(configPath, {
			encoding: "utf8",
			signal: options.signal
		});
	} catch (error) {
		if (options.signal?.aborted) throw error;
		return { kind: "absent" };
	}
	let legacy;
	try {
		const parsed = JSON.parse(content);
		const buildCommand = parsed.buildCommand === null || typeof parsed.buildCommand === "string" ? typeof parsed.buildCommand === "string" ? parsed.buildCommand.trim() || null : null : void 0;
		const outputDirectory = typeof parsed.outputDirectory === "string" ? normalizeRelativePath(parsed.outputDirectory) : void 0;
		if (buildCommand === void 0 || !outputDirectory) return {
			kind: "invalid",
			configPath
		};
		legacy = {
			buildCommand,
			outputDirectory
		};
	} catch {
		return {
			kind: "invalid",
			configPath
		};
	}
	return legacy.buildCommand === options.effective.buildCommand && legacy.outputDirectory === options.effective.outputDirectory ? {
		kind: "matching",
		configPath
	} : {
		kind: "custom",
		configPath,
		...legacy
	};
}
/** Resolves build settings purely from framework inference; nothing is read or written. */
async function resolveInferredAppBuildSettings(options) {
	return {
		status: "inferred",
		configPath: null,
		relativeConfigPath: null,
		settings: await resolveBuildSettings(options)
	};
}
/**
* Resolves build settings when the compute config owns them: configured
* fields win, omitted fields fall back to framework defaults.
*/
async function resolveConfiguredAppBuildSettings(options) {
	const configFilename = path.basename(options.configPath);
	const settings = await resolveConfiguredBuildSettings({
		appPath: options.appPath,
		buildType: options.buildType,
		configured: options.configured,
		source: `set by ${configFilename}`,
		signal: options.signal
	});
	return {
		status: "config",
		configPath: options.configPath,
		relativeConfigPath: configFilename,
		settings
	};
}
function normalizeRelativePath(value) {
	const raw = value.trim().replace(/\\/g, "/");
	if (raw.length === 0 || raw.split("/").includes("..")) return;
	if (/^[A-Za-z]:/.test(raw)) return;
	const normalized = path.posix.normalize(raw);
	const segments = normalized.split("/");
	if (path.win32.isAbsolute(value) || path.posix.isAbsolute(normalized) || segments.includes("..")) return;
	return normalized === "." ? "." : normalized;
}
//#endregion
export { PRISMA_APP_CONFIG_FILENAME, detectLegacyBuildSettings, resolveConfiguredAppBuildSettings, resolveInferredAppBuildSettings };
