// @prisma/cli/dist/lib/app/compute-config.js
import { CliError } from "../../shell/errors.js";
import { COMPUTE_CONFIG_FILENAME, COMPUTE_CONFIG_FILENAME as COMPUTE_CONFIG_FILENAME$1, ComputeConfigTargetRequiredError, computeTargetAppDir, frameworkByKey, inferComputeTargetFromCwd, loadComputeConfig, selectComputeDeployTarget } from "@prisma/compute-sdk/config";
import path from "node:path";
import { matchError } from "better-result";
//#region src/lib/app/compute-config.ts
/**
* Loads the nearest compute config, searching from `cwd` up to the source
* root (repository or workspace boundary). Thin adapter over the SDK loader
* keeping the CLI's positional-signal call shape.
*/
async function loadComputeConfig$1(cwd, signal) {
	return loadComputeConfig(cwd, { signal });
}
/** Local build/run strategy implied by a configured framework. */
function computeFrameworkToBuildType(framework) {
	return frameworkByKey(framework).buildType;
}
function mergeComputeDeployInputs(options) {
	const { cli, target, configFilename } = options;
	const configAnnotation = `set by ${configFilename}`;
	const framework = cli.framework ? {
		value: cli.framework,
		annotation: "set by --framework"
	} : target?.framework ? {
		value: target.framework,
		annotation: configAnnotation
	} : void 0;
	const entrypoint = cli.entrypoint ? {
		value: cli.entrypoint,
		annotation: "set by --entry"
	} : target?.entry ? {
		value: target.entry,
		annotation: configAnnotation
	} : void 0;
	const httpPort = cli.httpPort ? {
		value: cli.httpPort,
		annotation: "set by --http-port"
	} : target?.httpPort ? {
		value: String(target.httpPort),
		annotation: configAnnotation
	} : void 0;
	const cliEnvInputs = cli.envInputs && cli.envInputs.length > 0 ? cli.envInputs : void 0;
	const configEnvInputs = target && target.envInputs.length > 0 ? target.envInputs : void 0;
	const envInputs = cliEnvInputs ?? configEnvInputs;
	const configAppName = target?.name ? {
		value: target.name,
		annotation: configAnnotation
	} : target?.key ? {
		value: target.key,
		annotation: configAnnotation
	} : void 0;
	return {
		framework,
		entrypoint,
		httpPort,
		envInputs,
		envInputsFromConfig: !cliEnvInputs && configEnvInputs !== void 0,
		configAppName,
		appRoot: target?.root ?? void 0
	};
}
/**
* Merges CLI inputs for the local `app build` and `app run` commands with a
* selected config target. Explicit flags win; `--build-type auto` is the
* flag default and defers to the configured framework.
*/
function mergeComputeLocalInputs(options) {
	const { cli, target } = options;
	const cliBuildType = cli.buildType && cli.buildType !== "auto" ? cli.buildType : void 0;
	const configBuildType = target?.framework ? computeFrameworkToBuildType(target.framework) : void 0;
	return {
		entrypoint: cli.entrypoint ?? target?.entry ?? void 0,
		buildType: cliBuildType ?? configBuildType,
		buildTypeFromConfig: !cliBuildType && configBuildType !== void 0,
		port: cli.port ?? (target?.httpPort ? String(target.httpPort) : void 0),
		appRoot: target?.root ?? void 0
	};
}
function computeConfigErrorToCliError(error, commandName = "deploy") {
	const command = `prisma-cli app ${commandName}`;
	return matchError(error, {
		ComputeConfigAmbiguousError: (ambiguous) => new CliError({
			code: "COMPUTE_CONFIG_INVALID",
			domain: "app",
			summary: "Multiple compute config files found",
			why: ambiguous.message,
			fix: `Keep exactly one compute config file, preferably ${COMPUTE_CONFIG_FILENAME}.`,
			meta: { configPaths: ambiguous.configPaths },
			exitCode: 2,
			nextSteps: [command]
		}),
		ComputeConfigLoadError: (load) => new CliError({
			code: "COMPUTE_CONFIG_INVALID",
			domain: "app",
			summary: `Could not load ${path.basename(load.configPath)}`,
			why: load.message,
			fix: `Fix the error in ${path.basename(load.configPath)} and rerun the command.`,
			where: load.configPath,
			meta: { configPath: load.configPath },
			exitCode: 2,
			nextSteps: [command]
		}),
		ComputeConfigInvalidError: (invalid) => new CliError({
			code: "COMPUTE_CONFIG_INVALID",
			domain: "app",
			summary: `Invalid ${path.basename(invalid.configPath)}`,
			why: invalid.issues.join(" "),
			fix: `Edit ${path.basename(invalid.configPath)} so it default-exports defineComputeConfig({ app }) or defineComputeConfig({ apps }).`,
			where: invalid.configPath,
			meta: {
				configPath: invalid.configPath,
				issues: invalid.issues
			},
			exitCode: 2,
			nextSteps: [command]
		}),
		ComputeConfigTargetRequiredError: (required) => new CliError({
			code: "COMPUTE_CONFIG_TARGET_REQUIRED",
			domain: "app",
			summary: "App target required",
			why: required.message,
			fix: `Pass the app target, for example ${command} <target>.`,
			meta: {
				configPath: required.configPath,
				availableTargets: required.availableTargets
			},
			exitCode: 2,
			nextSteps: required.availableTargets.map((target) => `${command} ${target}`)
		}),
		ComputeConfigTargetUnknownError: (unknown) => new CliError({
			code: "COMPUTE_CONFIG_TARGET_UNKNOWN",
			domain: "app",
			summary: `Unknown app target "${unknown.requestedTarget}"`,
			why: unknown.message,
			fix: unknown.availableTargets.length > 0 ? `Pass one of the configured targets: ${unknown.availableTargets.join(", ")}.` : "Remove the target argument; this config defines a single app.",
			meta: {
				configPath: unknown.configPath,
				requestedTarget: unknown.requestedTarget,
				availableTargets: unknown.availableTargets
			},
			exitCode: 2,
			nextSteps: unknown.availableTargets.length > 0 ? unknown.availableTargets.map((target) => `${command} ${target}`) : [command]
		})
	});
}
//#endregion
export { COMPUTE_CONFIG_FILENAME$1 as COMPUTE_CONFIG_FILENAME, ComputeConfigTargetRequiredError, computeConfigErrorToCliError, computeFrameworkToBuildType, computeTargetAppDir, inferComputeTargetFromCwd, loadComputeConfig$1 as loadComputeConfig, mergeComputeDeployInputs, mergeComputeLocalInputs, selectComputeDeployTarget };
