// @prisma/cli/dist/lib/app/deploy-progress.js
import { renderDeployOutputRows } from "./deploy-output.js";
//#region src/lib/app/deploy-progress.ts
function createDeployProgressState() {
	return {
		buildStarted: false,
		buildCompleted: false,
		archiveReady: false,
		uploadCompleted: false,
		versionId: null,
		startRequested: false,
		containerLive: false,
		deploymentUrl: null,
		promotedUrl: null
	};
}
function createDeployProgress(output, ui, enabled, state = createDeployProgressState()) {
	const write = (line) => {
		if (!enabled) return;
		output.write(`${line}\n`);
	};
	const writeRows = (rows) => {
		for (const line of renderDeployOutputRows(ui, rows)) write(line);
	};
	return {
		onBuildStart() {
			state.buildStarted = true;
			write("Building locally...");
		},
		onBuildComplete() {
			state.buildCompleted = true;
		},
		onArchiveReady(byteLength) {
			state.archiveReady = true;
			writeRows([{
				label: "Built",
				value: formatArtifactSize(byteLength)
			}]);
		},
		onUploadStart() {
			write("Uploading...");
		},
		onVersionCreated(versionId) {
			state.versionId = versionId;
		},
		onUploadComplete() {
			state.uploadCompleted = true;
			writeRows([{ label: "Uploaded" }]);
		},
		onStartRequested() {
			state.startRequested = true;
			write("Deploying...");
		},
		onRunning(url) {
			state.containerLive = true;
			state.deploymentUrl = url;
			writeRows([{ label: "Deployed" }]);
		},
		onPromoted(url) {
			state.promotedUrl = url;
		}
	};
}
function formatArtifactSize(byteLength) {
	return `${(byteLength / 1024 / 1024).toFixed(1)} MB`;
}
function createPromoteProgress(output, enabled) {
	if (!enabled) return;
	const write = (line) => {
		output.write(`${line}\n`);
	};
	return {
		onVersionStarting(versionId) {
			write(`Starting deployment ${versionId}...`);
		},
		onVersionStartRequested() {
			write("Requesting deployment start...");
		},
		onStatusChange(status) {
			write(`Status: ${status}`);
		},
		onVersionRunning() {
			write("Deployment is running.");
		},
		onPromoteStart() {
			write("Promoting deployment...");
		},
		onPromoted(url) {
			if (url) {
				write(`Promoted to ${url}.`);
				return;
			}
			write("Promotion complete.");
		},
		onPromoteFailed(error) {
			write(`Promotion failed${error?.message ? `: ${error.message}` : "."}`);
		}
	};
}
//#endregion
export { createDeployProgress, createDeployProgressState, createPromoteProgress };
