// @prisma/cli/dist/lib/app/deploy-plan.js
//#region src/lib/app/deploy-plan.ts
/**
* Decides whether a deploy covers one app or the whole system. A multi-app
* config with no target named or inferred deploys every target in declaration
* order; everything else (single-app config, an explicit/inferred target, or
* no config at all) is a single deploy.
*/
function planAppDeploy(options) {
	const { config, requestedTarget, hasCreateProject } = options;
	if (!(config !== null && config.kind === "multi" && !requestedTarget && config.targets.length > 1)) return { mode: "single" };
	const total = config.targets.length;
	return {
		mode: "all",
		targets: config.targets.map((target, index) => ({
			targetKey: target.key,
			index,
			total,
			bindsCreateProject: index === 0 && hasCreateProject
		}))
	};
}
/**
* Names the per-app inputs present in a deploy-all invocation, in the order
* they should be reported. An empty list means the run is unambiguous; the
* caller renders the usage error when it is not.
*/
function perAppInputsForDeployAll(inputs) {
	return [
		["--app", inputs.appName],
		["--framework", inputs.framework],
		["--entry", inputs.entrypoint],
		["--http-port", inputs.httpPort],
		["--env", inputs.envAssignments?.length ? inputs.envAssignments : void 0],
		[inputs.appIdEnvVar.name, inputs.appIdEnvVar.value]
	].filter(([, value]) => value !== void 0).map(([flag]) => flag);
}
/**
* Describes a deploy-all run that stopped at the first failing target: which
* targets already went live and which were never attempted, so the caller can
* surface a converge path. Pure; the caller folds this into the CliError.
*/
function describeDeployAllFailure(options) {
	const { targetKeys, failedIndex, completed } = options;
	const failedTarget = targetKeys[failedIndex];
	const notAttempted = targetKeys.slice(failedIndex + 1);
	return {
		failedTarget,
		completed,
		notAttempted,
		contextLines: [
			`Deploying all apps stopped at "${failedTarget}" (${failedIndex + 1}/${targetKeys.length}).`,
			...completed.length > 0 ? [`Already live: ${completed.map((entry) => entry.target).join(", ")}.`] : [],
			...notAttempted.length > 0 ? [`Not attempted: ${notAttempted.join(", ")}.`] : []
		]
	};
}
//#endregion
export { describeDeployAllFailure, perAppInputsForDeployAll, planAppDeploy };
