// @prisma/cli/dist/lib/app/env-file.js
import { usageError } from "../../shell/errors.js";
import { validateKey } from "./env-config.js";
import { readFile } from "node:fs/promises";
import path from "node:path";
import { parse } from "dotenv";
//#region src/lib/app/env-file.ts
const ASSIGNMENT_KEY_PATTERN = /^\s*(?:export\s+)?([^#=\s]+)\s*=/;
async function readEnvFileAssignments(cwd, filePath, command) {
	const resolvedPath = path.resolve(cwd, filePath);
	let contents;
	try {
		contents = await readFile(resolvedPath, "utf8");
	} catch (error) {
		throw usageError(`Failed to read env file "${filePath}"`, error instanceof Error ? error.message : "The file could not be read.", "Pass a readable dotenv file path.", [command === "deploy" ? "prisma-cli app deploy --env .env" : `prisma-cli project env ${command} --file .env --role preview`], "app");
	}
	return parseEnvFileContents(contents, filePath, command);
}
function parseEnvFileContents(contents, filePath, command) {
	const parsedKeys = extractParsedKeys(contents);
	if (parsedKeys.length === 0) throw usageError(`No environment variables found in "${filePath}"`, "The file does not contain any KEY=VALUE assignments.", "Pass a dotenv file with at least one non-empty variable.", [], "app");
	const seen = /* @__PURE__ */ new Map();
	for (const entry of parsedKeys) {
		validateEnvFileKey(entry.key, entry.line, filePath, command);
		const firstLine = seen.get(entry.key);
		if (firstLine !== void 0) throw usageError(`Duplicate environment variable "${entry.key}" in "${filePath}"`, `Lines ${firstLine} and ${entry.line} both define ${entry.key}.`, "Keep one assignment for each key before importing the file.", [], "app");
		seen.set(entry.key, entry.line);
	}
	const parsedValues = parse(contents);
	return parsedKeys.map(({ key }) => {
		const value = parsedValues[key];
		if (typeof value !== "string" || value.length === 0) {
			const line = seen.get(key);
			throw usageError(`Environment variable "${key}" in "${filePath}" has an empty value`, line === void 0 ? `${key} has an empty value.` : `Line ${line} defines ${key} with an empty value.`, "Pass a non-empty value, or omit the key from the file.", [], "app");
		}
		return {
			key,
			value
		};
	});
}
function extractParsedKeys(contents) {
	const keys = [];
	let multilineQuote = null;
	const lines = contents.split(/\n/);
	for (const [index, line] of lines.entries()) {
		const lineNumber = index + 1;
		if (multilineQuote !== null) {
			if (hasClosingQuote(line, multilineQuote, 0)) multilineQuote = null;
			continue;
		}
		const match = ASSIGNMENT_KEY_PATTERN.exec(line);
		if (!match) continue;
		const key = match[1];
		keys.push({
			key,
			line: lineNumber
		});
		const valueStart = line.slice(match[0].length).trimStart();
		const openingQuote = valueStart[0];
		if ((openingQuote === "\"" || openingQuote === "'" || openingQuote === "`") && !hasClosingQuote(valueStart, openingQuote, 1)) multilineQuote = openingQuote;
	}
	return keys;
}
function validateEnvFileKey(key, line, filePath, command) {
	try {
		validateKey(key, command === "deploy" ? "add" : command);
	} catch (error) {
		const reason = error instanceof Error && error.message.length > 0 ? error.message : "Invalid environment variable key.";
		throw usageError(`Invalid environment variable "${key}" in "${filePath}"`, `Line ${line}: ${reason}`, "Use a valid env-var key and retry the import.", [], "app");
	}
}
function hasClosingQuote(value, quote, startIndex) {
	for (let index = startIndex; index < value.length; index += 1) if (value[index] === quote && !isEscaped(value, index)) return true;
	return false;
}
function isEscaped(value, index) {
	let backslashes = 0;
	for (let cursor = index - 1; cursor >= 0 && value[cursor] === "\\"; cursor -= 1) backslashes += 1;
	return backslashes % 2 === 1;
}
//#endregion
export { readEnvFileAssignments };
