// @prisma/cli/dist/lib/app/local-dev.js
import { resolveBunEntrypoint } from "./bun-project.js";
import "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";
//#region src/lib/app/local-dev.ts
const DEFAULT_LOCAL_DEV_PORT = 3e3;
async function runLocalApp(options) {
	const spawnImpl = options.spawnImpl ?? spawn;
	if (options.buildType === "nextjs") {
		const command = await runWithFallback([
			{
				command: path.join(options.appPath, "node_modules", ".bin", process.platform === "win32" ? "next.cmd" : "next"),
				args: [
					"dev",
					"--port",
					String(options.port)
				],
				display: `next dev --port ${options.port}`
			},
			{
				command: "npx",
				args: [
					"next",
					"dev",
					"--port",
					String(options.port)
				],
				display: `next dev --port ${options.port}`
			},
			{
				command: "bunx",
				args: [
					"next",
					"dev",
					"--port",
					String(options.port)
				],
				display: `next dev --port ${options.port}`
			}
		], {
			cwd: options.appPath,
			env: {
				...options.env,
				PORT: String(options.port)
			},
			signal: options.signal
		}, spawnImpl, "Could not find the Next.js CLI. Install it with `npm install next` or ensure npx/bunx is available.");
		return {
			framework: "nextjs",
			entrypoint: null,
			port: options.port,
			command: command.display,
			exitCode: command.exitCode,
			signal: command.signal
		};
	}
	const entrypoint = await resolveBunEntrypoint(options.appPath, options.entrypoint, options.signal);
	const command = await runWithFallback([{
		command: "bun",
		args: ["--watch", entrypoint],
		display: `bun --watch ${entrypoint}`
	}], {
		cwd: options.appPath,
		env: {
			...options.env,
			PORT: String(options.port)
		},
		signal: options.signal
	}, spawnImpl, "Bun is required to run this app locally. Install it from https://bun.sh.");
	return {
		framework: "bun",
		entrypoint,
		port: options.port,
		command: command.display,
		exitCode: command.exitCode,
		signal: command.signal
	};
}
async function runWithFallback(candidates, options, spawnImpl, missingCommandMessage) {
	for (const candidate of candidates) try {
		const result = await spawnCommand(candidate, options, spawnImpl);
		return {
			display: candidate.display,
			exitCode: result.exitCode,
			signal: result.signal
		};
	} catch (error) {
		if (error.code === "ENOENT") continue;
		throw error;
	}
	throw new Error(missingCommandMessage);
}
function spawnCommand(candidate, options, spawnImpl) {
	return new Promise((resolve, reject) => {
		const child = spawnImpl(candidate.command, candidate.args, {
			...options,
			stdio: "inherit"
		});
		child.once("error", reject);
		child.once("exit", (code, signal) => {
			resolve({
				exitCode: code ?? 1,
				signal
			});
		});
	});
}
//#endregion
export { DEFAULT_LOCAL_DEV_PORT, runLocalApp };
