// @prisma/cli/dist/lib/app/production-deploy-gate.js
import { CliError } from "../../shell/errors.js";
import { confirmPrompt } from "../../shell/prompt.js";
import { canPrompt } from "../../shell/runtime.js";
//#region src/lib/app/production-deploy-gate.ts
async function enforceProductionDeployGate(context, provider, options) {
	if (options.branchKind !== "production") return { firstProductionDeploy: false };
	if (!options.appId) {
		renderFirstProductionDeployLine(context, options.appName);
		return { firstProductionDeploy: true };
	}
	const currentLiveDeployment = resolveCurrentProductionDeployment(await provider.listDeployments(options.appId).catch((error) => {
		throw productionDeployInspectionFailedError(error);
	}));
	if (!currentLiveDeployment) {
		renderFirstProductionDeployLine(context, options.appName);
		return { firstProductionDeploy: true };
	}
	if (!options.prod) throw productionDeployRequiresFlagError();
	if (context.flags.yes) {
		renderProductionDeployYesLine(context);
		return { firstProductionDeploy: false };
	}
	if (!canPrompt(context)) throw productionDeployConfirmationRequiredError(options.appName);
	renderProductionDeployConfirmation(context, currentLiveDeployment);
	if (!await confirmPrompt({
		input: context.runtime.stdin,
		output: context.output.stderr,
		message: "Deploy to production?",
		initialValue: false
	})) throw productionDeployCancelledError();
	return { firstProductionDeploy: false };
}
function resolveCurrentProductionDeployment(result) {
	if (result.deployments.length === 0) return null;
	if (result.app.liveDeploymentId) {
		const live = result.deployments.find((deployment) => deployment.id === result.app.liveDeploymentId);
		if (live) return live;
	}
	return result.deployments.find((deployment) => deployment.live === true) ?? result.deployments[0] ?? null;
}
function renderFirstProductionDeployLine(context, appName) {
	if (context.flags.json || context.flags.quiet) return;
	context.output.stderr.write(`First deploy of "${appName}" -- promoting to production.\n\n`);
}
function renderProductionDeployYesLine(context) {
	if (context.flags.json || context.flags.quiet) return;
	context.output.stderr.write("Deploying to production (--prod --yes).\n\n");
}
function renderProductionDeployConfirmation(context, currentLiveDeployment) {
	if (context.flags.json || context.flags.quiet) return;
	const lines = [
		"This will deploy to production and replace the live deployment.",
		"",
		`  Current live:  ${currentLiveDeployment.id} deployed ${formatDeploymentAge(currentLiveDeployment.createdAt)}`,
		"  New deploy:    will be built from your local code",
		""
	];
	context.output.stderr.write(`${lines.join("\n")}\n`);
}
function formatDeploymentAge(createdAt) {
	const createdAtMs = Date.parse(createdAt);
	if (!Number.isFinite(createdAtMs)) return createdAt;
	const elapsedMs = Math.max(0, Date.now() - createdAtMs);
	for (const unit of [
		{
			label: "day",
			ms: 1440 * 60 * 1e3
		},
		{
			label: "hour",
			ms: 3600 * 1e3
		},
		{
			label: "minute",
			ms: 60 * 1e3
		}
	]) if (elapsedMs >= unit.ms) {
		const value = Math.floor(elapsedMs / unit.ms);
		return `${value} ${unit.label}${value === 1 ? "" : "s"} ago`;
	}
	return "less than a minute ago";
}
function productionDeployRequiresFlagError() {
	return new CliError({
		code: "PROD_DEPLOY_REQUIRES_FLAG",
		domain: "app",
		summary: "Production deploy requires --prod",
		why: "The resolved Branch is production and this App already has a production deployment.",
		fix: "Re-run with --prod, or deploy from a preview Branch.",
		exitCode: 2,
		nextActions: [{
			kind: "run-command",
			journey: "deploy-app",
			label: "Deploy to production",
			command: "prisma-cli app deploy --prod"
		}, {
			kind: "run-command",
			journey: "deploy-app",
			label: "Create a preview branch",
			commands: ["git checkout -b <branch-name>", "prisma-cli app deploy"]
		}],
		humanLines: [
			"This would deploy to production.",
			"",
			"Production deploys require explicit intent. Re-run with:",
			"",
			"  prisma-cli app deploy --prod",
			"",
			"Or deploy a preview from a feature branch:",
			"",
			"  git checkout -b <branch-name>",
			"  prisma-cli app deploy"
		]
	});
}
function productionDeployConfirmationRequiredError(appName) {
	return new CliError({
		code: "CONFIRMATION_REQUIRED",
		domain: "app",
		summary: "Production deploy requires confirmation in the current mode",
		why: "This command cannot prompt for production deploy confirmation in the current mode.",
		fix: `Pass --prod --yes to confirm deployment of "${appName}" to production.`,
		exitCode: 1,
		nextSteps: ["prisma-cli app deploy --prod --yes"],
		nextActions: [{
			kind: "run-command",
			journey: "deploy-app",
			label: "Deploy to production non-interactively",
			command: "prisma-cli app deploy --prod --yes"
		}]
	});
}
function productionDeployCancelledError() {
	return new CliError({
		code: "CONFIRMATION_REQUIRED",
		domain: "app",
		summary: "Production deploy cancelled",
		why: null,
		fix: null,
		exitCode: 0,
		humanLines: ["Cancelled."]
	});
}
function productionDeployInspectionFailedError(error) {
	return new CliError({
		code: "DEPLOY_FAILED",
		domain: "app",
		summary: "Failed to inspect production deployments",
		why: error instanceof Error ? error.message : String(error),
		fix: "Retry the command, or rerun with --trace for more detailed diagnostics.",
		debug: formatDebugDetails(error),
		exitCode: 1,
		nextSteps: ["prisma-cli app list-deploys"]
	});
}
function formatDebugDetails(error) {
	if (error instanceof Error) return error.stack ?? error.message;
	return typeof error === "string" ? error : null;
}
//#endregion
export { enforceProductionDeployGate };
