// @prisma/cli/dist/presenters/app-env.js
import { renderVerboseBlock } from "../shell/ui.js";
import { renderList, renderShow, serializeList } from "../output/patterns.js";
import { renderResolvedProjectContextBlock, stripVerboseContext } from "./verbose-context.js";
//#region src/presenters/app-env.ts
function scopeLabel(scope) {
	if (scope.kind === "role") return scope.role ?? "unknown";
	if (scope.kind === "overview") return "overview";
	return `branch:${scope.branchName ?? scope.branchId ?? "unknown"}`;
}
function listTargetLabel(result) {
	const target = result.target;
	if (target.source === "overview") return "overview";
	if (target.branchName) {
		const suffix = target.branchExists === false ? " (not created yet)" : "";
		return `branch:${target.branchName} -> ${target.envMap}${suffix}`;
	}
	return scopeLabel(result.scope);
}
function renderEnvVerboseBlocks(context, result) {
	return [...renderEnvResolvedContextBlock(context, result.verboseContext), ...renderEnvTargetBlock(context, result)];
}
function renderEnvResolvedContextBlock(context, verboseContext) {
	return renderResolvedProjectContextBlock(context.ui, verboseContext);
}
function renderEnvTargetBlock(context, result) {
	return renderVerboseBlock(context.ui, envTargetRows(result), { title: "Env target" });
}
function envTargetRows(result) {
	return [
		{
			key: "project id",
			value: result.projectId,
			tone: "dim"
		},
		{
			key: "scope",
			value: scopeLabel(result.scope)
		},
		...envListTargetRows(result),
		...envFileRows(result),
		{
			key: "keys",
			value: formatKeyNames(envResultKeys(result)),
			tone: envResultKeys(result).length > 0 ? "default" : "dim"
		}
	];
}
function envListTargetRows(result) {
	if (!("target" in result)) return [];
	return [
		{
			key: "target source",
			value: result.target.source
		},
		{
			key: "env map",
			value: result.target.envMap
		},
		...result.target.branchName ? [{
			key: "branch",
			value: result.target.branchName
		}] : [],
		...result.target.branchId ? [{
			key: "branch id",
			value: result.target.branchId,
			tone: "dim"
		}] : [],
		...result.target.branchExists === false ? [{
			key: "branch state",
			value: "not created yet",
			tone: "warning"
		}] : []
	];
}
function envFileRows(result) {
	if (!("file" in result) || !result.file) return [];
	return [{
		key: "file",
		value: result.file.path
	}, {
		key: "file count",
		value: String(result.file.count)
	}];
}
function envResultKeys(result) {
	if ("variables" in result && result.variables) return result.variables.map((variable) => variable.key).sort((left, right) => left.localeCompare(right));
	if ("variable" in result && result.variable) return [result.variable.key];
	if ("key" in result) return [result.key];
	return [];
}
function formatKeyNames(keys) {
	return keys.length > 0 ? keys.join(", ") : "none";
}
function renderEnvAdd(context, descriptor, result) {
	if (result.variables !== void 0) {
		const lines = renderList({
			title: "Setting new environment variables from file.",
			descriptor,
			parentContext: {
				key: "target",
				value: `${scopeLabel(result.scope)} from ${result.file.path}`
			},
			items: result.variables.map((variable) => ({
				noun: "variable",
				label: `${variable.key} (${variable.source})`,
				id: variable.id,
				status: variable.isManagedBySystem ? "default" : null
			})),
			emptyMessage: "No environment variables imported."
		}, context.ui);
		lines.push(...renderEnvVerboseBlocks(context, result));
		return lines;
	}
	const lines = renderShow({
		title: "Setting a new environment variable.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "scope",
				value: scopeLabel(result.scope)
			},
			{
				key: "key",
				value: result.variable.key
			},
			{
				key: "id",
				value: result.variable.id,
				tone: "dim"
			},
			{
				key: "last updated",
				value: result.variable.updatedAt,
				tone: "dim"
			}
		]
	}, context.ui);
	lines.push(...renderEnvVerboseBlocks(context, result));
	return lines;
}
function serializeEnvAdd(result) {
	return stripVerboseContext(result);
}
function renderEnvUpdate(context, descriptor, result) {
	if (result.variables !== void 0) {
		const lines = renderList({
			title: "Replacing environment variable values from file.",
			descriptor,
			parentContext: {
				key: "target",
				value: `${scopeLabel(result.scope)} from ${result.file.path}`
			},
			items: result.variables.map((variable) => ({
				noun: "variable",
				label: `${variable.key} (${variable.source})`,
				id: variable.id,
				status: variable.isManagedBySystem ? "default" : null
			})),
			emptyMessage: "No environment variables updated."
		}, context.ui);
		lines.push(...renderEnvVerboseBlocks(context, result));
		return lines;
	}
	const lines = renderShow({
		title: "Replacing the environment variable's value.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "scope",
				value: scopeLabel(result.scope)
			},
			{
				key: "key",
				value: result.variable.key
			},
			{
				key: "id",
				value: result.variable.id,
				tone: "dim"
			},
			{
				key: "last updated",
				value: result.variable.updatedAt,
				tone: "dim"
			}
		]
	}, context.ui);
	lines.push(...renderEnvVerboseBlocks(context, result));
	return lines;
}
function serializeEnvUpdate(result) {
	return stripVerboseContext(result);
}
function renderEnvList(context, descriptor, result) {
	const lines = renderList({
		title: "Listing environment variables for the selected scope.",
		descriptor,
		parentContext: {
			key: "target",
			value: listTargetLabel(result)
		},
		items: result.variables.map((variable) => ({
			noun: "variable",
			label: `${variable.key} (${variable.source})`,
			id: variable.id,
			status: variable.isManagedBySystem ? "default" : null
		})),
		emptyMessage: "No environment variables defined in this scope."
	}, context.ui);
	lines.push(...renderEnvVerboseBlocks(context, result));
	return lines;
}
function serializeEnvList(result) {
	const serializable = stripVerboseContext(result);
	return {
		projectId: serializable.projectId,
		scope: serializable.scope,
		target: serializable.target,
		...serializeList({
			context: { target: listTargetLabel(serializable) },
			items: serializable.variables.map((variable) => ({
				noun: "variable",
				label: `${variable.key} (${variable.source})`,
				id: variable.id,
				status: variable.isManagedBySystem ? "default" : null
			}))
		}),
		variables: serializable.variables
	};
}
function renderEnvRm(context, descriptor, result) {
	const lines = renderShow({
		title: "Removing the environment variable from the scope.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "scope",
				value: scopeLabel(result.scope)
			},
			{
				key: "key",
				value: result.key
			}
		]
	}, context.ui);
	lines.push(...renderEnvVerboseBlocks(context, result));
	return lines;
}
function serializeEnvRm(result) {
	return stripVerboseContext(result);
}
//#endregion
export { renderEnvAdd, renderEnvList, renderEnvRm, renderEnvUpdate, serializeEnvAdd, serializeEnvList, serializeEnvRm, serializeEnvUpdate };
