// @prisma/cli/dist/presenters/app.js
import { renderVerboseBlock } from "../shell/ui.js";
import { renderDeployOutputRows } from "../lib/app/deploy-output.js";
import { formatDomainFailureFix } from "../lib/app/domain-guidance.js";
import { renderList, renderShow, serializeList } from "../output/patterns.js";
import { renderResolvedProjectContextBlock, stripVerboseContext } from "./verbose-context.js";
//#region src/presenters/app.ts
function renderAppBuild(context, descriptor, result) {
	return renderShow({
		title: "Building the local app artifact.",
		descriptor,
		fields: [
			{
				key: "build type",
				value: result.buildType
			},
			{
				key: "entrypoint",
				value: result.entrypoint ?? "none",
				tone: result.entrypoint ? "default" : "dim"
			},
			{
				key: "directory",
				value: result.directory
			}
		]
	}, context.ui);
}
function serializeAppBuild(result) {
	return result;
}
function renderAppDeploy(context, descriptor, result, options) {
	const logsCommand = options?.logsTarget ? `prisma-cli app logs ${options.logsTarget}` : "prisma-cli app logs";
	return [
		`Live in ${formatDuration(result.durationMs)}`,
		...result.deployment.url ? [context.ui.link(result.deployment.url)] : [],
		...renderBranchDatabaseDeploySummary(context, result),
		"",
		...renderDeployOutputRows(context.ui, [{
			label: "Logs",
			value: logsCommand
		}]),
		...renderDeployResolvedContextBlock(context, result),
		...renderDeploySettingsBlock(context, result)
	];
}
function isAppDeployAllResult(result) {
	return "deployments" in result;
}
function renderAppDeployAll(context, descriptor, result) {
	const lines = [];
	for (const deployment of result.deployments) {
		lines.push(deployment.target);
		lines.push(...renderAppDeploy(context, descriptor, deployment.result, { logsTarget: deployment.target }).map((line) => line ? `  ${line}` : line));
		lines.push("");
	}
	lines.push(...renderDeployOutputRows(context.ui, result.deployments.map((deployment) => ({
		label: deployment.target,
		value: deployment.result.deployment.url ?? deployment.result.deployment.id
	}))));
	return lines;
}
function serializeAppDeployAll(result) {
	return {
		count: result.deployments.length,
		deployments: result.deployments.map((deployment) => ({
			target: deployment.target,
			...serializeAppDeploy(deployment.result)
		}))
	};
}
function serializeAppDeploy(result) {
	const { deploySettings, localPin: _localPin, ...serialized } = result;
	const { id: _branchId, ...branch } = serialized.branch;
	return {
		...serialized,
		branch,
		deploySettings: {
			config: deploySettings.config,
			buildCommand: deploySettings.buildCommand,
			outputDirectory: deploySettings.outputDirectory
		}
	};
}
function renderBranchDatabaseDeploySummary(context, result) {
	if (!result.branchDatabase || result.branchDatabase.status !== "created") return [];
	return ["", ...renderDeployOutputRows(context.ui, [{
		label: "Database",
		value: result.branchDatabase.database?.name ?? "created"
	}, {
		label: "Env",
		value: result.branchDatabase.envVars.join(", ")
	}])];
}
function formatDuration(durationMs) {
	if (durationMs < 1e3) return `${durationMs}ms`;
	return `${(durationMs / 1e3).toFixed(1)}s`;
}
function renderDeployResolvedContextBlock(context, result) {
	return renderResolvedProjectContextBlock(context.ui, {
		workspace: result.workspace,
		project: result.project,
		resolution: result.resolution,
		branch: {
			id: result.branch.id,
			name: result.branch.name,
			kind: result.branch.kind
		}
	}, { extraRows: [
		{
			key: "app",
			value: result.app.name
		},
		{
			key: "app id",
			value: result.app.id,
			tone: "dim"
		},
		{
			key: "deployment id",
			value: result.deployment.id,
			tone: "dim"
		},
		{
			key: "deployment status",
			value: result.deployment.status
		},
		...result.localPin ? [{
			key: "local pin",
			value: result.localPin.path
		}] : [],
		{
			key: "deploy duration",
			value: formatDuration(result.durationMs)
		}
	] });
}
function renderDeploySettingsBlock(context, result) {
	return renderVerboseBlock(context.ui, [...deploySettingsRows(result.deploySettings), ...branchDatabaseRows(result.branchDatabase)], { title: "Deploy settings" });
}
function deploySettingsRows(settings) {
	return [
		{
			key: "framework",
			value: `${settings.framework.name} (${settings.framework.buildType})`
		},
		{
			key: "framework source",
			value: settings.framework.source,
			tone: "dim"
		},
		{
			key: "entrypoint",
			value: settings.entrypoint ?? "derived from build output",
			tone: settings.entrypoint ? "default" : "dim"
		},
		{
			key: "http port",
			value: String(settings.httpPort)
		},
		{
			key: "region",
			value: settings.region ?? "existing app region",
			tone: settings.region ? "default" : "dim"
		},
		{
			key: "env vars",
			value: formatEnvVarNames(settings.envVars),
			tone: settings.envVars.length > 0 ? "default" : "dim"
		}
	];
}
function branchDatabaseRows(branchDatabase) {
	if (!branchDatabase) return [{
		key: "branch db",
		value: "not configured",
		tone: "dim"
	}];
	return [{
		key: "branch db",
		value: branchDatabase.status === "created" ? `created${branchDatabase.database ? ` (${branchDatabase.database.name})` : ""}` : `skipped${branchDatabase.reason ? ` (${branchDatabase.reason})` : ""}`,
		tone: branchDatabase.status === "created" ? "success" : "dim"
	}, ...branchDatabase.envVars.length > 0 ? [{
		key: "branch db env",
		value: branchDatabase.envVars.join(", ")
	}] : []];
}
function formatEnvVarNames(envVars) {
	return envVars.length > 0 ? envVars.join(", ") : "none";
}
function renderAppListDeploys(context, descriptor, result) {
	const lines = renderList({
		title: "Listing deployments for the selected app.",
		descriptor,
		parentContext: {
			key: "app",
			value: result.app?.name ?? "not selected"
		},
		items: result.deployments.map((deployment) => ({
			noun: "deployment",
			label: deployment.id,
			id: deployment.id,
			status: deployment.live ? "active" : null
		})),
		emptyMessage: result.app ? "No deployments found." : "No apps found."
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppListDeploys(result) {
	const { verboseContext: _verboseContext, ...serializable } = result;
	if (!serializable.app) return {
		projectId: serializable.projectId,
		app: null,
		items: [],
		count: 0
	};
	return {
		projectId: serializable.projectId,
		app: serializable.app,
		...serializeList({
			context: { app: serializable.app.name },
			items: serializable.deployments.map((deployment) => ({
				noun: "deployment",
				label: deployment.id,
				id: deployment.id,
				status: deployment.live ? "active" : null
			}))
		})
	};
}
function renderAppShow(context, descriptor, result) {
	const lines = renderShow({
		title: "Showing the selected app state.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "app",
				value: result.app?.name ?? "not selected",
				tone: result.app ? "default" : "dim"
			},
			{
				key: "live deployment",
				value: result.liveDeployment?.id ?? "none",
				tone: result.liveDeployment ? toneForStatus(result.liveDeployment.status) : "dim"
			},
			{
				key: "live url",
				value: result.liveUrl ?? "unavailable",
				tone: result.liveUrl ? "link" : "dim"
			},
			{
				key: "recent deployments",
				value: formatRecentDeployments(result.recentDeployments),
				tone: result.recentDeployments.length > 0 ? "default" : "dim"
			}
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppShow(result) {
	return stripVerboseContext(result);
}
function renderAppShowDeploy(context, descriptor, result) {
	return renderShow({
		title: "Showing deployment details.",
		descriptor,
		fields: [
			...result.app ? [{
				key: "app",
				value: result.app.name
			}] : [],
			{
				key: "deployment",
				value: result.deployment.id
			},
			{
				key: "status",
				value: result.deployment.status,
				tone: toneForStatus(result.deployment.status)
			},
			...result.deployment.url ? [{
				key: "url",
				value: result.deployment.url,
				tone: "link"
			}] : [],
			...result.deployment.live === null ? [] : [{
				key: "live",
				value: result.deployment.live ? "yes" : "no",
				tone: result.deployment.live ? "success" : "dim"
			}],
			{
				key: "created",
				value: result.deployment.createdAt,
				tone: "dim"
			}
		]
	}, context.ui);
}
function serializeAppShowDeploy(result) {
	return result;
}
function renderAppOpen(context, descriptor, result) {
	const lines = renderShow({
		title: result.opened ? "Opening the live URL for the selected app." : "Resolving the live URL for the selected app.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "app",
				value: result.app.name
			},
			{
				key: "url",
				value: result.url,
				tone: "link"
			},
			{
				key: "opened",
				value: result.opened ? "yes" : "no",
				tone: result.opened ? "success" : "dim"
			}
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppOpen(result) {
	return stripVerboseContext(result);
}
function renderAppDomainAdd(context, descriptor, result) {
	return renderShow({
		title: result.existing ? "Showing the existing custom domain for the selected app." : "Adding a custom domain to the selected app.",
		descriptor,
		fields: [
			...domainTargetFields(result),
			{
				key: "hostname",
				value: result.domain.hostname
			},
			{
				key: "status",
				value: result.domain.status,
				tone: toneForDomainStatus(result.domain.status)
			},
			...domainDnsFields(result.domain)
		]
	}, context.ui);
}
function serializeAppDomainAdd(result) {
	return result;
}
function renderAppDomainShow(context, descriptor, result) {
	return renderShow({
		title: "Showing custom domain status.",
		descriptor,
		fields: [
			...domainTargetFields(result),
			{
				key: "hostname",
				value: result.domain.hostname
			},
			{
				key: "status",
				value: result.domain.status,
				tone: toneForDomainStatus(result.domain.status)
			},
			...domainFailureFields(result.domain),
			{
				key: "cert expires",
				value: formatOptionalUtcDate(result.domain.certExpiresAt),
				tone: result.domain.certExpiresAt ? "default" : "dim"
			},
			{
				key: "created",
				value: formatUtcDate(result.domain.createdAt),
				tone: "dim"
			},
			...domainDnsFields(result.domain)
		]
	}, context.ui);
}
function serializeAppDomainShow(result) {
	return result;
}
function renderAppDomainRemove(context, descriptor, result) {
	return renderShow({
		title: "Removing a custom domain from the selected app.",
		descriptor,
		fields: [
			...domainTargetFields(result),
			{
				key: "hostname",
				value: result.hostname
			},
			{
				key: "removed",
				value: result.removed ? "yes" : "no",
				tone: result.removed ? "success" : "dim"
			}
		]
	}, context.ui);
}
function serializeAppDomainRemove(result) {
	return result;
}
function renderAppDomainRetry(context, descriptor, result) {
	return renderShow({
		title: "Retrying custom domain verification.",
		descriptor,
		fields: [
			...domainTargetFields(result),
			{
				key: "hostname",
				value: result.domain.hostname
			},
			{
				key: "status",
				value: result.domain.status,
				tone: toneForDomainStatus(result.domain.status)
			},
			...domainFailureFields(result.domain),
			...domainDnsFields(result.domain)
		]
	}, context.ui);
}
function serializeAppDomainRetry(result) {
	return result;
}
function renderAppPromote(context, descriptor, result) {
	const lines = renderShow({
		title: "Switching the live deployment for the selected app.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "app",
				value: result.app.name
			},
			{
				key: "deployment",
				value: result.deployment.id
			},
			{
				key: "status",
				value: result.deployment.status,
				tone: toneForStatus(result.deployment.status)
			},
			...result.deployment.url ? [{
				key: "url",
				value: result.deployment.url,
				tone: "link"
			}] : [],
			{
				key: "live",
				value: result.deployment.live ? "yes" : "no",
				tone: result.deployment.live ? "success" : "dim"
			},
			{
				key: "created",
				value: result.deployment.createdAt,
				tone: "dim"
			}
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppPromote(result) {
	return stripVerboseContext(result);
}
function renderAppRollback(context, descriptor, result) {
	const lines = renderShow({
		title: "Restoring the selected app to an earlier deployment.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "app",
				value: result.app.name
			},
			{
				key: "deployment",
				value: result.deployment.id
			},
			{
				key: "status",
				value: result.deployment.status,
				tone: toneForStatus(result.deployment.status)
			},
			...result.deployment.url ? [{
				key: "url",
				value: result.deployment.url,
				tone: "link"
			}] : [],
			{
				key: "live",
				value: result.deployment.live ? "yes" : "no",
				tone: result.deployment.live ? "success" : "dim"
			},
			{
				key: "created",
				value: result.deployment.createdAt,
				tone: "dim"
			},
			...result.previousLiveDeploymentId ? [{
				key: "replaced",
				value: result.previousLiveDeploymentId,
				tone: "dim"
			}] : []
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppRollback(result) {
	return stripVerboseContext(result);
}
function renderAppRun(_context, _descriptor, _result) {
	return [];
}
function serializeAppRun(result) {
	return result;
}
function renderAppRemove(context, descriptor, result) {
	const lines = renderShow({
		title: "Removing the selected app.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectId
			},
			{
				key: "app",
				value: result.app.name
			},
			{
				key: "removed",
				value: result.removed ? "yes" : "no",
				tone: result.removed ? "success" : "dim"
			}
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeAppRemove(result) {
	return stripVerboseContext(result);
}
function toneForStatus(status) {
	if (status === "running" || status === "ready" || status === "healthy") return "success";
	if (status === "provisioning" || status === "building" || status === "starting") return "warning";
	if (status === "failed" || status === "error") return "error";
	return "default";
}
function toneForDomainStatus(status) {
	if (status === "active") return "success";
	if (status === "failed") return "error";
	if (status === "pending_dns" || status === "verifying" || status === "provisioning_tls" || status === "verified_routing_blocked") return "warning";
	return "default";
}
function domainTargetFields(result) {
	return [
		{
			key: "workspace",
			value: result.workspace.name
		},
		{
			key: "project",
			value: result.project.name
		},
		{
			key: "branch",
			value: result.branch.name
		},
		{
			key: "app",
			value: result.app.name
		}
	];
}
function domainDnsFields(domain) {
	const records = domain.dnsRecords;
	if (records.length === 0) return [{
		key: "dns record",
		value: "not provided by platform",
		tone: "dim"
	}];
	return [{
		key: "dns record",
		value: records.map((record) => {
			const ttl = record.ttl ? ` ttl ${record.ttl}` : "";
			return `${record.type} ${record.name} -> ${record.value}${ttl}`;
		}).join(", ")
	}];
}
function formatDomainFailure(domain) {
	if (!domain.failureReason) return domain.failureCategory ?? "none";
	return domain.failureCategory ? `${domain.failureCategory} - ${domain.failureReason}` : domain.failureReason;
}
function domainFailureFields(domain) {
	const tone = hasDomainFailure(domain) ? "error" : "dim";
	return [{
		key: "failure",
		value: formatDomainFailure(domain),
		tone
	}, ...domainFixFields(domain)];
}
function hasDomainFailure(domain) {
	return Boolean(domain.failureCategory || domain.failureReason);
}
function domainFixFields(domain) {
	const fix = formatDomainFailureFix(domain);
	return fix ? [{
		key: "fix",
		value: fix
	}] : [];
}
function formatOptionalUtcDate(value) {
	return value ? formatUtcDate(value) : "-";
}
function formatUtcDate(value) {
	const date = new Date(value);
	if (Number.isNaN(date.getTime())) return value;
	return `${date.getUTCFullYear()}-${String(date.getUTCMonth() + 1).padStart(2, "0")}-${String(date.getUTCDate()).padStart(2, "0")} ${String(date.getUTCHours()).padStart(2, "0")}:${String(date.getUTCMinutes()).padStart(2, "0")} UTC`;
}
function formatRecentDeployments(deployments) {
	if (deployments.length === 0) return "none";
	return deployments.map((deployment) => `${deployment.id}${deployment.live ? " (live)" : ""}`).join(", ");
}
//#endregion
export { isAppDeployAllResult, renderAppBuild, renderAppDeploy, renderAppDeployAll, renderAppDomainAdd, renderAppDomainRemove, renderAppDomainRetry, renderAppDomainShow, renderAppListDeploys, renderAppOpen, renderAppPromote, renderAppRemove, renderAppRollback, renderAppRun, renderAppShow, renderAppShowDeploy, serializeAppBuild, serializeAppDeploy, serializeAppDeployAll, serializeAppDomainAdd, serializeAppDomainRemove, serializeAppDomainRetry, serializeAppDomainShow, serializeAppListDeploys, serializeAppOpen, serializeAppPromote, serializeAppRemove, serializeAppRollback, serializeAppRun, serializeAppShow, serializeAppShowDeploy };
