// @prisma/cli/dist/presenters/branch.js
import { formatDescriptorLabel } from "../shell/command-meta.js";
import { formatColumns } from "../shell/ui.js";
import { renderResolvedProjectContextBlock } from "./verbose-context.js";
//#region src/presenters/branch.ts
function renderBranchList(context, descriptor, result) {
	const ui = context.ui;
	const lines = [`${ui.strong(formatDescriptorLabel(descriptor))} ${ui.dim("→")} ${ui.dim("Listing branches for the resolved project.")}`, ""];
	const rail = ui.dim("│");
	lines.push(`${rail}  ${ui.accent("project:")}  ${result.projectName}`);
	lines.push(rail);
	if (result.branches.length === 0) {
		lines.push(`${rail}  ${ui.dim("No branches found.")}`);
		lines.push(...renderBranchResolvedContextBlock(context, result));
		return lines;
	}
	const widths = [
		Math.max(4, ...result.branches.map((branch) => branch.name.length)),
		Math.max(4, ...result.branches.map((branch) => branch.role.length)),
		Math.max(7, ...result.branches.map((branch) => branch.envMap.length))
	];
	lines.push(`${rail}  ${ui.accent(formatColumns([
		"Name",
		"Role",
		"Env map"
	], widths))}`);
	for (const branch of result.branches) lines.push(`${rail}  ${formatColumns([
		branch.name,
		branch.role,
		branch.envMap
	], widths)}`);
	lines.push(...renderBranchResolvedContextBlock(context, result));
	return lines;
}
function serializeBranchList(result) {
	const { verboseContext: _verboseContext, ...serializable } = result;
	return {
		projectId: serializable.projectId,
		projectName: serializable.projectName,
		branches: serializable.branches
	};
}
function renderBranchResolvedContextBlock(context, result) {
	return renderResolvedProjectContextBlock(context.ui, result.verboseContext);
}
//#endregion
export { renderBranchList, serializeBranchList };
