// @prisma/cli/dist/presenters/database.js
import { formatDescriptorLabel } from "../shell/command-meta.js";
import { formatColumns, renderSummaryLine } from "../shell/ui.js";
import { renderMutate, renderShow, serializeList } from "../output/patterns.js";
import { renderResolvedProjectContextBlock, stripVerboseContext } from "./verbose-context.js";
//#region src/presenters/database.ts
function renderDatabaseList(context, descriptor, result) {
	const ui = context.ui;
	const lines = [`${ui.strong(formatDescriptorLabel(descriptor))} ${ui.dim("→")} ${ui.dim("Listing databases for the resolved project.")}`, ""];
	const rail = ui.dim("│");
	lines.push(`${rail}  ${ui.accent("project:")}  ${result.projectName}`);
	if (result.branchName) lines.push(`${rail}  ${ui.accent("branch:")}   ${result.branchName}`);
	lines.push(rail);
	if (result.databases.length === 0) {
		lines.push(`${rail}  ${ui.dim("No databases found.")}`);
		lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
		return lines;
	}
	const rows = result.databases.map((database) => [
		database.name,
		database.branchName ?? "unscoped",
		database.region ?? "unknown",
		formatStatus(database),
		database.id
	]);
	const widths = [
		Math.max(4, ...rows.map((row) => row[0].length)),
		Math.max(6, ...rows.map((row) => row[1].length)),
		Math.max(6, ...rows.map((row) => row[2].length)),
		Math.max(6, ...rows.map((row) => row[3].length)),
		Math.max(2, ...rows.map((row) => row[4].length))
	];
	lines.push(`${rail}  ${ui.accent(formatColumns([
		"Name",
		"Branch",
		"Region",
		"Status",
		"Id"
	], widths))}`);
	for (const row of rows) lines.push(`${rail}  ${formatColumns(row, widths)}`);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeDatabaseList(result) {
	return {
		...serializeList({
			context: {
				project: result.projectName,
				...result.branchName ? { branch: result.branchName } : {}
			},
			items: result.databases.map((database) => ({
				noun: "database",
				label: database.name,
				id: database.id,
				status: database.isDefault ? "default" : null
			}))
		}),
		projectId: result.projectId,
		branchName: result.branchName,
		databases: result.databases
	};
}
function renderDatabaseShow(context, descriptor, result) {
	const lines = renderShow({
		title: "Showing database metadata.",
		descriptor,
		fields: [
			{
				key: "project",
				value: result.projectName
			},
			{
				key: "database",
				value: result.database.name
			},
			{
				key: "id",
				value: result.database.id,
				tone: "dim"
			},
			{
				key: "branch",
				value: result.database.branchName ?? "unscoped",
				tone: result.database.branchName ? "default" : "dim"
			},
			{
				key: "region",
				value: result.database.region ?? "unknown",
				tone: result.database.region ? "default" : "dim"
			},
			{
				key: "status",
				value: formatStatus(result.database)
			},
			{
				key: "connections",
				value: String(result.connections.length)
			}
		]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeDatabaseShow(result) {
	return stripVerboseContext(result);
}
function renderDatabaseCreateStdout(_context, _descriptor, result) {
	return [result.connectionString];
}
function renderDatabaseCreate(context, _descriptor, result) {
	const ui = context.ui;
	const lines = [
		"Creating database...",
		renderSummaryLine(ui, "success", `Created database "${result.database.name}" in ${formatDatabaseTarget(result.projectName, result.database.branchName)}.`),
		"  The connection URL below is shown once, so save it now."
	];
	if (ui.verbose) {
		lines.push("");
		lines.push(...renderDatabaseCreateVerboseRows(context, result));
	}
	return lines;
}
function serializeDatabaseCreate(result) {
	return stripVerboseContext(result);
}
function renderDatabaseRemove(context, descriptor, result) {
	const lines = renderMutate({
		title: "Removing database.",
		descriptor,
		context: [
			{
				key: "project",
				value: result.projectName
			},
			{
				key: "database",
				value: result.database.name
			},
			{
				key: "id",
				value: result.database.id,
				tone: "dim"
			}
		],
		operationDescription: "Removing database",
		operationCount: 1,
		details: ["Database and its connection metadata were removed."]
	}, context.ui);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeDatabaseRemove(result) {
	return stripVerboseContext(result);
}
function renderDatabaseConnectionList(context, descriptor, result) {
	const ui = context.ui;
	const lines = [`${ui.strong(formatDescriptorLabel(descriptor))} ${ui.dim("→")} ${ui.dim("Listing database connection metadata.")}`, ""];
	const rail = ui.dim("│");
	lines.push(`${rail}  ${ui.accent("database:")}  ${result.database.name}`);
	lines.push(rail);
	if (result.connections.length === 0) {
		lines.push(`${rail}  ${ui.dim("No database connections found.")}`);
		lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
		return lines;
	}
	const rows = result.connections.map((connection) => [
		connection.name,
		connection.id,
		connection.createdAt ?? "unknown"
	]);
	const widths = [
		Math.max(4, ...rows.map((row) => row[0].length)),
		Math.max(2, ...rows.map((row) => row[1].length)),
		Math.max(7, ...rows.map((row) => row[2].length))
	];
	lines.push(`${rail}  ${ui.accent(formatColumns([
		"Name",
		"Id",
		"Created"
	], widths))}`);
	for (const row of rows) lines.push(`${rail}  ${formatColumns(row, widths)}`);
	lines.push(...renderResolvedProjectContextBlock(context.ui, result.verboseContext));
	return lines;
}
function serializeDatabaseConnectionList(result) {
	return {
		...serializeList({
			context: {
				project: result.projectName,
				database: result.database.name
			},
			items: result.connections.map((connection) => ({
				noun: "connection",
				label: connection.name,
				id: connection.id,
				status: null
			}))
		}),
		projectId: result.projectId,
		database: result.database,
		connections: result.connections
	};
}
function renderDatabaseConnectionCreateStdout(_context, _descriptor, result) {
	return [result.connectionString];
}
function renderDatabaseConnectionCreate(context, _descriptor, result) {
	const ui = context.ui;
	const lines = [
		"Creating connection...",
		renderSummaryLine(ui, "success", `Added a connection to "${result.database.name}" in ${formatDatabaseTarget(result.projectName, result.database.branchName)}.`),
		"  The connection URL below is shown once, so save it now."
	];
	if (ui.verbose) {
		lines.push("");
		lines.push(...renderDatabaseConnectionCreateVerboseRows(context, result));
	}
	return lines;
}
function serializeDatabaseConnectionCreate(result) {
	return stripVerboseContext(result);
}
function renderDatabaseConnectionRemove(context, descriptor, result) {
	return renderMutate({
		title: "Removing database connection.",
		descriptor,
		context: [{
			key: "connection",
			value: result.connection.id,
			tone: "dim"
		}],
		operationDescription: "Removing database connection",
		operationCount: 1,
		details: ["The connection metadata was removed. Existing one-time secrets were not shown."]
	}, context.ui);
}
function serializeDatabaseConnectionRemove(result) {
	return { connection: result.connection };
}
function formatStatus(database) {
	return database.status ?? (database.isDefault ? "default" : "unknown");
}
function formatDatabaseTarget(projectName, branchName) {
	return branchName ? `${projectName} / ${branchName}` : projectName;
}
function renderDatabaseCreateVerboseRows(context, result) {
	return renderMetadataRows([
		...renderWorkspaceProjectRows(result),
		["branch", result.database.branchName ?? "unscoped"],
		["database", formatResourceWithId(context, result.database.name, result.database.id)],
		["region", result.database.region ?? "unknown"],
		["status", formatStatus(result.database)],
		["connection", formatResourceWithId(context, result.connection.name, result.connection.id)]
	]);
}
function renderDatabaseConnectionCreateVerboseRows(context, result) {
	return renderMetadataRows([
		...renderWorkspaceProjectRows(result),
		["branch", result.database.branchName ?? "unscoped"],
		["database", formatResourceWithId(context, result.database.name, result.database.id)],
		["connection", formatResourceWithId(context, result.connection.name, result.connection.id)]
	]);
}
function renderWorkspaceProjectRows(result) {
	return [...result.verboseContext ? [["workspace", result.verboseContext.workspace.name]] : [], ["project", result.projectName]];
}
function formatResourceWithId(context, name, id) {
	return `${name}  ${context.ui.dim(`(${id})`)}`;
}
function renderMetadataRows(rows) {
	const widths = [Math.max(...rows.map((row) => row[0].length)), 0];
	return rows.map(([key, value]) => `  ${formatColumns([key, value], widths)}`);
}
//#endregion
export { renderDatabaseConnectionCreate, renderDatabaseConnectionCreateStdout, renderDatabaseConnectionList, renderDatabaseConnectionRemove, renderDatabaseCreate, renderDatabaseCreateStdout, renderDatabaseList, renderDatabaseRemove, renderDatabaseShow, serializeDatabaseConnectionCreate, serializeDatabaseConnectionList, serializeDatabaseConnectionRemove, serializeDatabaseCreate, serializeDatabaseList, serializeDatabaseRemove, serializeDatabaseShow };
