// @prisma/cli/dist/presenters/verbose-context.js
import { renderVerboseBlock } from "../shell/ui.js";
//#region src/presenters/verbose-context.ts
function renderResolvedProjectContextBlock(ui, context, options = {}) {
	if (!context) return [];
	return renderVerboseBlock(ui, [...projectResolutionRows(context), ...options.extraRows ?? []], { title: options.title ?? "Resolved context" });
}
function projectResolutionRows(context) {
	return [
		{
			key: "workspace",
			value: context.workspace.name
		},
		{
			key: "workspace id",
			value: context.workspace.id,
			tone: "dim"
		},
		{
			key: "project",
			value: context.project.name
		},
		{
			key: "project id",
			value: context.project.id,
			tone: "dim"
		},
		{
			key: "project source",
			value: formatProjectSource(context.resolution.projectSource)
		},
		...context.resolution.targetName ? [{
			key: "target name",
			value: formatTargetName(context.resolution)
		}] : [],
		...context.branch ? [{
			key: "branch",
			value: `${context.branch.name} (${context.branch.kind})`
		}, ...context.branch.id ? [{
			key: "branch id",
			value: context.branch.id,
			tone: "dim"
		}] : []] : []
	];
}
function stripVerboseContext(result) {
	const { verboseContext: _verboseContext, ...serialized } = result;
	return serialized;
}
function formatProjectSource(source) {
	switch (source) {
		case "explicit": return "--project";
		case "env": return "environment";
		case "local-pin": return ".prisma/local.json";
		case "platform-mapping": return "platform mapping";
		case "created": return "created";
		case "prompt": return "prompt";
		case "unbound": return "unbound";
	}
}
function formatTargetName(resolution) {
	return resolution.targetNameSource ? `${resolution.targetName} (${resolution.targetNameSource})` : String(resolution.targetName);
}
//#endregion
export { renderResolvedProjectContextBlock, stripVerboseContext };
