// @prisma/cli/dist/lib/auth/auth-ops.js
import { SERVICE_TOKEN_ENV_VAR } from "./client.js";
import { FileTokenStorage } from "../../adapters/token-storage.js";
import { requireComputeAuth } from "./guard.js";
import { login } from "./login.js";
//#region src/lib/auth/auth-ops.ts
const WORKSPACE_SUB_PREFIX = "workspace:";
function decodeJwtPayload(token) {
	try {
		const payload = token.split(".")[1];
		if (!payload) return {};
		return JSON.parse(Buffer.from(payload, "base64url").toString("utf8"));
	} catch {
		return {};
	}
}
function emailFromClaims(claims) {
	const email = claims.email;
	return typeof email === "string" && email.trim().length > 0 ? email.trim() : null;
}
function workspaceIdFromClaims(claims) {
	const sub = claims.sub;
	if (typeof sub !== "string") return null;
	if (!sub.startsWith(WORKSPACE_SUB_PREFIX)) return null;
	const id = sub.slice(10).trim();
	return id.length > 0 ? id : null;
}
async function performLogin(env, signal) {
	await login({
		tokenStorage: new FileTokenStorage(env, signal, { activateOnSetTokens: true }),
		env,
		signal
	});
}
async function readAuthState(env, signal) {
	const rawServiceToken = env[SERVICE_TOKEN_ENV_VAR];
	if (rawServiceToken !== void 0) {
		const serviceToken = rawServiceToken.trim();
		if (serviceToken.length === 0) throw new Error(`${SERVICE_TOKEN_ENV_VAR} is set but empty. Provide a valid token or unset the variable.`);
		return readServiceTokenAuthState(serviceToken, env, signal);
	}
	const tokenStorage = new FileTokenStorage(env, signal);
	const tokens = await tokenStorage.getTokens();
	if (!tokens) return {
		authenticated: false,
		provider: null,
		user: null,
		workspace: null,
		credential: null
	};
	const client = await requireComputeAuth(env, signal);
	const currentPrincipal = await readCurrentPrincipalAuthState(client, signal);
	if (currentPrincipal) {
		if (currentPrincipal.authenticated && currentPrincipal.workspace) await tokenStorage.rememberWorkspace?.(tokens.workspaceId, currentPrincipal.workspace);
		return currentPrincipal;
	}
	const claims = decodeJwtPayload(tokens.accessToken);
	const authState = await buildAuthState({
		workspaceIdFromCredential: tokens.workspaceId,
		claims,
		env,
		client,
		signal
	});
	if (authState.authenticated && authState.workspace) await tokenStorage.rememberWorkspace?.(tokens.workspaceId, authState.workspace);
	return authState;
}
async function readServiceTokenAuthState(token, env, signal) {
	const client = await requireComputeAuth(env, signal);
	const currentPrincipal = await readCurrentPrincipalAuthState(client, signal);
	if (currentPrincipal) return currentPrincipal;
	const claims = decodeJwtPayload(token);
	const workspaceId = workspaceIdFromClaims(claims);
	if (!workspaceId) return {
		authenticated: false,
		provider: null,
		user: null,
		workspace: null,
		credential: null
	};
	return buildAuthState({
		workspaceIdFromCredential: workspaceId,
		claims,
		env,
		client,
		signal
	});
}
async function buildAuthState({ workspaceIdFromCredential, claims, env, client, signal }) {
	let workspaceId = workspaceIdFromCredential;
	let workspaceName = workspaceIdFromCredential;
	client ??= await requireComputeAuth(env, signal);
	if (client) try {
		const { data, response } = await client.GET("/v1/workspaces/{id}", {
			params: { path: { id: workspaceIdFromCredential } },
			signal
		});
		if (response?.status === 401) return {
			authenticated: false,
			provider: null,
			user: null,
			workspace: null,
			credential: null
		};
		if (data?.data?.id) {
			workspaceId = data.data.id;
			workspaceName = data.data.id;
		}
		if (data?.data?.name) workspaceName = data.data.name;
	} catch {
		signal?.throwIfAborted();
	}
	const email = emailFromClaims(claims);
	return {
		authenticated: true,
		provider: null,
		user: email ? { email } : null,
		workspace: {
			id: workspaceId,
			name: workspaceName
		},
		credential: null
	};
}
async function readCurrentPrincipalAuthState(client, signal) {
	if (!client) return null;
	try {
		const { data, response } = await client.GET("/v1/me", { signal });
		if (response?.status === 401) return {
			authenticated: false,
			provider: null,
			user: null,
			workspace: null,
			credential: null
		};
		const principal = data?.data;
		if (!principal) return null;
		if (!principal.credential) return null;
		return {
			authenticated: true,
			provider: null,
			user: principal.user ? {
				id: principal.user.id,
				email: principal.user.email,
				name: principal.user.name
			} : null,
			workspace: principal.workspace,
			credential: principal.credential
		};
	} catch {
		signal?.throwIfAborted();
		return null;
	}
}
async function performLogout(env, signal) {
	await new FileTokenStorage(env, signal).clearTokens();
}
//#endregion
export { performLogin, performLogout, readAuthState };
