// @prisma/cli/dist/lib/auth/guard.js
import { CLIENT_ID, SERVICE_TOKEN_ENV_VAR, getApiBaseUrl } from "./client.js";
import { FileTokenStorage } from "../../adapters/token-storage.js";
import { createManagementApiClient, createManagementApiSdk } from "@prisma/management-api-sdk";
//#region src/lib/auth/guard.ts
/**
* Resolve authentication and return a ManagementApiClient.
*
* Priority:
*   1. PRISMA_SERVICE_TOKEN env var → service token (CI / headless)
*   2. Stored OAuth tokens     → SDK with auto-refresh
*
* Returns null if not authenticated.
*/
async function requireComputeAuth(env = process.env, signal) {
	signal?.throwIfAborted();
	const rawToken = env[SERVICE_TOKEN_ENV_VAR];
	if (rawToken !== void 0) {
		const token = rawToken.trim();
		if (token.length === 0) throw new Error(`${SERVICE_TOKEN_ENV_VAR} is set but empty. Provide a valid token or unset the variable.`);
		return createManagementApiClient({
			baseUrl: getApiBaseUrl(env),
			token
		});
	}
	const tokenStorage = new FileTokenStorage(env, signal, {
		activateOnSetTokens: false,
		lockSetTokens: false
	});
	if (!await tokenStorage.getTokens()) return null;
	return createManagementApiSdk({
		clientId: CLIENT_ID,
		redirectUri: "http://localhost:0/auth/callback",
		tokenStorage,
		apiBaseUrl: getApiBaseUrl(env)
	}).client;
}
//#endregion
export { requireComputeAuth };
