// @prisma/compute-sdk/dist/config/frameworks.js
export const NEXT_CONFIG_FILENAMES = [
    "next.config.js",
    "next.config.mjs",
    "next.config.ts",
    "next.config.mts",
];
export const NUXT_CONFIG_FILENAMES = [
    "nuxt.config.js",
    "nuxt.config.mjs",
    "nuxt.config.cjs",
    "nuxt.config.ts",
    "nuxt.config.mts",
];
export const ASTRO_CONFIG_FILENAMES = [
    "astro.config.js",
    "astro.config.mjs",
    "astro.config.cjs",
    "astro.config.ts",
    "astro.config.mts",
];
// Detection checks frameworks in this order; keep more specific signals first.
export const FRAMEWORKS = [
    {
        key: "nextjs",
        displayName: "Next.js",
        buildType: "nextjs",
        aliases: ["nextjs", "next", "next.js"],
        detectPackages: ["next"],
        detectConfigFiles: NEXT_CONFIG_FILENAMES,
        usesEntrypoint: false,
        defaultEntrypoint: null,
        hasLocalDevServer: true,
        defaultHttpPort: 3000,
    },
    {
        key: "nuxt",
        displayName: "Nuxt",
        buildType: "nuxt",
        aliases: ["nuxt", "nuxtjs", "nuxt.js"],
        detectPackages: ["nuxt"],
        detectConfigFiles: NUXT_CONFIG_FILENAMES,
        usesEntrypoint: false,
        defaultEntrypoint: null,
        hasLocalDevServer: false,
        defaultHttpPort: 3000,
    },
    {
        key: "astro",
        displayName: "Astro",
        buildType: "astro",
        aliases: ["astro"],
        detectPackages: ["astro"],
        detectConfigFiles: ASTRO_CONFIG_FILENAMES,
        usesEntrypoint: false,
        defaultEntrypoint: null,
        hasLocalDevServer: false,
        defaultHttpPort: 4321,
    },
    {
        key: "hono",
        displayName: "Hono",
        buildType: "bun",
        aliases: ["hono"],
        detectPackages: ["hono"],
        detectConfigFiles: [],
        usesEntrypoint: true,
        defaultEntrypoint: "src/index.ts",
        hasLocalDevServer: true,
        defaultHttpPort: 3000,
    },
    {
        key: "nestjs",
        displayName: "NestJS",
        buildType: "nestjs",
        aliases: ["nestjs", "nest"],
        detectPackages: ["@nestjs/core"],
        detectConfigFiles: ["nest-cli.json"],
        usesEntrypoint: false,
        defaultEntrypoint: null,
        hasLocalDevServer: false,
        defaultHttpPort: 3000,
    },
    {
        key: "tanstack-start",
        displayName: "TanStack Start",
        buildType: "tanstack-start",
        aliases: [
            "tanstack-start",
            "tanstack",
            "@tanstack/react-start",
            "@tanstack/solid-start",
        ],
        detectPackages: ["@tanstack/react-start", "@tanstack/solid-start"],
        detectConfigFiles: [],
        usesEntrypoint: false,
        defaultEntrypoint: null,
        hasLocalDevServer: false,
        defaultHttpPort: 3000,
    },
    {
        key: "bun",
        displayName: "Bun",
        buildType: "bun",
        aliases: ["bun"],
        detectPackages: [],
        detectConfigFiles: [],
        usesEntrypoint: true,
        defaultEntrypoint: null,
        hasLocalDevServer: true,
        defaultHttpPort: 3000,
    },
];
export const FRAMEWORK_KEYS = FRAMEWORKS.map((framework) => framework.key);
/**
 * Build types whose preview build consumes committed build settings. The
 * others (nuxt, astro) run their framework CLI and stage fixed output, so a
 * config `build` block has nothing to apply to.
 */
export const CONFIG_BACKED_BUILD_TYPES = [
    "nextjs",
    "tanstack-start",
    "bun",
];
/** Build types that consume a user-provided source entrypoint. */
export const ENTRYPOINT_BUILD_TYPES = [
    ...new Set(FRAMEWORKS.filter((framework) => framework.usesEntrypoint).map((framework) => framework.buildType)),
];
/** Build types `app run` can start a local dev server for. */
export const LOCAL_DEV_BUILD_TYPES = [
    ...new Set(FRAMEWORKS.filter((framework) => framework.hasLocalDevServer).map((framework) => framework.buildType)),
];
const DEFAULT_HTTP_PORT_BY_BUILD_TYPE = (() => {
    const ports = new Map();
    for (const framework of FRAMEWORKS) {
        const existing = ports.get(framework.buildType);
        if (existing !== undefined && existing !== framework.defaultHttpPort) {
            throw new Error(`Conflicting defaultHttpPort for build type "${framework.buildType}".`);
        }
        ports.set(framework.buildType, framework.defaultHttpPort);
    }
    return Object.fromEntries(ports);
})();
/**
 * Default HTTP port for a build type, sourced from the framework registry so
 * build strategies and deploy front doors agree on where the gateway routes.
 */
export function defaultHttpPortForBuildType(buildType) {
    const port = DEFAULT_HTTP_PORT_BY_BUILD_TYPE[buildType];
    if (port === undefined) {
        throw new Error(`Unknown framework build type "${buildType}".`);
    }
    return port;
}
export function frameworkByKey(key) {
    const framework = FRAMEWORKS.find((candidate) => candidate.key === key);
    if (!framework) {
        throw new Error(`Unknown framework key "${key}".`);
    }
    return framework;
}
export function frameworkFromAlias(value) {
    const normalized = value.trim().toLowerCase();
    return (FRAMEWORKS.find((framework) => framework.aliases.includes(normalized)) ??
        null);
}
export function isConfigBackedBuildType(value) {
    return CONFIG_BACKED_BUILD_TYPES.includes(value);
}
