// @prisma/compute-sdk/dist/config/index.js
/**
 * The compute config contract and its tooling: types, validation, discovery,
 * loading, and serialization for `prisma.compute.ts`.
 *
 * Consumers (CLI, build-runner, scaffolding) share this module so every part
 * of the platform reads the same deploy graph. Consumer-specific concerns —
 * project/branch resolution, prompts, flag precedence — stay with consumers.
 */
export { ASTRO_CONFIG_FILENAMES, CONFIG_BACKED_BUILD_TYPES, defaultHttpPortForBuildType, ENTRYPOINT_BUILD_TYPES, FRAMEWORK_KEYS, FRAMEWORKS, frameworkByKey, frameworkFromAlias, isConfigBackedBuildType, LOCAL_DEV_BUILD_TYPES, NEXT_CONFIG_FILENAMES, NUXT_CONFIG_FILENAMES, } from "./frameworks.js";
export { COMPUTE_CONFIG_FILENAME, COMPUTE_CONFIG_FILENAMES, ComputeConfigAmbiguousError, ComputeConfigLoadError, findComputeConfigCandidates, findComputeConfigDir, loadComputeConfig, } from "./load.js";
export { ComputeConfigInvalidError, ComputeConfigTargetRequiredError, ComputeConfigTargetUnknownError, computeTargetAppDir, inferComputeTargetFromCwd, normalizeComputeConfig, selectComputeDeployTarget, } from "./normalize.js";
export { serializeComputeConfig } from "./serialize.js";
export { resolveSourceRoot, sourceRootLineage } from "./source-root.js";
export { COMPUTE_FRAMEWORKS, defineComputeConfig, } from "./types.js";
