// @prisma/cli/dist/controllers/app-env-api.js
import { CliError, authRequiredError } from "../shell/errors.js";
//#region src/controllers/app-env-api.ts
async function findVariableByNaturalKey(client, projectId, key, resolved, signal) {
	const { data, error, response } = await client.GET("/v1/environment-variables", {
		params: { query: {
			projectId,
			class: resolved.apiTarget.class,
			key
		} },
		signal
	});
	if (error || !data) throw apiCallError(`Failed to look up ${key}`, response, error);
	return data.data.filter((row) => rowMatchesExactScope(row, resolved))[0] ?? null;
}
function toMetadata(row, requestedScope) {
	const rowScope = row.branchId === null ? {
		kind: "role",
		role: row.class
	} : requestedScope;
	return {
		id: row.id,
		key: row.key,
		scope: rowScope,
		source: formatDescriptorLabel(rowScope),
		isManagedBySystem: row.isManagedBySystem,
		updatedAt: row.updatedAt
	};
}
function rowMatchesExactScope(row, resolved) {
	return row.class === resolved.apiTarget.class && row.branchId === resolved.apiTarget.branchId;
}
function apiCallError(summary, response, error) {
	const status = response?.status ?? 0;
	const apiCode = error?.error?.code;
	const apiMessage = error?.error?.message;
	const apiHint = error?.error?.hint;
	if (status === 401 || status === 403) return authRequiredError(["prisma auth login"]);
	return new CliError({
		code: apiCode ?? "ENV_API_ERROR",
		domain: "app",
		summary,
		why: apiMessage ?? `The Management API returned status ${status || "unknown"}.`,
		fix: apiHint ?? "Re-run with --trace for the underlying API response details.",
		exitCode: 1,
		nextSteps: []
	});
}
function formatDescriptorLabel(scope) {
	if (scope.kind === "role") return scope.role ?? "unknown";
	if (scope.kind === "overview") return "overview";
	return `branch:${scope.branchName ?? scope.branchId ?? "unknown"}`;
}
//#endregion
export { apiCallError, findVariableByNaturalKey, toMetadata };
