// @prisma/cli/dist/controllers/auth.js
import { CLIENT_ID, SERVICE_TOKEN_ENV_VAR, getApiBaseUrl } from "../lib/auth/client.js";
import { FileTokenStorage, WorkspaceSelectionError } from "../adapters/token-storage.js";
import { authRequiredError, usageError, workspaceAmbiguousError, workspaceNotAuthenticatedError, workspaceSwitchUnavailableError } from "../shell/errors.js";
import { canPrompt } from "../shell/runtime.js";
import { performLogin, performLogout, readAuthState } from "../lib/auth/auth-ops.js";
import { createAuthUseCases } from "../use-cases/auth.js";
import { createCliUseCaseGateways } from "../use-cases/create-cli-gateways.js";
import { createSelectPromptPort } from "./select-prompt-port.js";
import { createManagementApiSdk } from "@prisma/management-api-sdk";
//#region src/controllers/auth.ts
function isRealMode(context) {
	return !context.runtime.fixturePath && !context.runtime.env.PRISMA_CLI_MOCK_FIXTURE_PATH;
}
async function runAuthLogin(context, options) {
	let result;
	if (isRealMode(context)) {
		await performLogin(context.runtime.env, context.runtime.signal);
		result = await readAuthState(context.runtime.env, context.runtime.signal);
	} else result = await loginWithSelectionFlow(context, createAuthUseCases(createCliUseCaseGateways(context)), options);
	return createAuthSuccess("auth.login", result, ["prisma-cli auth whoami", "prisma-cli project list"]);
}
async function runAuthLogout(context) {
	let result;
	if (isRealMode(context)) {
		await performLogout(context.runtime.env, context.runtime.signal);
		result = await readAuthState(context.runtime.env, context.runtime.signal);
	} else result = await createAuthUseCases(createCliUseCaseGateways(context)).logout();
	return createAuthSuccess("auth.logout", result, ["prisma-cli auth login"]);
}
async function runAuthWhoAmI(context) {
	let result;
	if (isRealMode(context)) result = await readAuthState(context.runtime.env, context.runtime.signal);
	else result = await createAuthUseCases(createCliUseCaseGateways(context)).whoami();
	return createAuthSuccess("auth.whoami", result, result.authenticated ? [] : ["prisma-cli auth login"]);
}
async function runAuthWorkspaceList(context) {
	const result = isRealMode(context) ? await listRealAuthWorkspaces(context) : await createAuthUseCases(createCliUseCaseGateways(context)).listWorkspaces();
	return {
		command: "auth.workspace.list",
		result,
		warnings: [],
		nextSteps: result.workspaces.length === 0 ? ["prisma-cli auth login"] : []
	};
}
async function runAuthWorkspaceUse(context, workspaceRef) {
	const trimmedWorkspaceRef = workspaceRef?.trim();
	const selectedWorkspaceRef = trimmedWorkspaceRef ? trimmedWorkspaceRef : await selectWorkspaceSession(context);
	return {
		command: "auth.workspace.use",
		result: isRealMode(context) ? await useRealAuthWorkspace(context, selectedWorkspaceRef) : await createAuthUseCases(createCliUseCaseGateways(context)).useWorkspace(selectedWorkspaceRef),
		warnings: [],
		nextSteps: ["prisma-cli auth whoami", "prisma-cli project list"]
	};
}
async function runAuthWorkspaceLogout(context, workspaceRef) {
	if (!workspaceRef?.trim()) throw usageError("Workspace required", "auth workspace logout needs a workspace id or cached workspace name.", "Pass a workspace from prisma-cli auth workspace list.", ["prisma-cli auth workspace list"], "auth");
	const result = isRealMode(context) ? await logoutRealAuthWorkspace(context, workspaceRef) : await createAuthUseCases(createCliUseCaseGateways(context)).logoutWorkspace(workspaceRef);
	return {
		command: "auth.workspace.logout",
		result,
		warnings: [],
		nextSteps: result.activeWorkspace ? ["prisma-cli auth workspace list"] : ["prisma-cli auth workspace list", "prisma-cli auth workspace use <id>"]
	};
}
async function requireAuthenticatedAuthState(context) {
	if (isRealMode(context)) {
		const current = await readAuthState(context.runtime.env, context.runtime.signal);
		if (current.authenticated) return current;
		if (!canPrompt(context)) throw authRequiredError();
		await performLogin(context.runtime.env, context.runtime.signal);
		return readAuthState(context.runtime.env, context.runtime.signal);
	}
	const useCases = createAuthUseCases(createCliUseCaseGateways(context));
	const current = await useCases.whoami();
	if (current.authenticated) return current;
	if (!canPrompt(context)) throw authRequiredError();
	return loginWithSelectionFlow(context, useCases, {});
}
async function listRealAuthWorkspaces(context) {
	const rawServiceToken = context.runtime.env[SERVICE_TOKEN_ENV_VAR];
	const storage = new FileTokenStorage(context.runtime.env, context.runtime.signal);
	const localWorkspaces = await hydrateLocalAuthWorkspaces(context, storage, await storage.listWorkspaces());
	if (rawServiceToken !== void 0) {
		const authState = await readAuthState(context.runtime.env, context.runtime.signal);
		return {
			authSource: authState.authenticated ? "service_token" : "none",
			activeWorkspace: authState.workspace,
			workspaces: [...authState.workspace ? [{
				...authState.workspace,
				credentialWorkspaceId: null,
				active: true,
				source: "service_token",
				switchable: false,
				lastSeenAt: null
			}] : [], ...localWorkspaces.map((workspace) => ({
				...toAuthWorkspace(workspace),
				credentialWorkspaceId: workspace.credentialWorkspaceId,
				active: false,
				source: "oauth",
				switchable: false,
				lastSeenAt: workspace.lastSeenAt
			}))]
		};
	}
	const active = localWorkspaces.find((workspace) => workspace.active) ?? null;
	return {
		authSource: localWorkspaces.length > 0 ? "oauth" : "none",
		activeWorkspace: active ? toAuthWorkspace(active) : null,
		workspaces: localWorkspaces.map((workspace) => ({
			...toAuthWorkspace(workspace),
			credentialWorkspaceId: workspace.credentialWorkspaceId,
			active: workspace.active,
			source: "oauth",
			switchable: true,
			lastSeenAt: workspace.lastSeenAt
		}))
	};
}
async function useRealAuthWorkspace(context, workspaceRef) {
	if (context.runtime.env["PRISMA_SERVICE_TOKEN"] !== void 0) throw workspaceSwitchUnavailableError();
	const storage = new FileTokenStorage(context.runtime.env, context.runtime.signal);
	await hydrateLocalAuthWorkspaces(context, storage, await storage.listWorkspaces());
	try {
		const result = await storage.useWorkspace(workspaceRef);
		return {
			previousWorkspace: result.previous ? toAuthWorkspace(result.previous) : null,
			workspace: toAuthWorkspace(result.selected)
		};
	} catch (error) {
		if (error instanceof WorkspaceSelectionError) {
			if (error.reason === "ambiguous") throw workspaceAmbiguousError(error.workspaceRef ?? workspaceRef, error.matches.map((match) => ({
				id: match.id,
				name: match.name,
				credentialWorkspaceId: match.credentialWorkspaceId
			})));
			throw workspaceNotAuthenticatedError(error.workspaceRef ?? workspaceRef);
		}
		throw error;
	}
}
async function logoutRealAuthWorkspace(context, workspaceRef) {
	const storage = new FileTokenStorage(context.runtime.env, context.runtime.signal);
	await hydrateLocalAuthWorkspaces(context, storage, await storage.listWorkspaces());
	try {
		const result = await storage.logoutWorkspace(workspaceRef);
		return {
			workspace: toAuthWorkspace(result.workspace),
			wasActive: result.wasActive,
			activeWorkspace: result.activeWorkspace ? toAuthWorkspace(result.activeWorkspace) : null
		};
	} catch (error) {
		if (error instanceof WorkspaceSelectionError) {
			if (error.reason === "ambiguous") throw workspaceAmbiguousError(error.workspaceRef ?? workspaceRef, error.matches.map((match) => ({
				id: match.id,
				name: match.name,
				credentialWorkspaceId: match.credentialWorkspaceId
			})));
			throw workspaceNotAuthenticatedError(error.workspaceRef ?? workspaceRef);
		}
		throw error;
	}
}
async function selectWorkspaceSession(context) {
	const realMode = isRealMode(context);
	if (realMode && context.runtime.env["PRISMA_SERVICE_TOKEN"] !== void 0) throw workspaceSwitchUnavailableError();
	const workspaces = (realMode ? await listRealAuthWorkspaces(context) : await createAuthUseCases(createCliUseCaseGateways(context)).listWorkspaces()).workspaces.filter((workspace) => workspace.switchable);
	if (workspaces.length === 0) throw usageError("No authenticated workspaces", "There are no local OAuth workspace sessions to select.", "Run prisma-cli auth login and authorize a workspace.", ["prisma-cli auth login"], "auth");
	if (workspaces.length === 1) return workspaces[0].id;
	if (!canPrompt(context)) throw usageError("Interactive workspace selection unavailable", "auth workspace use needs an interactive terminal when no workspace is provided and more than one workspace is available.", "Run prisma-cli auth workspace use <id-or-name> with a workspace from prisma-cli auth workspace list.", ["prisma-cli auth workspace list"], "auth");
	return (await createSelectPromptPort(context).select({
		message: "Select a workspace",
		choices: workspaces.map((workspace) => ({
			label: `${workspace.name} (${workspace.id})${workspace.active ? " active" : ""}`,
			value: workspace
		}))
	})).id;
}
async function hydrateLocalAuthWorkspaces(context, storage, workspaces) {
	const candidates = workspaces.filter(needsWorkspaceMetadataHydration);
	if (candidates.length === 0) return workspaces;
	const tokensByCredentialWorkspaceId = new Map((await storage.listWorkspaceTokens()).map((tokens) => [tokens.workspaceId, tokens]));
	let nextWorkspaces = workspaces;
	for (const workspace of candidates) {
		const tokens = tokensByCredentialWorkspaceId.get(workspace.credentialWorkspaceId);
		if (!tokens) continue;
		const resolved = await resolveOAuthWorkspaceMetadata(context, tokens);
		if (!resolved) continue;
		await rememberResolvedWorkspaceMetadata(context, storage, tokens, resolved);
		nextWorkspaces = nextWorkspaces.map((candidate) => candidate.credentialWorkspaceId === workspace.credentialWorkspaceId ? {
			...candidate,
			id: resolved.id,
			name: resolved.name,
			lastSeenAt: (/* @__PURE__ */ new Date()).toISOString()
		} : candidate);
	}
	return nextWorkspaces;
}
async function rememberResolvedWorkspaceMetadata(context, storage, tokens, resolved) {
	try {
		await storage.rememberWorkspace(tokens.workspaceId, resolved);
	} catch {
		context.runtime.signal?.throwIfAborted();
	}
}
function needsWorkspaceMetadataHydration(workspace) {
	return workspace.id === workspace.credentialWorkspaceId || workspace.name === "Unknown workspace" || workspace.name === workspace.credentialWorkspaceId;
}
async function resolveOAuthWorkspaceMetadata(context, tokens) {
	const sdk = createManagementApiSdk({
		clientId: CLIENT_ID,
		redirectUri: "http://localhost:0/auth/callback",
		tokenStorage: createSingleWorkspaceTokenStorage(new FileTokenStorage(context.runtime.env, context.runtime.signal, { activateOnSetTokens: false }), tokens),
		apiBaseUrl: getApiBaseUrl(context.runtime.env)
	});
	try {
		const { data } = await sdk.client.GET("/v1/workspaces/{id}", {
			params: { path: { id: tokens.workspaceId } },
			signal: context.runtime.signal
		});
		const id = stringOrNull(data?.data?.id) ?? tokens.workspaceId;
		const name = stringOrNull(data?.data?.name) ?? id;
		if (id === tokens.workspaceId && name === tokens.workspaceId) return null;
		return {
			id,
			name
		};
	} catch {
		context.runtime.signal?.throwIfAborted();
		return null;
	}
}
function createSingleWorkspaceTokenStorage(storage, initialTokens) {
	let currentTokens = initialTokens;
	return {
		getTokens: async () => currentTokens,
		setTokens: async (tokens) => {
			currentTokens = tokens;
			await storage.setTokens(tokens);
		},
		clearTokens: async () => {
			const tokens = currentTokens;
			currentTokens = null;
			if (tokens) await storage.clearTokensIfCurrent(tokens);
		}
	};
}
function stringOrNull(value) {
	return typeof value === "string" && value.trim().length > 0 ? value.trim() : null;
}
function toAuthWorkspace(workspace) {
	return {
		id: workspace.id,
		name: workspace.name
	};
}
async function loginWithSelectionFlow(context, useCases, options) {
	const selection = await resolveLoginSelection(useCases, canPrompt(context) ? createSelectPromptPort(context) : null, options);
	return useCases.login(selection);
}
async function resolveLoginSelection(useCases, prompt, options) {
	const provider = options.provider ? (await useCases.resolveProvider(options.provider)).id : (await selectProvider(useCases, prompt)).id;
	const user = options.user ? await useCases.resolveUserForProvider(provider, options.user) : await selectUser(useCases, prompt, provider);
	const workspace = options.workspace ? await useCases.resolveWorkspaceForUser(user.id, options.workspace) : await selectWorkspace(useCases, prompt, user.id);
	return {
		provider,
		userId: user.id,
		workspaceId: workspace.id
	};
}
async function selectProvider(useCases, prompt) {
	if (!prompt) throw nonInteractiveLoginError("Re-run prisma-cli auth login in a TTY, or provide --provider and --user, and --workspace when required.");
	const providers = await useCases.listProviders();
	return prompt.select({
		message: "Select a provider",
		choices: providers.map((provider) => ({
			label: provider.name,
			value: provider
		}))
	});
}
async function selectUser(useCases, prompt, provider) {
	const users = await useCases.listUsersForProvider(provider);
	if (!prompt) throw nonInteractiveLoginError("Re-run prisma-cli auth login in a TTY, or provide --provider and --user, and --workspace when required.");
	return prompt.select({
		message: "Select a user",
		choices: users.map((user) => ({
			label: `${user.name} <${user.email}>`,
			value: user
		}))
	});
}
async function selectWorkspace(useCases, prompt, userId) {
	const workspaces = await useCases.listWorkspacesForUser(userId);
	if (workspaces.length === 1) return workspaces[0];
	if (!prompt) throw usageError("Login requires explicit selectors in non-interactive mode", "The selected mock user belongs to more than one workspace and the shell cannot prompt in the current mode.", "Re-run prisma-cli auth login in a TTY, or provide --workspace.", ["prisma-cli auth login"], "auth");
	return prompt.select({
		message: "Select a workspace",
		choices: workspaces.map((workspace) => ({
			label: `${workspace.name} (${workspace.id})`,
			value: workspace
		}))
	});
}
function nonInteractiveLoginError(fix) {
	return usageError("Login requires explicit selectors in non-interactive mode", "The fixture mode cannot prompt in the current mode.", fix, ["prisma-cli auth login"], "auth");
}
function createAuthSuccess(command, result, nextSteps) {
	return {
		command,
		result,
		warnings: [],
		nextSteps
	};
}
//#endregion
export { requireAuthenticatedAuthState, runAuthLogin, runAuthLogout, runAuthWhoAmI, runAuthWorkspaceList, runAuthWorkspaceLogout, runAuthWorkspaceUse };
