// @prisma/cli/dist/controllers/branch.js
import { CliError, authRequiredError, workspaceRequiredError } from "../shell/errors.js";
import { projectResolutionErrorToCliError, resolveProjectTarget } from "../lib/project/resolution.js";
import { requireComputeAuth } from "../lib/auth/guard.js";
import { createCliUseCaseGateways } from "../use-cases/create-cli-gateways.js";
import { requireAuthenticatedAuthState } from "./auth.js";
import { listRealWorkspaceProjects } from "./project.js";
import { createBranchUseCases } from "../use-cases/branch.js";
//#region src/controllers/branch.ts
function isRealMode(context) {
	return !context.runtime.fixturePath && !context.runtime.env.PRISMA_CLI_MOCK_FIXTURE_PATH;
}
async function runBranchList(context) {
	if (isRealMode(context)) return {
		command: "branch.list",
		result: await listRealBranches(context),
		warnings: [],
		nextSteps: []
	};
	return {
		command: "branch.list",
		result: await createBranchUseCases(createCliUseCaseGateways(context)).list(),
		warnings: [],
		nextSteps: []
	};
}
async function listRealBranches(context) {
	const authState = await requireAuthenticatedAuthState(context);
	const client = await requireComputeAuth(context.runtime.env, context.runtime.signal);
	if (!client) throw authRequiredError(["prisma-cli auth login"]);
	const workspace = authState.workspace;
	if (!workspace) throw workspaceRequiredError();
	const targetResult = await resolveProjectTarget({
		context,
		workspace,
		listProjects: () => listRealWorkspaceProjects(client, workspace, context.runtime.signal)
	});
	if (targetResult.isErr()) throw projectResolutionErrorToCliError(targetResult.error);
	const target = targetResult.value;
	const branches = await listBranches(client, target.project.id, context.runtime.signal);
	return {
		projectId: target.project.id,
		projectName: target.project.name,
		verboseContext: {
			workspace,
			project: target.project,
			resolution: target.resolution
		},
		branches: sortBranches(branches.map(toBranchSummary))
	};
}
function sortBranches(branches) {
	return branches.slice().sort((left, right) => {
		const leftRank = branchOrder(left);
		const rightRank = branchOrder(right);
		if (leftRank !== rightRank) return leftRank - rightRank;
		return left.name.localeCompare(right.name);
	});
}
function branchOrder(branch) {
	return branch.role === "production" ? 0 : 1;
}
async function listBranches(client, projectId, signal) {
	const collected = [];
	let cursor;
	while (true) {
		const query = {};
		if (cursor !== void 0) query.cursor = cursor;
		const { data, error, response } = await client.GET("/v1/projects/{projectId}/branches", {
			params: {
				path: { projectId },
				query
			},
			signal
		});
		if (error || !data) throw branchApiError("Failed to list branches", response, error);
		collected.push(...data.data);
		if (!data.pagination.hasMore || !data.pagination.nextCursor) break;
		cursor = data.pagination.nextCursor;
	}
	return collected;
}
function toBranchSummary(branch) {
	return {
		id: branch.id,
		name: branch.gitName,
		role: branch.role,
		envMap: branch.role
	};
}
function branchApiError(summary, response, error) {
	const status = response?.status ?? 0;
	return new CliError({
		code: error?.error?.code ?? "BRANCH_API_ERROR",
		domain: "branch",
		summary,
		why: error?.error?.message ?? `The Management API returned status ${status || "unknown"}.`,
		fix: error?.error?.hint ?? "Re-run with --trace for the underlying API response details.",
		exitCode: 1,
		nextSteps: []
	});
}
//#endregion
export { runBranchList };
