// @prisma/cli/dist/commands/project/index.js
import { attachCommandDescriptor } from "../../shell/command-meta.js";
import { addCompactGlobalFlags, addGlobalFlags } from "../../shell/global-flags.js";
import { configureRuntimeCommand } from "../../shell/runtime.js";
import { runProjectCreate, runProjectLink, runProjectList, runProjectShow } from "../../controllers/project.js";
import { runCommand } from "../../shell/command-runner.js";
import { renderProjectList, renderProjectSetup, renderProjectShow, serializeProjectList, serializeProjectSetup, serializeProjectShow } from "../../presenters/project.js";
import { createEnvCommand } from "../env.js";
import { Command } from "commander";
//#region src/commands/project/index.ts
function createProjectCommand(runtime) {
	const project = attachCommandDescriptor(configureRuntimeCommand(new Command("project"), runtime), "project");
	addCompactGlobalFlags(project);
	project.addCommand(createProjectListCommand(runtime));
	project.addCommand(createProjectShowCommand(runtime));
	project.addCommand(createProjectCreateCommand(runtime));
	project.addCommand(createProjectLinkCommand(runtime));
	project.addCommand(createEnvCommand(runtime));
	return project;
}
function createProjectCreateCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("create"), runtime), "project.create");
	command.argument("<name>", "Project name");
	addGlobalFlags(command);
	command.action(async (name, options) => {
		await runCommand(runtime, "project.create", options, (context) => runProjectCreate(context, String(name)), {
			renderHuman: (context, descriptor, result) => renderProjectSetup(context, descriptor, result),
			renderJson: (result) => serializeProjectSetup(result)
		});
	});
	return command;
}
function createProjectLinkCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("link"), runtime), "project.link");
	command.argument("[id-or-name]", "Project id or name");
	addGlobalFlags(command);
	command.action(async (projectRef, options) => {
		await runCommand(runtime, "project.link", options, (context) => runProjectLink(context, typeof projectRef === "string" ? projectRef : void 0), {
			renderHuman: (context, descriptor, result) => renderProjectSetup(context, descriptor, result),
			renderJson: (result) => serializeProjectSetup(result)
		});
	});
	return command;
}
function createProjectListCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("list"), runtime), "project.list");
	addGlobalFlags(command);
	command.action(async (options) => {
		await runCommand(runtime, "project.list", options, (context) => runProjectList(context), {
			renderHuman: (context, descriptor, result) => renderProjectList(context, descriptor, result),
			renderJson: (result) => serializeProjectList(result)
		});
	});
	return command;
}
function createProjectShowCommand(runtime) {
	const command = attachCommandDescriptor(configureRuntimeCommand(new Command("show"), runtime), "project.show");
	command.option("--project <id-or-name>", "Project id or name");
	addGlobalFlags(command);
	command.action(async (options) => {
		const projectRef = options.project;
		await runCommand(runtime, "project.show", options, (context) => runProjectShow(context, projectRef), {
			renderHuman: (context, descriptor, result) => renderProjectShow(context, descriptor, result),
			renderJson: (result) => serializeProjectShow(result)
		});
	});
	return command;
}
//#endregion
export { createProjectCommand };
