// @prisma/cli/dist/lib/git/local-status.js
import { execFile } from "node:child_process";
//#region src/lib/git/local-status.ts
async function readLocalGitState(cwd, signal) {
	signal.throwIfAborted();
	if ((await runGit(cwd, ["rev-parse", "--is-inside-work-tree"], signal))?.trim() !== "true") return null;
	const [ref, sha, status] = await Promise.all([
		runGit(cwd, [
			"symbolic-ref",
			"--quiet",
			"--short",
			"HEAD"
		], signal),
		runGit(cwd, [
			"rev-parse",
			"--short",
			"HEAD"
		], signal),
		runGit(cwd, ["status", "--porcelain"], signal)
	]);
	return {
		ref: cleanGitValue(ref),
		sha: cleanGitValue(sha),
		dirty: status === null ? null : status.trim().length > 0
	};
}
function runGit(cwd, args, signal) {
	return new Promise((resolve, reject) => {
		signal.throwIfAborted();
		execFile("git", args, {
			cwd,
			signal,
			timeout: 2e3
		}, (error, stdout) => {
			if (signal.aborted) {
				reject(error);
				return;
			}
			if (error) {
				resolve(null);
				return;
			}
			resolve(stdout);
		}).on("error", (error) => {
			if (signal.aborted) {
				reject(error);
				return;
			}
			resolve(null);
		});
	});
}
function cleanGitValue(value) {
	const cleaned = value?.trim();
	return cleaned ? cleaned : null;
}
//#endregion
export { readLocalGitState };
