// @prisma/cli/dist/lib/database/provider.js
import { CliError } from "../../shell/errors.js";
//#region src/lib/database/provider.ts
function createManagementDatabaseProvider(client) {
	return {
		async listDatabases(options) {
			const databases = [];
			let cursor;
			while (true) {
				const result = await client.GET("/v1/databases", {
					params: { query: {
						projectId: options.projectId,
						branchGitName: options.branchName,
						cursor
					} },
					signal: options.signal
				});
				if (result.error || !result.data) throw databaseApiError("Failed to list databases", result.response, result.error);
				databases.push(...result.data.data);
				if (!result.data.pagination.hasMore || !result.data.pagination.nextCursor) break;
				cursor = result.data.pagination.nextCursor;
			}
			return databases.map((database) => normalizeDatabase(database, options.projectId));
		},
		async showDatabase(databaseId, options) {
			const result = await client.GET("/v1/databases/{databaseId}", {
				params: { path: { databaseId } },
				signal: options?.signal
			});
			if (result.response?.status === 404) return null;
			if (result.error || !result.data) throw databaseApiError("Failed to show database", result.response, result.error);
			const database = result.data.data;
			return normalizeDatabase(database, requireDatabaseProjectId(database, options?.projectId));
		},
		async createDatabase(options) {
			const result = await client.POST("/v1/databases", {
				body: {
					projectId: options.projectId,
					name: options.name,
					source: { type: "empty" },
					...options.branchName ? { branchGitName: options.branchName } : {},
					...options.region ? { region: options.region } : {}
				},
				signal: options.signal
			});
			if (result.error || !result.data) throw databaseApiError("Failed to create database", result.response, result.error);
			return normalizeCreatedDatabase(result.data.data, options.projectId);
		},
		async removeDatabase(databaseId, options) {
			const result = await client.DELETE("/v1/databases/{databaseId}", {
				params: { path: { databaseId } },
				signal: options?.signal
			});
			if (result.error) throw databaseApiError("Failed to remove database", result.response, result.error);
		},
		async listConnections(databaseId, options) {
			const result = await client.GET("/v1/databases/{databaseId}/connections", {
				params: { path: { databaseId } },
				signal: options?.signal
			});
			if (result.error || !result.data) throw databaseApiError("Failed to list database connections", result.response, result.error);
			return result.data.data.map((connection) => normalizeConnection(connection, databaseId));
		},
		async createConnection(options) {
			const result = await client.POST("/v1/databases/{databaseId}/connections", {
				params: { path: { databaseId: options.databaseId } },
				body: { name: options.name },
				signal: options.signal
			});
			if (result.error || !result.data) throw databaseApiError("Failed to create database connection", result.response, result.error);
			return normalizeCreatedConnection(result.data.data, options.databaseId);
		},
		async removeConnection(connectionId, options) {
			const result = await client.DELETE("/v1/connections/{id}", {
				params: { path: { id: connectionId } },
				signal: options?.signal
			});
			if (result.error) throw databaseApiError("Failed to remove database connection", result.response, result.error);
		}
	};
}
function normalizeDatabase(database, fallbackProjectId) {
	return {
		id: database.id,
		name: database.name,
		projectId: database.projectId ?? fallbackProjectId,
		branchId: database.branchId ?? database.branch?.id ?? null,
		branchName: database.branchGitName ?? database.branchName ?? database.branch?.gitName ?? database.branch?.name ?? null,
		region: normalizeRegion(database),
		status: database.status ?? null,
		isDefault: database.isDefault ?? null,
		createdAt: database.createdAt ?? null
	};
}
function normalizeConnection(connection, fallbackDatabaseId) {
	return {
		id: connection.id,
		name: connection.name ?? connection.id,
		databaseId: connection.databaseId ?? fallbackDatabaseId,
		createdAt: connection.createdAt ?? null
	};
}
function normalizeCreatedDatabase(database, fallbackProjectId) {
	const rawConnection = database.connections?.[0];
	if (!rawConnection) throw new CliError({
		code: "DATABASE_CONNECTION_MISSING",
		domain: "database",
		summary: "Created database did not return a connection string",
		why: "The Management API created the database but did not include the one-time connection payload.",
		fix: "Create a connection explicitly with prisma-cli database connection create <database>.",
		exitCode: 1,
		nextSteps: [`prisma-cli database connection create ${database.id}`]
	});
	return {
		database: normalizeDatabase(database, fallbackProjectId),
		...normalizeCreatedConnection(rawConnection, database.id)
	};
}
function normalizeCreatedConnection(connection, fallbackDatabaseId) {
	const connectionString = extractConnectionString(connection);
	if (!connectionString) throw new CliError({
		code: "DATABASE_CONNECTION_STRING_MISSING",
		domain: "database",
		summary: "Created connection did not return a connection string",
		why: "Database connection strings are one-time-view secrets, but the Management API did not include one in this create response.",
		fix: "Create another database connection and store the returned URL immediately.",
		exitCode: 1,
		nextSteps: [`prisma-cli database connection create ${fallbackDatabaseId}`]
	});
	return {
		connection: normalizeConnection(connection, fallbackDatabaseId),
		connectionString
	};
}
function normalizeRegion(database) {
	if (typeof database.region === "string") return database.region;
	return database.region?.id ?? database.regionId ?? null;
}
function requireDatabaseProjectId(database, fallbackProjectId) {
	const projectId = database.projectId ?? fallbackProjectId;
	if (projectId) return projectId;
	throw new CliError({
		code: "DATABASE_API_ERROR",
		domain: "database",
		summary: "Database response did not include a project id",
		why: "The Management API returned database metadata without project context.",
		fix: "Re-run with --trace for the underlying API response details.",
		exitCode: 1,
		nextSteps: []
	});
}
function extractConnectionString(connection) {
	return connection.endpoints?.pooled?.connectionString ?? connection.connectionString ?? connection.endpoints?.direct?.connectionString ?? connection.endpoints?.accelerate?.connectionString ?? null;
}
function databaseApiError(summary, response, error) {
	const status = response?.status ?? 0;
	return new CliError({
		code: error?.error?.code ?? "DATABASE_API_ERROR",
		domain: "database",
		summary,
		why: error?.error?.message ?? `The Management API returned status ${status || "unknown"}.`,
		fix: error?.error?.hint ?? "Re-run with --trace for the underlying API response details.",
		exitCode: 1,
		nextSteps: []
	});
}
//#endregion
export { createManagementDatabaseProvider, normalizeConnection, normalizeDatabase };
