// @prisma/cli/dist/shell/command-runner.js
import { CliError, authConfigInvalidError, authRequiredError, commandCanceledError } from "./errors.js";
import { getCommandDescriptor } from "./command-meta.js";
import { resolveGlobalFlags } from "./global-flags.js";
import { createCommandContext } from "./runtime.js";
import { cliErrorToJson, writeHumanError, writeHumanLines, writeJsonError, writeJsonEvent, writeJsonSuccess } from "./output.js";
import { collectCommandDiagnostics } from "../lib/diagnostics.js";
import { renderCommandDiagnostics } from "./diagnostics-output.js";
import { AuthError } from "@prisma/management-api-sdk";
//#region src/shell/command-runner.ts
function toCliError(error, runtime) {
	if (isCancellationError(error) || runtime.signal.aborted) return commandCanceledError();
	if (error instanceof CliError) return error;
	if (error instanceof AuthError) return authRequiredError(["prisma-cli auth login"], { debug: error.message });
	if (error instanceof Error && error.message.startsWith("PRISMA_SERVICE_TOKEN is set but empty")) return authConfigInvalidError(error.message);
	return null;
}
function isCancellationError(error) {
	if (!(error instanceof Error)) return false;
	return error.name === "AbortError" || error.name === "CancelledError";
}
async function runCommand(runtime, commandName, options, handler, presenter) {
	const context = await createCommandContext(runtime, resolveGlobalFlags(runtime.argv, options));
	const descriptor = getCommandDescriptor(commandName);
	const startedAt = Date.now();
	try {
		await writeCommandSuccess(context, descriptor, await handler(context), presenter, Date.now() - startedAt);
	} catch (error) {
		const cliError = toCliError(error, runtime);
		if (cliError) {
			writeCommandError(context, commandName, cliError);
			process.exitCode = cliError.exitCode;
			return;
		}
		throw error;
	}
}
async function writeCommandSuccess(context, descriptor, success, presenter, durationMs) {
	if (context.flags.json) {
		writeJsonSuccess(context.output, {
			...success,
			result: presenter.renderJson ? presenter.renderJson(success.result) : success.result
		});
		return;
	}
	const stdout = presenter.renderStdout?.(context, descriptor, success.result) ?? [];
	if (context.flags.quiet) {
		writeStdoutLines(context, stdout);
		return;
	}
	const rendered = presenter.renderHuman(context, descriptor, success.result);
	const diagnostics = await renderBestEffortCommandDiagnostics(context, {
		enabled: context.flags.verbose && rendered.length > 0,
		durationMs
	});
	const humanLines = [...rendered, ...diagnostics];
	if (stdout.length > 0 && humanLines.length > 0) humanLines.push("");
	writeHumanLines(context.output, humanLines);
	writeStdoutLines(context, stdout);
}
function writeCommandError(context, commandName, cliError) {
	if (context.flags.json) {
		writeJsonError(context.output, commandName, cliError);
		return;
	}
	writeHumanError(context.output, context.ui, cliError, { trace: context.flags.trace });
}
function writeStdoutLines(context, lines) {
	if (lines.length > 0) context.output.stdout.write(`${lines.join("\n")}\n`);
}
async function renderBestEffortCommandDiagnostics(context, options) {
	if (!options.enabled) return [];
	try {
		return renderCommandDiagnostics(context, await collectCommandDiagnostics(context, { durationMs: options.durationMs }));
	} catch {
		return [];
	}
}
async function runStreamingCommand(runtime, commandName, options, handler) {
	const flags = resolveGlobalFlags(runtime.argv, options);
	const context = await createCommandContext(runtime, flags);
	try {
		await handler(context);
		if (flags.json) writeJsonEvent(context.output, {
			type: "success",
			command: commandName,
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			result: null,
			warnings: [],
			nextSteps: [],
			nextActions: []
		});
	} catch (error) {
		const cliError = toCliError(error, runtime);
		if (cliError) {
			if (flags.json) writeJsonEvent(context.output, {
				type: "error",
				command: commandName,
				timestamp: (/* @__PURE__ */ new Date()).toISOString(),
				error: cliErrorToJson(cliError),
				warnings: [],
				nextSteps: cliError.nextSteps,
				nextActions: cliError.nextActions
			});
			else writeHumanError(context.output, context.ui, cliError, { trace: flags.trace });
			process.exitCode = cliError.exitCode;
			return;
		}
		throw error;
	}
}
//#endregion
export { runCommand, runStreamingCommand };
