// @prisma/cli/dist/shell/global-flags.js
import { Option } from "commander";
//#region src/shell/global-flags.ts
const COMPACT_GLOBAL_OPTION_FLAGS = [
	"--json",
	"-q, --quiet",
	"-v, --verbose",
	"--trace",
	"--no-interactive",
	"-y, --yes",
	"--version"
];
function addGlobalFlags(command) {
	return command.option("--json", "Emit structured JSON output.").option("-q, --quiet", "Reduce human-oriented output.").option("-v, --verbose", "Increase human-oriented output detail.").option("--trace", "Show deeper diagnostics for failures.").option("-y, --yes", "Accept supported confirmation prompts.").addOption(new Option("--interactive", "Force interactive behavior when prompts are supported.")).addOption(new Option("--no-interactive", "Disable interactive behavior and prompts.")).addOption(new Option("--color", "Force color output in supported terminals.")).addOption(new Option("--no-color", "Disable color output."));
}
function addCompactGlobalFlags(command) {
	return command.option("--json", "Emit structured JSON output.").option("-q, --quiet", "Reduce human-oriented output.").option("-v, --verbose", "Increase human-oriented output detail.").option("--trace", "Show deeper diagnostics for failures.").addOption(new Option("--no-interactive", "Disable interactive behavior and prompts.")).option("-y, --yes", "Accept supported confirmation prompts.");
}
function getExplicitBoolean(argv, positive, negative) {
	for (let index = argv.length - 1; index >= 0; index -= 1) {
		const value = argv[index];
		if (value === positive) return true;
		if (value === negative) return false;
	}
}
function resolveGlobalFlags(argv, options) {
	return {
		json: options.json === true || argv.includes("--json"),
		quiet: options.quiet === true || argv.includes("--quiet") || argv.includes("-q"),
		verbose: options.verbose === true || argv.includes("--verbose") || argv.includes("-v"),
		trace: options.trace === true || argv.includes("--trace"),
		yes: options.yes === true || argv.includes("--yes") || argv.includes("-y"),
		interactive: getExplicitBoolean(argv, "--interactive", "--no-interactive"),
		color: getExplicitBoolean(argv, "--color", "--no-color")
	};
}
//#endregion
export { COMPACT_GLOBAL_OPTION_FLAGS, addCompactGlobalFlags, addGlobalFlags, resolveGlobalFlags };
