// @prisma/cli/dist/shell/help.js
import { formatDescriptorLabel, getDescriptorForCommand } from "./command-meta.js";
import { COMPACT_GLOBAL_OPTION_FLAGS, resolveGlobalFlags } from "./global-flags.js";
import { createShellUi, padDisplay, wrapText } from "./ui.js";
//#region src/shell/help.ts
function renderHelp(command, runtime) {
	const descriptor = getDescriptorForCommand(command);
	const ui = createShellUi(runtime, resolveGlobalFlags(runtime.argv, {}));
	const rail = ui.isTTY ? ui.dim("│") : "│";
	const lines = [`${formatDescriptorLabel(descriptor)} ${ui.dim("→")} ${ui.dim(descriptor.description)}`, ""];
	const visibleCommands = command.commands.filter((candidate) => candidate.name() !== "help" && !candidate.hidden);
	const visibleOptions = command.options.filter((candidate) => !candidate.hidden);
	if (visibleCommands.length > 0) lines.push(...renderCommandRows(rail, ui, visibleCommands));
	lines.push(...renderLongDescription(rail, ui, descriptor.longDescription));
	lines.push(...renderVisibleOptions(rail, ui, visibleCommands, visibleOptions));
	lines.push(...renderExamples(rail, descriptor.examples));
	lines.push(...renderDocsPath(rail, ui, descriptor.docsPath));
	lines.push("");
	return `${lines.join("\n")}`;
}
function renderLongDescription(rail, ui, longDescription) {
	if (!longDescription) return [];
	return [`${rail}`, ...wrapText(longDescription, Math.max(ui.width - 3, 40)).map((line) => `${rail}  ${line}`)];
}
function renderVisibleOptions(rail, ui, visibleCommands, visibleOptions) {
	if (visibleOptions.length === 0) return [];
	const lines = visibleCommands.length > 0 ? [`${rail}`] : [];
	if (shouldLabelGlobalOptions(visibleCommands, visibleOptions)) lines.push(`${rail}  Global options:`);
	return [...lines, ...renderOptionRows(rail, ui, visibleOptions)];
}
function shouldLabelGlobalOptions(visibleCommands, visibleOptions) {
	return visibleCommands.length > 0 && visibleOptions.every((option) => COMPACT_GLOBAL_OPTION_FLAGS.includes(option.flags));
}
function renderExamples(rail, examples) {
	if (!examples || examples.length === 0) return [];
	return [
		`${rail}`,
		`${rail}  Examples:`,
		...examples.map((example) => `${rail}    $ ${example}`)
	];
}
function renderDocsPath(rail, ui, docsPath) {
	if (!docsPath) return [];
	return [`${rail}`, `${rail}  ${ui.accent(padDisplay("Read more", 16))}  ${ui.link(docsPath)}`];
}
function renderCommandRows(rail, ui, commands) {
	return renderAlignedRows(rail, ui, commands.map((command) => {
		const descriptor = getDescriptorForCommand(command);
		return {
			term: renderCommandTerm(command),
			description: descriptor.description
		};
	}));
}
function renderOptionRows(rail, ui, options) {
	return renderAlignedRows(rail, ui, options.map((option) => ({
		term: option.flags,
		description: option.description || "",
		defaultValue: option.defaultValueDescription ?? formatDefaultValue(option.defaultValue)
	})));
}
function renderAlignedRows(rail, ui, rows) {
	const termWidth = rows.reduce((width, row) => Math.max(width, row.term.length), 0);
	const descriptionWidth = Math.max(ui.width - 3 - termWidth - 4, 30);
	const lines = [];
	for (const row of rows) {
		const [firstLine, ...rest] = wrapText(row.description, descriptionWidth, " ".repeat(termWidth + 2));
		lines.push(`${rail}  ${ui.accent(padDisplay(row.term, termWidth))}  ${firstLine}`);
		for (const line of rest) lines.push(`${rail}  ${" ".repeat(termWidth)}  ${line.trimStart()}`);
		if (row.defaultValue) lines.push(`${rail}  ${" ".repeat(termWidth)}  ${ui.dim(`default: ${row.defaultValue}`)}`);
	}
	return lines;
}
function renderCommandTerm(command) {
	const argumentsList = command.registeredArguments.map(renderArgumentLabel).join(" ");
	return `${command.name()}${argumentsList ? ` ${argumentsList}` : ""}`;
}
function renderArgumentLabel(argument) {
	const name = argument.name();
	return argument.required ? `<${name}>` : `[${name}]`;
}
function formatDefaultValue(value) {
	if (value === void 0) return null;
	if (Array.isArray(value)) return value.join(", ");
	if (typeof value === "boolean" || typeof value === "number") return String(value);
	if (typeof value === "string" && value.length > 0) return value;
	return null;
}
//#endregion
export { renderHelp };
