// @prisma/cli/dist/shell/output.js
import { renderNextSteps, renderSummaryLine } from "./ui.js";
//#region src/shell/output.ts
function writeJsonSuccess(output, success) {
	output.stdout.write(`${JSON.stringify({
		ok: true,
		nextActions: [],
		...success
	}, null, 2)}\n`);
}
function writeJsonEvent(output, event) {
	output.stdout.write(`${JSON.stringify(event)}\n`);
}
function cliErrorToJson(error) {
	return {
		code: error.code,
		domain: error.domain,
		severity: error.severity,
		summary: error.summary,
		why: error.why,
		fix: error.fix,
		where: error.where,
		meta: error.meta,
		docsUrl: error.docsUrl
	};
}
function formatUnexpectedError(error, trace) {
	const debug = error instanceof Error ? error.stack ?? error.message : String(error);
	if (trace) return `${debug}\n`;
	return [
		`Unexpected CLI error: ${error instanceof Error && error.message ? error.message : String(error)}`,
		"More: Re-run with --trace for deeper diagnostics",
		""
	].join("\n");
}
function writeJsonError(output, command, error) {
	output.stdout.write(`${JSON.stringify({
		ok: false,
		command,
		error: cliErrorToJson(error),
		warnings: [],
		nextSteps: error.nextSteps,
		nextActions: error.nextActions
	}, null, 2)}\n`);
}
function writeHumanLines(output, lines) {
	if (lines.length === 0) return;
	output.stderr.write(`${lines.join("\n")}\n`);
}
function writeHumanError(output, ui, error, options) {
	if (error.humanLines && error.humanLines.length > 0) {
		const lines = [...error.humanLines];
		if (options.trace && error.debug) {
			lines.push("");
			lines.push("Trace:");
			lines.push(...error.debug.trimEnd().split("\n"));
		}
		lines.push(...renderNextSteps(error.nextSteps));
		writeHumanLines(output, lines);
		return;
	}
	const lines = [renderSummaryLine(ui, "error", `${error.summary} [${error.code}]`)];
	if (error.where) lines.push(...["", `Where: ${error.where}`]);
	if (error.why) {
		if (!error.where) lines.push("");
		lines.push(`Why: ${error.why}`);
	}
	if (error.fix) lines.push(`Fix: ${error.fix}`);
	if (options.trace) {
		if (error.debug) {
			lines.push("");
			lines.push("Trace:");
			lines.push(...error.debug.trimEnd().split("\n"));
		}
	} else {
		lines.push("");
		lines.push("More: Re-run with --trace for deeper diagnostics");
	}
	lines.push(...renderNextSteps(error.nextSteps));
	writeHumanLines(output, lines);
}
//#endregion
export { cliErrorToJson, formatUnexpectedError, writeHumanError, writeHumanLines, writeJsonError, writeJsonEvent, writeJsonSuccess };
