// @prisma/cli/dist/shell/prompt.js
import { usageError } from "./errors.js";
import { confirm, isCancel, select, text } from "@clack/prompts";
//#region src/shell/prompt.ts
async function selectPrompt(options) {
	const promptOptions = options.choices.map((choice) => ({
		label: choice.label,
		value: choice.value
	}));
	const response = await select({
		input: options.input,
		output: options.output,
		message: options.message,
		options: promptOptions
	});
	if (isCancel(response)) throw usageError("Interactive prompt canceled", "The command was canceled before a selection was made.", "Re-run the command and choose an option to continue.");
	return response;
}
async function textPrompt(options) {
	const response = await text({
		input: options.input,
		output: options.output,
		message: options.message,
		placeholder: options.placeholder,
		validate: options.validate
	});
	if (isCancel(response)) throw usageError("Interactive prompt canceled", "The command was canceled before a value was entered.", "Re-run the command and provide a value to continue.");
	return response;
}
async function confirmPrompt(options) {
	const response = await confirm({
		input: options.input,
		output: options.output,
		message: options.message,
		initialValue: options.initialValue ?? false
	});
	if (isCancel(response)) throw usageError("Interactive prompt canceled", "The command was canceled before a confirmation was made.", "Re-run the command and choose an option to continue.");
	return response;
}
function disposePromptState(_input) {}
//#endregion
export { confirmPrompt, disposePromptState, selectPrompt, textPrompt };
