// @prisma/cli/dist/shell/runtime.js
import { LocalStateStore } from "../adapters/local-state.js";
import { MockApi } from "../adapters/mock-api.js";
import { createShellUi } from "./ui.js";
import { renderHelp } from "./help.js";
import { findComputeConfigDir } from "@prisma/compute-sdk/config";
import path from "node:path";
//#region src/shell/runtime.ts
const DEFAULT_STATE_DIR_NAME = path.join(".prisma", "cli");
function configureRuntimeCommand(command, runtime) {
	return command.helpCommand(false).configureHelp({ formatHelp: (configuredCommand) => renderHelp(configuredCommand, runtime) }).configureOutput({
		writeOut: (text) => {
			runtime.stderr.write(text);
		},
		writeErr: (text) => {
			runtime.stderr.write(text);
		},
		outputError: (text, write) => {
			write(text);
		}
	}).exitOverride();
}
async function createCommandContext(runtime, flags) {
	const fixturePath = runtime.fixturePath ?? runtime.env.PRISMA_CLI_MOCK_FIXTURE_PATH;
	const stateDir = await resolveStateDir(runtime);
	let loadedApi;
	if (fixturePath) loadedApi = await MockApi.load(fixturePath, runtime.signal);
	return {
		get api() {
			if (!loadedApi) throw new Error("context.api accessed in real mode. Set runtime.fixturePath or PRISMA_CLI_MOCK_FIXTURE_PATH to use fixture mode.");
			return loadedApi;
		},
		stateStore: new LocalStateStore(stateDir, runtime.signal),
		output: {
			stdout: runtime.stdout,
			stderr: runtime.stderr
		},
		flags,
		runtime,
		ui: createShellUi(runtime, flags)
	};
}
async function resolveStateDir(runtime) {
	const explicitStateDir = runtime.stateDir ?? runtime.env.PRISMA_CLI_STATE_DIR;
	if (explicitStateDir) return explicitStateDir;
	const projectDir = await findComputeConfigDir(runtime.cwd, runtime.signal);
	return path.join(projectDir ?? runtime.cwd, DEFAULT_STATE_DIR_NAME);
}
function canPrompt(context) {
	if (context.flags.json) return false;
	if (context.flags.interactive === false) return false;
	if (context.runtime.env.CI && context.flags.interactive !== true) return false;
	return Boolean(context.runtime.stdin.isTTY && context.runtime.stderr.isTTY);
}
//#endregion
export { canPrompt, configureRuntimeCommand, createCommandContext, resolveStateDir };
