// @prisma/cli/dist/shell/ui.js
import { createColors } from "colorette";
import stringWidth from "string-width";
import stripAnsi from "strip-ansi";
import wrapAnsi from "wrap-ansi";
//#region src/shell/ui.ts
const URL_CREDENTIALS_PATTERN = /:\/\/[^:@/\s]+:[^@/\s]+@/g;
const DEFAULT_WIDTH = 80;
function createShellUi(runtime, flags) {
	const isTTY = Boolean(runtime.stderr.isTTY);
	const colorEnabled = resolveColorEnabled(runtime, flags, isTTY);
	const colors = createColors({ useColor: colorEnabled });
	return {
		isTTY,
		colorEnabled,
		width: runtime.stderr.columns && runtime.stderr.columns > 0 ? runtime.stderr.columns : DEFAULT_WIDTH,
		quiet: flags.quiet,
		verbose: flags.verbose,
		trace: flags.trace,
		accent: (text) => colors.cyan(text),
		success: (text) => colors.greenBright(text),
		error: (text) => colors.redBright(text),
		warning: (text) => colors.yellow(text),
		info: (text) => colors.blue(text),
		link: (text) => colors.blue(text),
		dim: (text) => colors.dim(text),
		strong: (text) => colors.bold(text)
	};
}
function renderCommandHeader(ui, options) {
	if (!ui.isTTY) return [];
	const rows = options.rows ?? [];
	const lines = [`${ui.strong(options.commandLabel)} ${ui.dim("→")} ${ui.dim(options.description)}`, ""];
	const rail = ui.dim("│");
	const keyWidth = rows.length > 0 ? Math.max(...rows.map((row) => stringWidth(`${row.key}:`)), stringWidth("Read more")) : stringWidth("Read more");
	for (const row of rows) lines.push(`${rail}  ${ui.accent(padDisplay(`${row.key}:`, keyWidth))}  ${formatHeaderValue(ui, row)}`);
	if (rows.length > 0 || options.docsPath) lines.push(rail);
	if (options.docsPath) lines.push(`${rail}  ${ui.accent(padDisplay("Read more", keyWidth))}  ${ui.link(options.docsPath)}`);
	lines.push("");
	return lines;
}
function renderSummaryLine(ui, status, text) {
	return `${status === "success" ? ui.success("✔") : status === "error" ? ui.error("✘") : status === "warning" ? ui.warning("⚠") : ui.info("ℹ")} ${text}`;
}
function renderNextSteps(steps) {
	if (steps.length === 0) return [];
	return [
		"",
		steps.length === 1 ? "Next step:" : "Next steps:",
		...steps.map((step) => `- ${step}`)
	];
}
function renderVerboseBlock(ui, rows, options = {}) {
	if (!ui.verbose || rows.length === 0) return [];
	const title = options.title ?? "Details";
	const keyWidth = Math.max(...rows.map((row) => stringWidth(`${row.key}:`)));
	const rail = ui.dim("│");
	return [
		"",
		`${ui.dim(title)}:`,
		...rows.map((row) => `${rail}  ${ui.accent(padDisplay(`${row.key}:`, keyWidth))}  ${formatVerboseValue(ui, row)}`)
	];
}
function formatColumns(columns, widths) {
	return columns.map((value, index) => padDisplay(value, widths[index])).join("   ").trimEnd();
}
function wrapText(text, width, indent = "") {
	return wrapAnsi(text, width, {
		hard: false,
		trim: false
	}).split("\n").map((line, index) => index === 0 ? line : `${indent}${line}`);
}
function padDisplay(text, width) {
	const padding = Math.max(width - stringWidth(stripAnsi(text)), 0);
	return `${text}${" ".repeat(padding)}`;
}
function maskValue(value) {
	return maskEmailLocalParts(value).replace(URL_CREDENTIALS_PATTERN, "://****:****@");
}
function maskEmailLocalParts(value) {
	let masked = "";
	let segmentStart = 0;
	for (let index = 0; index < value.length; index += 1) {
		if (value[index] !== "@") continue;
		let localStart = index;
		while (localStart > segmentStart && isEmailLocalPartChar(value[localStart - 1])) localStart -= 1;
		if (localStart === index) continue;
		masked += `${value.slice(segmentStart, localStart)}****@`;
		segmentStart = index + 1;
	}
	return masked + value.slice(segmentStart);
}
function isEmailLocalPartChar(char) {
	return char >= "A" && char <= "Z" || char >= "a" && char <= "z" || char >= "0" && char <= "9" || char === "!" || char === "#" || char === "$" || char === "." || char === "&" || char === "'" || char === "*" || char === "%" || char === "+" || char === "-" || char === "/" || char === "=" || char === "?" || char === "^" || char === "_" || char === "`" || char === "{" || char === "|" || char === "}" || char === "~";
}
function resolveColorEnabled(runtime, flags, isTTY) {
	if (flags.color === true) return true;
	if (flags.color === false) return false;
	if (runtime.env.NO_COLOR !== void 0) return false;
	return isTTY;
}
function formatHeaderValue(ui, row) {
	const value = row.sensitive ? maskValue(row.value) : row.value;
	if (row.tone === "dim") return ui.dim(value);
	if (row.tone === "link") return ui.link(value);
	return value;
}
function formatVerboseValue(ui, row) {
	const value = row.sensitive ? maskValue(row.value) : row.value;
	if (row.tone === "dim") return ui.dim(value);
	if (row.tone === "success") return ui.success(value);
	if (row.tone === "warning") return ui.warning(value);
	if (row.tone === "link") return ui.link(value);
	return value;
}
//#endregion
export { createShellUi, formatColumns, maskValue, padDisplay, renderCommandHeader, renderNextSteps, renderSummaryLine, renderVerboseBlock, wrapText };
