// @prisma/cli/dist/cli2.js
import { CliError } from "./shell/errors.js";
import "./shell/prompt.js";
import { attachCommandDescriptor } from "./shell/command-meta.js";
import { addCompactGlobalFlags } from "./shell/global-flags.js";
import { createShellUi } from "./shell/ui.js";
import { configureRuntimeCommand, createCommandContext } from "./shell/runtime.js";
import { formatUnexpectedError, writeHumanError, writeJsonError, writeJsonSuccess } from "./shell/output.js";
import { createAppCommand } from "./commands/app/index.js";
import { createAuthCommand } from "./commands/auth/index.js";
import { createBranchCommand } from "./commands/branch/index.js";
import { createDatabaseCommand } from "./commands/database/index.js";
import { createGitCommand } from "./commands/git/index.js";
import { createProjectCommand } from "./commands/project/index.js";
import { getCliName, getCliVersion } from "./lib/version.js";
import { runVersion } from "./controllers/version.js";
import { createVersionCommand } from "./commands/version/index.js";
import { maybeWriteCachedUpdateNotification } from "./shell/update-check.js";
import process from "node:process";
import { Command, CommanderError, Option } from "commander";
//#region src/cli.ts
async function runCli(options = {}) {
	const runtime = resolveRuntime(options);
	const program = createProgram(runtime);
	process.exitCode = 0;
	try {
		await maybeWriteCachedUpdateNotification(runtime);
		if (runtime.argv.includes("--version")) return await handleVersionFlag(runtime);
		const bareHelpCommand = resolveBareHelpCommand(program, runtime.argv);
		if (bareHelpCommand) {
			runtime.stderr.write(bareHelpCommand.helpInformation());
			return 0;
		}
		await program.parseAsync(runtime.argv, { from: "user" });
		return process.exitCode ?? 0;
	} catch (error) {
		if (error instanceof CommanderError) return error.code === "commander.helpDisplayed" ? 0 : 2;
		runtime.stderr.write(formatUnexpectedError(error, runtime.argv.includes("--trace")));
		return 1;
	} finally {
		runtime.stdin;
	}
}
function createProgram(runtime) {
	const program = attachCommandDescriptor(configureRuntimeCommand(new Command(), runtime), "root");
	addCompactGlobalFlags(program);
	program.addOption(new Option("--version", "Print the CLI version and exit."));
	program.name("prisma").showSuggestionAfterError();
	program.addCommand(createVersionCommand(runtime));
	program.addCommand(createAuthCommand(runtime));
	program.addCommand(createProjectCommand(runtime));
	program.addCommand(createGitCommand(runtime));
	program.addCommand(createBranchCommand(runtime));
	program.addCommand(createDatabaseCommand(runtime));
	program.addCommand(createAppCommand(runtime));
	return program;
}
async function handleVersionFlag(runtime) {
	const wantsJson = runtime.argv.includes("--json");
	const output = {
		stdout: runtime.stdout,
		stderr: runtime.stderr
	};
	try {
		if (wantsJson) {
			const success = await runVersion(await createCommandContext(runtime, buildVersionFlagFlags(runtime)));
			writeJsonSuccess(output, {
				command: success.command,
				result: { version: success.result.cli.version },
				warnings: success.warnings,
				nextSteps: success.nextSteps
			});
			return 0;
		}
		const versionLine = `${getCliName()} ${getCliVersion()}`;
		runtime.stdout.write(`${versionLine}\n`);
		return 0;
	} catch (error) {
		if (error instanceof CliError) {
			if (wantsJson) writeJsonError(output, "version", error);
			else writeHumanError(output, createShellUi(runtime, buildVersionFlagFlags(runtime)), error, { trace: false });
			return error.exitCode;
		}
		throw error;
	}
}
function buildVersionFlagFlags(runtime) {
	return {
		json: runtime.argv.includes("--json"),
		quiet: false,
		verbose: false,
		trace: false,
		yes: false,
		interactive: void 0,
		color: void 0
	};
}
function resolveBareHelpCommand(program, argv) {
	if (argv.length === 0) return program;
	if (argv.length !== 1) return null;
	const candidate = program.commands.find((command) => command.name() === argv[0]) ?? null;
	if (!candidate) return null;
	if (candidate.commands.length === 0) return null;
	return candidate;
}
function resolveRuntime(options) {
	return {
		argv: options.argv ?? process.argv.slice(2),
		cwd: options.cwd ?? process.env.INIT_CWD ?? process.cwd(),
		env: options.env ?? process.env,
		signal: options.signal ?? new AbortController().signal,
		stdin: options.stdin ?? process.stdin,
		stdout: options.stdout ?? process.stdout,
		stderr: options.stderr ?? process.stderr,
		fixturePath: options.fixturePath,
		stateDir: options.stateDir
	};
}
//#endregion
export { runCli };
