// @prisma/cli/dist/lib/version.js
import { CliError } from "../shell/errors.js";
import { createRequire } from "node:module";
import process from "node:process";
//#region src/lib/version.ts
const requireFromHere = createRequire(import.meta.url);
function readPackageMetadata() {
	try {
		return requireFromHere("../../package.json");
	} catch {
		return {};
	}
}
function getCliVersion() {
	const pkg = readPackageMetadata();
	if (!pkg.version) throw new CliError({
		code: "VERSION_UNAVAILABLE",
		domain: "cli",
		summary: "CLI version metadata is missing from the installed package",
		why: "The bundled package.json could not be read or did not contain a version field.",
		fix: "Reinstall the CLI from the npm registry, or check your install path is intact.",
		exitCode: 1
	});
	return pkg.version;
}
function getCliName() {
	return "prisma-cli";
}
function detectInvocation(env, argv) {
	if (env.npm_config_user_agent?.startsWith("bun")) return "bunx";
	const normalizedExecPath = env.npm_execpath?.replace(/\\/g, "/").toLowerCase();
	const normalizedUserAgent = env.npm_config_user_agent?.toLowerCase();
	if (env.npm_lifecycle_event === "npx" || normalizedExecPath?.includes("/_npx/") || normalizedUserAgent?.includes("npx")) return "npx";
	const entry = (argv[1] ?? "").replace(/\\/g, "/").toLowerCase();
	if (entry.endsWith(".ts") || entry.includes("/tsx/")) return "dev";
	if (entry.includes("/_npx/")) return "npx";
	if (entry.includes("/.bun/")) return "bunx";
	if (entry.includes("/node_modules/.bin/") || /\/prisma-cli(\.cmd|\.exe)?$/.test(entry)) return "global";
	return "unknown";
}
function buildVersionResult(env, argv) {
	return {
		cli: {
			name: getCliName(),
			version: getCliVersion()
		},
		node: { version: process.version },
		os: {
			platform: process.platform,
			arch: process.arch
		},
		invocation: detectInvocation(env, argv)
	};
}
//#endregion
export { buildVersionResult, getCliName, getCliVersion };
