// @prisma/cli/dist/output/patterns.js
import { formatDescriptorLabel } from "../shell/command-meta.js";
import { maskValue, padDisplay, renderSummaryLine } from "../shell/ui.js";
import stringWidth from "string-width";
//#region src/output/patterns.ts
function renderList(input, ui) {
	const keyWidth = Math.max(stringWidth(`${input.parentContext.key}:`), ...input.items.map((item) => stringWidth(`⚬ ${item.noun}:`)), ...readMoreWidth(input.descriptor));
	const lines = renderCardTitle(input.descriptor, input.title, ui);
	lines.push(renderCardRow(ui, keyWidth, input.parentContext.key, input.parentContext.value));
	if (input.items.length === 0) lines.push(renderPlainCardLine(ui, ui.dim(input.emptyMessage)));
	else for (const item of input.items) lines.push(renderCardRow(ui, keyWidth, `⚬ ${item.noun}`, formatListItemValue(ui, item)));
	pushReadMore(lines, ui, keyWidth, input.descriptor);
	return lines;
}
function serializeList(input) {
	return {
		context: input.context,
		items: input.items.map((item) => ({
			name: item.label,
			id: item.id,
			status: item.status
		})),
		count: input.items.length
	};
}
function renderShow(input, ui) {
	const keyWidth = Math.max(0, ...input.fields.map((field) => stringWidth(`${field.key}:`)), ...readMoreWidth(input.descriptor));
	const lines = renderCardTitle(input.descriptor, input.title, ui);
	for (const field of input.fields) lines.push(renderCardRow(ui, keyWidth, field.key, formatValue(ui, field.value, field.tone, field.sensitive)));
	pushReadMore(lines, ui, keyWidth, input.descriptor);
	return lines;
}
function renderMutate(input, ui) {
	const rows = input.context;
	const keyWidth = Math.max(0, ...rows.map((row) => stringWidth(`${row.key}:`)), ...readMoreWidth(input.descriptor));
	const lines = renderCardTitle(input.descriptor, input.title, ui);
	for (const row of rows) lines.push(renderCardRow(ui, keyWidth, row.key, formatValue(ui, row.value, row.tone, row.sensitive)));
	pushReadMore(lines, ui, keyWidth, input.descriptor);
	lines.push("");
	lines.push(`${ui.warning("◇")} ${input.operationDescription}...`);
	lines.push(renderSummaryLine(ui, "success", `Applied ${input.operationCount} operation(s)`));
	for (const detail of input.details) lines.push(`  ${detail}`);
	for (const alert of input.alerts ?? []) {
		lines.push("");
		lines.push(renderSummaryLine(ui, alert.tone, alert.text));
	}
	return lines;
}
function renderCardTitle(descriptor, title, ui) {
	return [`${ui.strong(formatDescriptorLabel(descriptor))} ${ui.dim("→")} ${ui.dim(title)}`, ""];
}
function renderCardRow(ui, keyWidth, key, value) {
	return `${renderCardRail(ui)}  ${ui.accent(padDisplay(`${key}:`, keyWidth))}  ${value}`;
}
function renderPlainCardLine(ui, text) {
	return `${renderCardRail(ui)}  ${text}`;
}
function renderCardDivider(ui) {
	return renderCardRail(ui);
}
function readMoreWidth(descriptor) {
	return descriptor.docsPath ? [stringWidth("Read more")] : [];
}
function pushReadMore(lines, ui, keyWidth, descriptor) {
	if (!descriptor.docsPath) return;
	lines.push(renderCardDivider(ui));
	lines.push(`${renderCardRail(ui)}  ${ui.accent(padDisplay("Read more", keyWidth))}  ${ui.link(descriptor.docsPath)}`);
}
function renderCardRail(ui) {
	return ui.dim("│");
}
function formatListItemValue(ui, item) {
	const annotation = renderAnnotation(ui, item.status);
	return annotation ? `${item.label} ${annotation}` : item.label;
}
function renderAnnotation(ui, status) {
	if (status === "active") return ui.success("(active)");
	if (status === "default") return ui.dim("(default)");
	return "";
}
function formatValue(ui, value, tone = "default", sensitive = false) {
	const resolvedValue = sensitive ? maskValue(value) : value;
	if (tone === "success") return ui.success(resolvedValue);
	if (tone === "warning") return ui.warning(resolvedValue);
	if (tone === "error") return ui.error(resolvedValue);
	if (tone === "link") return ui.link(resolvedValue);
	if (tone === "dim") return ui.dim(resolvedValue);
	return resolvedValue;
}
//#endregion
export { renderList, renderMutate, renderShow, serializeList };
