// @prisma/compute-sdk/dist/build-strategy.js
import { readdir, readFile, stat } from "node:fs/promises";
import path from "node:path";
import { buildCommandEnv, runChildProcess } from "./workspace.js";
/**
 * Reads directly from a pre-built application directory without copying.
 * Validates the entrypoint exists.
 */
export class PreBuilt {
    #appPath;
    #entrypoint;
    constructor(options) {
        this.#appPath = options.appPath;
        this.#entrypoint = options.entrypoint;
    }
    async canBuild(_signal) {
        return true;
    }
    async execute(signal) {
        signal?.throwIfAborted();
        const normalized = path.normalize(this.#entrypoint);
        if (path.isAbsolute(normalized)) {
            throw new Error("Entrypoint must be a relative path");
        }
        if (normalized.startsWith("..")) {
            throw new Error("Entrypoint must not escape the application directory");
        }
        const fullPath = path.join(this.#appPath, normalized);
        const stats = await stat(fullPath).catch(() => null);
        if (!stats?.isFile()) {
            throw new Error(`Entrypoint not found in pre-built artifact: ${this.#entrypoint}`);
        }
        return {
            directory: this.#appPath,
            entrypoint: normalized.split(path.sep).join("/"),
        };
    }
}
// ---------------------------------------------------------------------------
// Shared helpers used by framework strategies (Next, Nuxt, Astro, TanStack
// Start). Live here because they're tightly tied to the BuildStrategy surface
// (framework detection + CLI dispatch) and don't justify their own module.
// ---------------------------------------------------------------------------
/**
 * True if any of `filenames` exists at the root of `appPath`. Swallows
 * readdir errors (missing/unreadable dir) and returns false.
 */
export async function hasRootFile(appPath, filenames, signal) {
    signal?.throwIfAborted();
    let entries;
    try {
        entries = await readdir(appPath);
    }
    catch {
        return false;
    }
    return entries.some((entry) => filenames.includes(entry));
}
/**
 * True if any of `packageNames` is listed in `package.json`'s `dependencies`
 * or `devDependencies`. Swallows missing-file / invalid-JSON errors and
 * returns false.
 */
export async function hasPackageDependency(appPath, packageNames, signal) {
    signal?.throwIfAborted();
    let content;
    try {
        content = await readFile(path.join(appPath, "package.json"), "utf-8");
    }
    catch {
        return false;
    }
    let parsed;
    try {
        parsed = JSON.parse(content);
    }
    catch {
        return false;
    }
    const deps = isRecord(parsed.dependencies) ? parsed.dependencies : {};
    const devDeps = isRecord(parsed.devDependencies)
        ? parsed.devDependencies
        : {};
    return packageNames.some((name) => name in deps || name in devDeps);
}
/**
 * Run a package's CLI inside `appPath`. Tries the project-local bin first,
 * then `npx <cliName>`, then `bunx <cliName>`. Each candidate that fails
 * with ENOENT is skipped; other failures (e.g., build errors) re-throw
 * with stderr surfaced.
 *
 * Throws with `missingMessage` if none of the launchers is available.
 */
export async function runPackageCli(opts) {
    // Every ancestor `node_modules/.bin` on PATH plus any injected deploy vars,
    // so a workspace-hoisted CLI resolves and the build sees its deploy env.
    const env = await buildCommandEnv(opts.appPath, { ...process.env, ...opts.env }, opts.signal);
    const localBin = path.join(opts.appPath, "node_modules", ".bin", opts.cliName);
    const candidates = [
        { command: localBin, args: opts.args },
        { command: "npx", args: [opts.cliName, ...opts.args] },
        { command: "bunx", args: [opts.cliName, ...opts.args] },
    ];
    for (const { command, args } of candidates) {
        opts.signal?.throwIfAborted();
        try {
            await runChildProcess({
                command,
                args,
                cwd: opts.appPath,
                env,
                failurePrefix: opts.failurePrefix,
                onOutput: opts.onOutput,
                signal: opts.signal,
            });
            return;
        }
        catch (error) {
            if (isENOENT(error))
                continue;
            throw error;
        }
    }
    throw new Error(opts.missingMessage);
}
function isRecord(value) {
    return typeof value === "object" && value !== null;
}
function isENOENT(error) {
    return (error instanceof Error &&
        "code" in error &&
        error.code === "ENOENT");
}
