// @prisma/compute-sdk/dist/build.js
import { AstroBuild } from "./astro-build.js";
import { BunBuild } from "./bun-build.js";
import { FRAMEWORKS } from "./config/frameworks.js";
import { NestjsBuild } from "./nestjs-build.js";
import { NextjsBuild } from "./nextjs-build.js";
import { NuxtBuild } from "./nuxt-build.js";
import { TanstackStartBuild } from "./tanstack-start-build.js";
/**
 * Detection order, derived from the framework registry so a new framework is
 * picked up automatically. `bun` is forced last — it is the universal fallback
 * (its `canBuild` always returns true).
 */
const DETECTION_ORDER = (() => {
    const buildTypes = [
        ...new Set(FRAMEWORKS.map((framework) => framework.buildType)),
    ];
    return [...buildTypes.filter((type) => type !== "bun"), "bun"];
})();
/**
 * Constructs the build strategy for an explicit framework build type, wiring
 * through optional committed build settings and (for bun) the entrypoint.
 */
export function createBuildStrategy(options) {
    const { appPath, buildSettings, io } = options;
    switch (options.buildType) {
        case "nextjs":
            return new NextjsBuild({
                appPath,
                buildSettings,
                io,
                requireStandalone: options.requireStandalone,
            });
        case "nuxt":
            return new NuxtBuild({ appPath, io });
        case "astro":
            return new AstroBuild({ appPath, io });
        case "nestjs":
            return new NestjsBuild({ appPath, buildSettings, io });
        case "tanstack-start":
            return new TanstackStartBuild({ appPath, buildSettings, io });
        case "bun":
            return new BunBuild({
                appPath,
                entrypoint: options.entrypoint,
                buildSettings,
                io,
            });
    }
}
/**
 * Resolves the build strategy and the concrete build type for an app. With an
 * explicit `buildType` the matching strategy is returned directly; with
 * `"auto"` (or omitted) each framework is tried in detection order and the
 * first match wins, with bun as the fallback. Returning the resolved build
 * type lets a consumer report what it detected.
 */
export async function resolveBuildStrategy(options) {
    const shared = {
        appPath: options.appPath,
        entrypoint: options.entrypoint,
        buildSettings: options.buildSettings,
        io: options.io,
        requireStandalone: options.requireStandalone,
    };
    if (options.buildType && options.buildType !== "auto") {
        return {
            buildType: options.buildType,
            strategy: createBuildStrategy({
                ...shared,
                buildType: options.buildType,
            }),
        };
    }
    for (const buildType of DETECTION_ORDER) {
        if (buildType === "bun") {
            break;
        }
        const strategy = createBuildStrategy({ ...shared, buildType });
        options.signal?.throwIfAborted();
        if (await strategy.canBuild(options.signal)) {
            return { buildType, strategy };
        }
    }
    return {
        buildType: "bun",
        strategy: createBuildStrategy({ ...shared, buildType: "bun" }),
    };
}
