// @prisma/compute-sdk/dist/build-settings.js
import { readFile } from "node:fs/promises";
import path from "node:path";
import { parseModule } from "magicast";
import { readBuildScript, readPackageManifest, resolvePackageManager, } from "./workspace.js";
/**
 * Resolves build settings from framework inference: the user's
 * `package.json` build script (run via the detected package manager) over the
 * framework default, plus the framework's output directory (reading
 * `next.config` `distDir` for Next.js).
 */
export async function resolveBuildSettings(options) {
    switch (options.buildType) {
        // Nuxt and Astro run their framework CLI and stage fixed output; these
        // settings only describe that for display.
        case "nuxt":
            return {
                buildCommand: "nuxt build",
                buildCommandSource: "Nuxt default",
                outputDirectory: ".output",
                outputDirectorySource: "Nuxt output",
            };
        case "astro":
            return {
                buildCommand: "astro build",
                buildCommandSource: "Astro default",
                outputDirectory: "dist",
                outputDirectorySource: "Astro output",
            };
        case "nextjs": {
            const manifest = await readPackageManifest(options.appPath, options.signal);
            const buildCommand = await resolveFrameworkBuildCommand(options.appPath, manifest, { command: "next build", source: "Next.js default" }, options.signal);
            const outputRoot = await resolveNextOutputRoot(options.appPath, options.signal);
            return {
                buildCommand: buildCommand.command,
                buildCommandSource: buildCommand.source,
                outputDirectory: joinPosix(outputRoot, "standalone"),
                outputDirectorySource: outputRoot === ".next" ? "Next.js output" : "next.config distDir",
            };
        }
        case "nestjs": {
            const manifest = await readPackageManifest(options.appPath, options.signal);
            const buildCommand = await resolveFrameworkBuildCommand(options.appPath, manifest, { command: "nest build", source: "NestJS default" }, options.signal);
            return {
                buildCommand: buildCommand.command,
                buildCommandSource: buildCommand.source,
                outputDirectory: "dist",
                outputDirectorySource: "NestJS output",
            };
        }
        case "tanstack-start": {
            const manifest = await readPackageManifest(options.appPath, options.signal);
            const buildCommand = await resolveFrameworkBuildCommand(options.appPath, manifest, { command: "vite build", source: "TanStack Start default" }, options.signal);
            return {
                buildCommand: buildCommand.command,
                buildCommandSource: buildCommand.source,
                outputDirectory: ".output",
                outputDirectorySource: "TanStack Start output",
            };
        }
        case "bun": {
            const manifest = await readPackageManifest(options.appPath, options.signal);
            const buildCommand = await resolveFrameworkBuildCommand(options.appPath, manifest, { command: null, source: null }, options.signal);
            return {
                buildCommand: buildCommand.command,
                buildCommandSource: buildCommand.source,
                outputDirectory: ".",
                outputDirectorySource: "app root",
            };
        }
    }
}
/**
 * Build settings when committed config owns them: configured fields win,
 * omitted fields fall back to framework inference. `source` labels configured
 * values with the config filename so a consumer can show provenance.
 */
export async function resolveConfiguredBuildSettings(options) {
    const needsFallback = options.configured.command === undefined ||
        options.configured.outputDirectory === undefined;
    const fallback = needsFallback ? await resolveBuildSettings(options) : null;
    return {
        buildCommand: options.configured.command !== undefined
            ? options.configured.command
            : fallback.buildCommand,
        buildCommandSource: options.configured.command !== undefined
            ? options.source
            : fallback.buildCommandSource,
        outputDirectory: options.configured.outputDirectory ??
            fallback.outputDirectory,
        outputDirectorySource: options.configured.outputDirectory !== undefined
            ? options.source
            : fallback.outputDirectorySource,
    };
}
async function resolveFrameworkBuildCommand(appPath, manifest, fallback, signal) {
    const buildScript = readBuildScript(manifest);
    if (buildScript) {
        const packageManager = await resolvePackageManager(appPath, signal);
        return {
            command: packageManager ? `${packageManager} run build` : buildScript,
            source: "package.json scripts.build",
        };
    }
    return fallback;
}
/** Joins posix-style path segments, collapsing duplicate separators. */
export function joinPosix(...parts) {
    return parts.join("/").replace(/\/+/g, "/");
}
/**
 * Maps a Next.js standalone output directory (e.g. ".next/standalone") back to
 * its output root (".next"), so static assets can be staged alongside it.
 */
export function nextOutputRootFromStandaloneDirectory(outputDirectory) {
    const normalized = outputDirectory.replace(/\/+$/g, "");
    if (normalized === "standalone") {
        return ".";
    }
    if (normalized.endsWith("/standalone")) {
        const outputRoot = normalized.slice(0, -"/standalone".length);
        return outputRoot.length > 0 ? outputRoot : ".";
    }
    const dirname = path.posix.dirname(normalized);
    return dirname === "." ? "." : dirname;
}
const NEXT_CONFIG_FILENAMES = [
    "next.config.js",
    "next.config.mjs",
    "next.config.ts",
    "next.config.mts",
];
async function resolveNextOutputRoot(appPath, signal) {
    const config = await readNextConfig(appPath, signal);
    return config.distDir ?? ".next";
}
async function readNextConfig(appPath, signal) {
    for (const fileName of NEXT_CONFIG_FILENAMES) {
        signal?.throwIfAborted();
        let content;
        try {
            content = await readFile(path.join(appPath, fileName), {
                encoding: "utf8",
                signal,
            });
        }
        catch (error) {
            if (error.code === "ENOENT") {
                continue;
            }
            throw error;
        }
        return readStaticNextConfig(content);
    }
    return {};
}
/** Best-effort static read of `distDir`/`output` from a next.config source. */
export function readStaticNextConfig(content) {
    try {
        const module = parseModule(content);
        const program = asAstNode(module.$ast);
        const bindings = program
            ? collectStaticBindings(program)
            : new Map();
        const configObject = program
            ? findExportedConfigObject(program, bindings)
            : null;
        if (!configObject) {
            return {};
        }
        const rawDistDir = readStaticStringProperty(configObject, "distDir");
        const output = readStaticStringProperty(configObject, "output");
        return {
            distDir: rawDistDir ? normalizeRelativePath(rawDistDir) : undefined,
            output: output === "standalone" || output === "export" ? output : undefined,
        };
    }
    catch {
        return {};
    }
}
function normalizeRelativePath(value) {
    const raw = value.trim().replace(/\\/g, "/");
    if (raw.length === 0 || raw.split("/").includes("..")) {
        return undefined;
    }
    if (/^[A-Za-z]:/.test(raw)) {
        return undefined;
    }
    const normalized = path.posix.normalize(raw);
    if (path.win32.isAbsolute(value) ||
        path.posix.isAbsolute(normalized) ||
        normalized.split("/").includes("..")) {
        return undefined;
    }
    return normalized === "." ? "." : normalized;
}
function asAstNode(value) {
    if (!value || typeof value !== "object") {
        return null;
    }
    const type = value.type;
    return typeof type === "string" ? value : null;
}
function astNodes(value) {
    if (!Array.isArray(value)) {
        return [];
    }
    return value.map(asAstNode).filter((node) => Boolean(node));
}
function collectStaticBindings(program) {
    const bindings = new Map();
    for (const statement of astNodes(program.body)) {
        if (statement.type !== "VariableDeclaration") {
            continue;
        }
        for (const declaration of astNodes(statement.declarations)) {
            const id = asAstNode(declaration.id);
            const init = asAstNode(declaration.init);
            if (id?.type === "Identifier" && typeof id.name === "string" && init) {
                bindings.set(id.name, init);
            }
        }
    }
    return bindings;
}
function findExportedConfigObject(program, bindings) {
    for (const statement of astNodes(program.body)) {
        if (statement.type === "ExportDefaultDeclaration") {
            return resolveConfigObject(statement.declaration, bindings);
        }
        if (statement.type !== "ExpressionStatement") {
            continue;
        }
        const expression = asAstNode(statement.expression);
        if (expression?.type !== "AssignmentExpression" ||
            expression.operator !== "=") {
            continue;
        }
        if (isModuleExports(expression.left)) {
            return resolveConfigObject(expression.right, bindings);
        }
    }
    return null;
}
function resolveConfigObject(value, bindings, depth = 0) {
    if (depth > 4) {
        return null;
    }
    const node = unwrapStaticExpression(asAstNode(value));
    if (!node) {
        return null;
    }
    if (node.type === "ObjectExpression") {
        return node;
    }
    if (node.type === "Identifier" && typeof node.name === "string") {
        return resolveConfigObject(bindings.get(node.name), bindings, depth + 1);
    }
    if (node.type === "CallExpression") {
        return resolveConfigObject(astNodes(node.arguments)[0], bindings, depth + 1);
    }
    return null;
}
function unwrapStaticExpression(node) {
    let current = node;
    while (current?.type === "TSAsExpression" ||
        current?.type === "TSSatisfiesExpression" ||
        current?.type === "TSNonNullExpression") {
        current = asAstNode(current.expression);
    }
    return current;
}
function isModuleExports(value) {
    const node = asAstNode(value);
    if (node?.type !== "MemberExpression" || node.computed === true) {
        return false;
    }
    const object = asAstNode(node.object);
    const property = asAstNode(node.property);
    return (object?.type === "Identifier" &&
        object.name === "module" &&
        property?.type === "Identifier" &&
        property.name === "exports");
}
function readStaticStringProperty(objectExpression, propertyName) {
    for (const property of astNodes(objectExpression.properties)) {
        if (property.type !== "ObjectProperty" || property.computed === true) {
            continue;
        }
        if (propertyKeyName(property.key) !== propertyName) {
            continue;
        }
        const value = unwrapStaticExpression(asAstNode(property.value));
        if (value?.type === "StringLiteral" && typeof value.value === "string") {
            const trimmed = value.value.trim();
            return trimmed.length > 0 ? trimmed : undefined;
        }
    }
    return undefined;
}
function propertyKeyName(value) {
    const key = asAstNode(value);
    if (key?.type === "Identifier" && typeof key.name === "string") {
        return key.name;
    }
    if (key?.type === "StringLiteral" && typeof key.value === "string") {
        return key.value;
    }
    return undefined;
}
