// @prisma/compute-sdk/dist/errors.js
import { TaggedError } from "better-result";
export class AuthenticationError extends TaggedError("AuthenticationError")() {
}
export class ApiError extends TaggedError("ApiError")() {
}
export class MissingArgumentError extends TaggedError("MissingArgumentError")() {
    constructor(args) {
        super({
            field: args.field,
            message: `Missing required argument: ${args.field}`,
        });
    }
}
export class BuildError extends TaggedError("BuildError")() {
}
export class ArtifactError extends TaggedError("ArtifactError")() {
}
export class TimeoutError extends TaggedError("TimeoutError")() {
    constructor(args) {
        super({
            ...args,
            message: `Timed out waiting for version ${args.versionId} to reach target state (last status: ${args.lastStatus}, elapsed: ${Math.round(args.elapsedMs / 1000)}s)`,
        });
    }
}
export class VersionFailedError extends TaggedError("VersionFailedError")() {
    constructor(args) {
        super({
            versionId: args.versionId,
            message: `Version ${args.versionId} transitioned to failed status`,
        });
    }
}
export class NoExistingVersionError extends TaggedError("NoExistingVersionError")() {
    constructor(args) {
        super({
            serviceId: args.serviceId,
            message: `Service ${args.serviceId} has no existing versions. Environment-only updates require at least one prior deployment.`,
        });
    }
}
export class NoVersionsFoundError extends TaggedError("NoVersionsFoundError")() {
    constructor(args) {
        super({
            serviceId: args.serviceId,
            message: `Service ${args.serviceId} has no versions`,
        });
    }
}
export class CancelledError extends TaggedError("CancelledError")() {
    constructor(args) {
        super({ message: args?.message ?? "Operation cancelled" });
    }
}
export class DestroyAggregateError extends TaggedError("DestroyAggregateError")() {
    constructor(args) {
        const failedIds = args.failures.map((f) => f.versionId).join(", ");
        super({
            ...args,
            message: `Failed to destroy ${args.failures.length} version(s): ${failedIds}`,
        });
    }
}
export class InvalidOptionsError extends TaggedError("InvalidOptionsError")() {
}
export class LogStreamError extends TaggedError("LogStreamError")() {
}
