// @prisma/compute-sdk/dist/index.js
// Re-export Result utilities so consumers don't need a direct better-result dependency
export { Err, matchError, matchErrorPartial, Ok, TaggedError, } from "better-result";
export { applyMigrations } from "./apply-migrations.js";
export { createArchive } from "./archive.js";
export { normalizeArtifactSymlinks } from "./artifact-postprocess.js";
export { hoistIsolatedStoreDependencies, stageStandaloneArtifact, } from "./artifact-stage.js";
export { AstroBuild } from "./astro-build.js";
export { AutoBuild } from "./auto-build.js";
export { createBuildStrategy, resolveBuildStrategy, } from "./build.js";
export { joinPosix, nextOutputRootFromStandaloneDirectory, readStaticNextConfig, resolveBuildSettings, resolveConfiguredBuildSettings, } from "./build-settings.js";
export { PreBuilt } from "./build-strategy.js";
export { BunBuild } from "./bun-build.js";
export { ComputeClient } from "./compute-client.js";
export { detectAppSchema } from "./detect-schema.js";
export { ApiError, ArtifactError, AuthenticationError, BuildError, CancelledError, DestroyAggregateError, InvalidOptionsError, LogStreamError, MissingArgumentError, NoExistingVersionError, NoVersionsFoundError, TimeoutError, VersionFailedError, } from "./errors.js";
export { streamLogs } from "./log-stream.js";
export { NestjsBuild } from "./nestjs-build.js";
export { NextjsBuild } from "./nextjs-build.js";
export { NuxtBuild } from "./nuxt-build.js";
export { TanstackStartBuild } from "./tanstack-start-build.js";
export { KNOWN_REGION_IDS, REGIONS } from "./types.js";
export { uploadArtifact } from "./upload-artifact.js";
export { buildCommandEnv, readBuildScript, readPackageManifest, resolvePackageManager, runBuildCommand, } from "./workspace.js";
