// @prisma/compute-sdk/dist/log-stream.js
import { Result } from "better-result";
import WebSocket from "ws";
import { CancelledError, LogStreamError } from "./errors.js";
const isBun = "Bun" in globalThis;
function parseErrorBody(body, statusCode) {
    try {
        const parsed = JSON.parse(body);
        return parsed.error?.message ?? parsed.message ?? `HTTP ${statusCode}`;
    }
    catch {
        return body || `HTTP ${statusCode}`;
    }
}
function toError(value) {
    if (value instanceof Error)
        return value;
    // Bun emits ErrorEvent instead of Error on WebSocket failures
    if (typeof value === "object" &&
        value !== null &&
        "message" in value &&
        typeof value.message === "string") {
        return new Error(value.message);
    }
    return new Error(String(value));
}
function buildWebSocketUrl(options) {
    const httpUrl = new URL(`/v1/compute-services/versions/${encodeURIComponent(options.versionId)}/logs`, options.baseUrl);
    httpUrl.protocol = httpUrl.protocol === "https:" ? "wss:" : "ws:";
    if (options.tail != null)
        httpUrl.searchParams.set("tail", String(options.tail));
    if (options.fromStart)
        httpUrl.searchParams.set("from_start", "true");
    if (options.cursor)
        httpUrl.searchParams.set("cursor", options.cursor);
    return httpUrl.toString();
}
function connectOnce(options, onRecord) {
    return new Promise((resolve, reject) => {
        if (options.signal?.aborted) {
            reject(new CancelledError());
            return;
        }
        const url = buildWebSocketUrl(options);
        const ws = new WebSocket(url, {
            headers: { Authorization: `Bearer ${options.token}` },
        });
        let terminal = null;
        let lastByteEnd = null;
        let settled = false;
        const cleanup = () => {
            options.signal?.removeEventListener("abort", onAbort);
        };
        const onAbort = () => {
            cleanup();
            if (!settled) {
                settled = true;
                ws.close();
                reject(new CancelledError());
            }
        };
        options.signal?.addEventListener("abort", onAbort, { once: true });
        if (options.signal?.aborted) {
            onAbort();
            return;
        }
        // Node.js ws: fires when the server responds with a non-101 status,
        // giving us access to the HTTP response for structured error messages.
        // Bun doesn't implement this event — errors go to the "error" handler.
        if (!isBun) {
            ws.on("unexpected-response", (_req, res) => {
                const chunks = [];
                res.on("data", (chunk) => {
                    chunks.push(chunk);
                });
                res.on("end", () => {
                    cleanup();
                    if (!settled) {
                        settled = true;
                        const statusCode = res.statusCode ?? 0;
                        const body = Buffer.concat(chunks).toString();
                        reject(new LogStreamError({
                            statusCode,
                            message: parseErrorBody(body, statusCode),
                        }));
                    }
                });
            });
        }
        ws.on("message", (data) => {
            let record;
            try {
                record = JSON.parse(data.toString());
            }
            catch {
                return;
            }
            if (record.type === "log") {
                lastByteEnd = record.byteEnd;
            }
            else if (record.type === "terminal") {
                terminal = record;
            }
            try {
                onRecord(record);
            }
            catch (err) {
                cleanup();
                if (!settled) {
                    settled = true;
                    ws.close();
                    reject(toError(err));
                }
            }
        });
        ws.on("close", () => {
            cleanup();
            if (!settled) {
                settled = true;
                resolve({ terminal, lastByteEnd });
            }
        });
        ws.on("error", (err) => {
            cleanup();
            if (!settled) {
                settled = true;
                const error = toError(err);
                reject(new LogStreamError({ statusCode: 0, message: error.message }));
            }
        });
    });
}
/**
 * Open a WebSocket to the log-streaming endpoint and deliver records via callback.
 *
 * Returns `Ok` with the stream result on graceful close, or `Err` with a
 * `LogStreamError` (connection/HTTP errors) or `CancelledError` (abort signal).
 */
export async function streamLogs(options, onRecord) {
    return Result.tryPromise({
        try: () => connectOnce(options, onRecord),
        catch: (e) => {
            if (CancelledError.is(e))
                return e;
            if (LogStreamError.is(e))
                return e;
            const error = toError(e);
            return new LogStreamError({ statusCode: 0, message: error.message });
        },
    });
}
