// @prisma/compute-sdk/dist/workspace.js
import { spawn } from "node:child_process";
import { readFile, stat } from "node:fs/promises";
import path from "node:path";
import { sourceRootLineage } from "./config/source-root.js";
/**
 * Detects the package manager governing `appPath` by walking from the app up
 * to the source root: the nearest `packageManager` field wins, then the
 * nearest lockfile. Workspace repos keep both at the root, so a package
 * deep in a monorepo still resolves correctly. Returns undefined when no
 * signal is found.
 */
export async function resolvePackageManager(appPath, signal) {
    for (const directory of await sourceRootLineage(appPath, signal)) {
        const manifest = await readPackageManifest(directory, signal);
        const fromField = packageManagerFromField(manifest?.packageManager);
        if (fromField) {
            return fromField;
        }
        if ((await pathExists(path.join(directory, "bun.lock"), signal)) ||
            (await pathExists(path.join(directory, "bun.lockb"), signal))) {
            return "bun";
        }
        if (await pathExists(path.join(directory, "pnpm-lock.yaml"), signal)) {
            return "pnpm";
        }
        if (await pathExists(path.join(directory, "yarn.lock"), signal)) {
            return "yarn";
        }
        if (await pathExists(path.join(directory, "package-lock.json"), signal)) {
            return "npm";
        }
    }
    return undefined;
}
/**
 * Build environment for `appPath`: every `node_modules/.bin` between the app
 * and its source root, prepended to PATH. Workspace repos hoist binaries like
 * `next` to the workspace root, so a build run from a package directory still
 * finds them. The returned object is suitable as the `env` for a child build
 * process.
 */
export async function buildCommandEnv(appPath, baseEnv, signal) {
    const binDirs = (await sourceRootLineage(appPath, signal)).map((directory) => path.join(directory, "node_modules", ".bin"));
    return {
        ...baseEnv,
        PATH: [...binDirs, baseEnv.PATH].filter(Boolean).join(path.delimiter),
    };
}
/**
 * Runs a build command string inside `appPath`, with every ancestor
 * `node_modules/.bin` on PATH so workspace-hoisted binaries resolve. Rejects
 * with the framework prefix and captured output on a non-zero exit.
 */
export async function runBuildCommand(options) {
    const env = await buildCommandEnv(options.appPath, { ...process.env, ...options.env }, options.signal);
    // `shell: true` preserves the platform shell so a command string like
    // `pnpm run build` works the same as the CLI's previous `exec`.
    await runChildProcess({
        command: options.command,
        args: [],
        cwd: options.appPath,
        env,
        shell: true,
        failurePrefix: options.failurePrefix,
        onOutput: options.onOutput,
        signal: options.signal,
    });
}
/**
 * Spawns a child build process, streaming each output line to `onOutput` and
 * keeping a bounded tail for the failure message instead of buffering the whole
 * build. Resolves on a zero exit; rejects on a non-zero exit (with the prefix,
 * exit code, and tail) or a spawn error (whose `code`, e.g. ENOENT, is
 * preserved on the rejection so a launcher ladder can branch on it).
 *
 * Shared by `runBuildCommand` (a shell command string) and `runPackageCli`
 * (an explicit binary + args).
 */
export function runChildProcess(options) {
    return new Promise((resolve, reject) => {
        const tail = createBoundedTail();
        const child = spawn(options.command, [...options.args], {
            cwd: options.cwd,
            env: options.env,
            signal: options.signal,
            shell: options.shell ?? false,
        });
        const streamed = Promise.all([
            streamLines(child.stdout, "stdout", tail, options.onOutput),
            streamLines(child.stderr, "stderr", tail, options.onOutput),
        ]);
        child.on("error", (error) => {
            // Preserve `code` (e.g. ENOENT) so a launcher ladder can fall through.
            reject(Object.assign(new Error(`${options.failurePrefix} build failed:\n${error.message}`), { code: error.code }));
        });
        child.on("close", (code) => {
            streamed.then(() => {
                if (code === 0) {
                    resolve();
                    return;
                }
                const detail = tail.text().trim();
                const exit = code != null ? ` (exit code ${code})` : "";
                reject(new Error(`${options.failurePrefix} build failed${exit}${detail ? `:\n${detail}` : "."}`));
            }, reject);
        });
    });
}
/** Splits a child stream into lines, taps each, and records it in the tail. */
function streamLines(stream, source, tail, onOutput) {
    return new Promise((resolve) => {
        if (!stream) {
            resolve();
            return;
        }
        stream.setEncoding("utf8");
        let buffer = "";
        const emit = (line) => {
            tail.push(line);
            onOutput?.(line, source);
        };
        stream.on("data", (chunk) => {
            buffer += chunk;
            let newline = buffer.indexOf("\n");
            while (newline !== -1) {
                emit(buffer.slice(0, newline));
                buffer = buffer.slice(newline + 1);
                newline = buffer.indexOf("\n");
            }
        });
        stream.on("end", () => {
            if (buffer.length > 0) {
                emit(buffer);
            }
            resolve();
        });
        // A stream error surfaces through the child's own error/close handlers.
        stream.on("error", () => resolve());
    });
}
/** A ring of the most recent output lines, for the failure message. */
function createBoundedTail(maxLines = 200) {
    const lines = [];
    return {
        push(line) {
            lines.push(line);
            if (lines.length > maxLines) {
                lines.shift();
            }
        },
        text() {
            return lines.join("\n");
        },
    };
}
/** Reads and parses `appPath/package.json`, returning null on any failure. */
export async function readPackageManifest(appPath, signal) {
    signal?.throwIfAborted();
    try {
        const content = await readFile(path.join(appPath, "package.json"), {
            encoding: "utf-8",
            signal,
        });
        const parsed = JSON.parse(content);
        return isRecord(parsed) ? parsed : null;
    }
    catch (error) {
        // A missing file or invalid JSON is expected; an abort must propagate.
        if (signal?.aborted)
            throw error;
        return null;
    }
}
/** The trimmed `scripts.build` string, or null when absent or empty. */
export function readBuildScript(manifest) {
    const build = manifest?.scripts?.build;
    if (typeof build !== "string") {
        return null;
    }
    const trimmed = build.trim();
    return trimmed.length > 0 ? trimmed : null;
}
function packageManagerFromField(value) {
    if (typeof value !== "string") {
        return null;
    }
    const name = value.split("@")[0];
    return name === "bun" || name === "pnpm" || name === "yarn" || name === "npm"
        ? name
        : null;
}
async function pathExists(targetPath, signal) {
    signal?.throwIfAborted();
    try {
        await stat(targetPath);
        return true;
    }
    catch {
        return false;
    }
}
function isRecord(value) {
    return typeof value === "object" && value !== null;
}
