// @prisma/management-api-sdk/dist/index.js
import createOpenAPIClient from "openapi-fetch";

//#region src/errors.ts
/**
* Error thrown when a network request fails.
*/
var FetchError = class extends Error {
	constructor(message, cause) {
		super(message);
		this.cause = cause;
		this.name = "FetchError";
	}
};
/**
* Error thrown when authentication fails.
*/
var AuthError = class extends Error {
	constructor(message, refreshTokenInvalid = false) {
		super(message);
		this.refreshTokenInvalid = refreshTokenInvalid;
		this.name = "AuthError";
	}
};

//#endregion
//#region src/auth.ts
function decodeJwtPayload(token) {
	const parts = token.split(".");
	const payload = parts[1];
	if (parts.length !== 3 || payload === void 0) throw new Error("Invalid JWT format");
	const base64 = payload.replace(/-/g, "+").replace(/_/g, "/");
	const padded = base64 + "=".repeat((4 - base64.length % 4) % 4);
	const decoded = atob(padded);
	return JSON.parse(decoded);
}
function extractWorkspaceId(accessToken) {
	const workspaceId = decodeJwtPayload(accessToken)["workspace_id"];
	if (typeof workspaceId !== "string") throw new Error("workspace_id not found in access token");
	return workspaceId;
}
function base64urlEncode(buffer) {
	const bytes = Array.from(buffer);
	return btoa(String.fromCharCode(...bytes)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
function generateRandomString(length) {
	const array = new Uint8Array(length);
	crypto.getRandomValues(array);
	return base64urlEncode(array);
}
async function generateCodeChallenge(verifier) {
	const data = new TextEncoder().encode(verifier);
	const hash = await crypto.subtle.digest("SHA-256", data);
	return base64urlEncode(new Uint8Array(hash));
}
function tokenUrl(config) {
	return `${config.authBaseUrl}/token`;
}
function formatOAuthError(data) {
	if (typeof data === "object" && data !== null) {
		const errorData = data;
		const error = errorData["error"];
		const description = errorData["error_description"];
		if (typeof error === "string") return typeof description === "string" ? `${error}: ${description}` : error;
	}
	return "Unknown error";
}
async function parseTokenResponse(response) {
	let data;
	try {
		data = await response.json();
	} catch {
		return {
			status: "error",
			error: `Failed to parse token response (status ${response.status})`,
			refreshTokenInvalid: false
		};
	}
	if (response.status >= 400 && response.status < 500) return {
		status: "error",
		error: formatOAuthError(data),
		refreshTokenInvalid: typeof data === "object" && data !== null && data["error"] === "invalid_grant"
	};
	if (response.status !== 200) return {
		status: "error",
		error: `Token request failed with status ${response.status}`,
		refreshTokenInvalid: false
	};
	if (typeof data !== "object" || !data || typeof data["access_token"] !== "string" || typeof data["refresh_token"] !== "string") return {
		status: "error",
		error: "Invalid token response format",
		refreshTokenInvalid: false
	};
	const tokenData = data;
	return {
		status: "success",
		accessToken: tokenData.access_token,
		refreshToken: tokenData.refresh_token
	};
}
async function refreshAccessToken(config, refreshToken) {
	const body = new URLSearchParams({
		grant_type: "refresh_token",
		refresh_token: refreshToken,
		client_id: config.clientId
	});
	let response;
	try {
		response = await fetch(tokenUrl(config), {
			method: "POST",
			headers: { "Content-Type": "application/x-www-form-urlencoded" },
			body
		});
	} catch {
		return {
			status: "error",
			error: "Network error during token refresh",
			refreshTokenInvalid: false
		};
	}
	return parseTokenResponse(response);
}
function tokensEqual(a, b) {
	return a?.workspaceId === b?.workspaceId && a?.accessToken === b?.accessToken && a?.refreshToken === b?.refreshToken;
}
function createTokenRefreshingFetch(config) {
	let refreshPromise = null;
	const doRefresh = async (tokensThatFailed) => {
		return (config.tokenStorage.withRefreshLock?.bind(config.tokenStorage) ?? ((fn) => fn()))(async () => {
			const currentTokens = await config.tokenStorage.getTokens();
			if (!currentTokens) throw new AuthError("No tokens available. Please log in first.");
			if (!tokensEqual(currentTokens, tokensThatFailed)) return currentTokens;
			if (!currentTokens.refreshToken) throw new AuthError("No refresh token available. Please log in again.");
			const result = await refreshAccessToken(config, currentTokens.refreshToken);
			if (result.status === "error") {
				if (result.refreshTokenInvalid) if (config.tokenStorage.clearTokensIfCurrent) await config.tokenStorage.clearTokensIfCurrent(currentTokens);
				else await config.tokenStorage.clearTokens();
				throw new AuthError(result.error, result.refreshTokenInvalid);
			}
			const newTokens = {
				accessToken: result.accessToken,
				refreshToken: result.refreshToken,
				workspaceId: extractWorkspaceId(result.accessToken)
			};
			await config.tokenStorage.setTokens(newTokens);
			return newTokens;
		});
	};
	const refreshTokens = async (tokensThatFailed) => {
		if (refreshPromise) return refreshPromise;
		refreshPromise = doRefresh(tokensThatFailed).finally(() => {
			refreshPromise = null;
		});
		return refreshPromise;
	};
	return async function tokenRefreshingFetch(request) {
		request.signal.throwIfAborted();
		const tokens = await config.tokenStorage.getTokens();
		if (!tokens) throw new AuthError("No tokens available. Please log in first.");
		const requestCloneForRetry = request.clone();
		request.headers.set("Authorization", `Bearer ${tokens.accessToken}`);
		let response;
		try {
			response = await fetch(request);
		} catch (error) {
			request.signal.throwIfAborted();
			throw new FetchError("Network request failed", error);
		}
		if (response.status === 401) {
			request.signal.throwIfAborted();
			const newTokens = await refreshTokens(tokens);
			requestCloneForRetry.headers.set("Authorization", `Bearer ${newTokens.accessToken}`);
			try {
				response = await fetch(requestCloneForRetry);
			} catch (error) {
				request.signal.throwIfAborted();
				throw new FetchError("Network request failed after token refresh", error);
			}
		}
		return response;
	};
}

//#endregion
//#region src/index.ts
const DEFAULT_API_BASE_URL = "https://api.prisma.io";
const DEFAULT_AUTH_BASE_URL = "https://auth.prisma.io";
/**
* Creates a raw API client without authentication handling.
* Useful if you want to manage authentication yourself.
*/
function createManagementApiClient(options) {
	const clientOptions = {
		...options,
		baseUrl: options.baseUrl ?? DEFAULT_API_BASE_URL,
		headers: {
			...options.headers,
			Authorization: options.token ? `Bearer ${options.token}` : void 0
		}
	};
	delete clientOptions.token;
	return createOpenAPIClient(clientOptions);
}
/**
* Creates a Management API instance with OAuth authentication and automatic token refresh.
*/
function createManagementApiSdk(config) {
	const fullConfig = {
		...config,
		apiBaseUrl: config.apiBaseUrl ?? DEFAULT_API_BASE_URL,
		authBaseUrl: config.authBaseUrl ?? DEFAULT_AUTH_BASE_URL
	};
	const client = createManagementApiClient({
		baseUrl: fullConfig.apiBaseUrl,
		fetch: createTokenRefreshingFetch(fullConfig)
	});
	client.use({ onError({ error, request }) {
		request.signal.throwIfAborted();
		if (error instanceof FetchError || error instanceof AuthError) return error;
		return new FetchError("API request failed", error);
	} });
	async function getLoginUrl(options, asyncOptions) {
		asyncOptions?.signal?.throwIfAborted();
		const state = generateRandomString(16);
		const verifier = generateRandomString(32);
		const challenge = await generateCodeChallenge(verifier);
		asyncOptions?.signal?.throwIfAborted();
		const authUrl = new URL(`${fullConfig.authBaseUrl}/authorize`);
		authUrl.searchParams.set("response_type", "code");
		authUrl.searchParams.set("client_id", fullConfig.clientId);
		authUrl.searchParams.set("redirect_uri", fullConfig.redirectUri);
		authUrl.searchParams.set("scope", options.scope);
		authUrl.searchParams.set("state", state);
		authUrl.searchParams.set("code_challenge", challenge);
		authUrl.searchParams.set("code_challenge_method", "S256");
		if (options.additionalParams) for (const [key, value] of Object.entries(options.additionalParams)) authUrl.searchParams.set(key, value);
		return {
			url: authUrl.toString(),
			state,
			verifier
		};
	}
	async function handleCallback(options, asyncOptions) {
		let url;
		try {
			url = typeof options.callbackUrl === "string" ? new URL(options.callbackUrl) : options.callbackUrl;
		} catch {
			throw new AuthError("Invalid callback URL");
		}
		const params = url.searchParams;
		const error = params.get("error");
		if (error) {
			const errorDescription = params.get("error_description");
			throw new AuthError(errorDescription ? `${error}: ${errorDescription}` : error);
		}
		const code = params.get("code");
		const state = params.get("state");
		if (!code) throw new AuthError("No authorization code found in callback URL");
		if (state !== options.expectedState) throw new AuthError("Invalid state parameter. Possible CSRF attack.");
		asyncOptions?.signal?.throwIfAborted();
		const body = new URLSearchParams({
			grant_type: "authorization_code",
			code,
			redirect_uri: fullConfig.redirectUri,
			client_id: fullConfig.clientId,
			code_verifier: options.verifier
		});
		let response;
		try {
			response = await fetch(tokenUrl(fullConfig), {
				method: "POST",
				headers: { "Content-Type": "application/x-www-form-urlencoded" },
				body
			});
		} catch (error$1) {
			throw new FetchError("Network error during token exchange", error$1);
		}
		const result = await parseTokenResponse(response);
		if (result.status === "error") throw new AuthError(result.error);
		await fullConfig.tokenStorage.setTokens({
			accessToken: result.accessToken,
			refreshToken: result.refreshToken,
			workspaceId: extractWorkspaceId(result.accessToken)
		});
	}
	async function logout(asyncOptions) {
		asyncOptions?.signal?.throwIfAborted();
		await fullConfig.tokenStorage.clearTokens();
	}
	return {
		client,
		getLoginUrl,
		handleCallback,
		logout
	};
}

//#endregion
export { AuthError, FetchError, createManagementApiClient, createManagementApiSdk };
