// @prisma/cli/dist/use-cases/auth.js
import { authRequiredError, usageError, workspaceAmbiguousError, workspaceNotAuthenticatedError } from "../shell/errors.js";
//#region src/use-cases/auth.ts
function createAuthUseCases(dependencies) {
	return {
		whoami: () => resolveCurrentAuthState(dependencies),
		login: async (selection) => {
			await dependencies.sessionGateway.writeAuthSession({
				provider: selection.provider,
				userId: selection.userId,
				workspaceId: selection.workspaceId
			});
			return resolveCurrentAuthState(dependencies);
		},
		logout: async () => {
			await dependencies.sessionGateway.clearAuthSession();
			return resolveCurrentAuthState(dependencies);
		},
		listWorkspaces: async () => {
			const session = await dependencies.sessionGateway.readAuthSession();
			if (!session) return {
				authSource: "none",
				activeWorkspace: null,
				workspaces: []
			};
			const workspaces = dependencies.identityGateway.listUserWorkspaces(session.userId);
			return {
				authSource: "oauth",
				activeWorkspace: workspaces.find((workspace) => workspace.id === session.workspaceId) ?? null,
				workspaces: workspaces.map((workspace) => ({
					...workspace,
					credentialWorkspaceId: workspace.id,
					active: workspace.id === session.workspaceId,
					source: "oauth",
					switchable: true,
					lastSeenAt: null
				}))
			};
		},
		useWorkspace: async (workspaceRef) => {
			const session = await dependencies.sessionGateway.readAuthSession();
			if (!session) throw authRequiredError(["prisma-cli auth login"]);
			const ref = workspaceRef.trim();
			const matches = dependencies.identityGateway.listUserWorkspaces(session.userId).filter((workspace) => workspaceMatchesRef(workspace, ref));
			if (matches.length === 0) throw workspaceNotAuthenticatedError(workspaceRef);
			if (matches.length > 1) throw workspaceAmbiguousError(workspaceRef, matches.map((workspace) => ({
				id: workspace.id,
				name: workspace.name,
				credentialWorkspaceId: workspace.id
			})));
			const selected = matches[0];
			const previousWorkspace = dependencies.identityGateway.getWorkspace(session.workspaceId) ?? null;
			await dependencies.sessionGateway.writeAuthSession({
				...session,
				workspaceId: selected.id
			});
			return {
				previousWorkspace,
				workspace: selected
			};
		},
		logoutWorkspace: async (workspaceRef) => {
			const session = await dependencies.sessionGateway.readAuthSession();
			if (!session) throw workspaceNotAuthenticatedError(workspaceRef);
			const ref = workspaceRef.trim();
			const matches = dependencies.identityGateway.listUserWorkspaces(session.userId).filter((workspace) => workspaceMatchesRef(workspace, ref));
			if (matches.length === 0) throw workspaceNotAuthenticatedError(workspaceRef);
			if (matches.length > 1) throw workspaceAmbiguousError(workspaceRef, matches.map((workspace) => ({
				id: workspace.id,
				name: workspace.name,
				credentialWorkspaceId: workspace.id
			})));
			const workspace = matches[0];
			const wasActive = workspace.id === session.workspaceId;
			const activeWorkspace = wasActive ? null : dependencies.identityGateway.getWorkspace(session.workspaceId) ?? null;
			if (wasActive) await dependencies.sessionGateway.clearAuthSession();
			return {
				workspace,
				wasActive,
				activeWorkspace
			};
		},
		listProviders: async () => dependencies.identityGateway.listProviders(),
		resolveProvider: async (providerId) => {
			const provider = dependencies.identityGateway.getProvider(providerId);
			if (!provider) throw usageError("Login requires a valid mock provider", `The mock provider "${providerId}" does not exist.`, "Use --provider github or --provider google.", ["prisma-cli auth login"], "auth");
			return provider;
		},
		listUsersForProvider: async (providerId) => {
			const users = dependencies.identityGateway.listUsersForProvider(providerId);
			if (users.length === 0) throw usageError("Login requires a valid mock user", `No mock users support provider "${providerId}".`, "Update the fixture data or choose a different provider.", ["prisma-cli auth login"], "auth");
			return users;
		},
		resolveUserForProvider: async (providerId, userId) => {
			const user = dependencies.identityGateway.getUserForProvider(providerId, userId);
			if (!user) throw usageError("Login requires a valid mock user", `The mock user "${userId}" is not available for provider "${providerId}".`, "Choose a user that supports the selected provider.", ["prisma-cli auth login"], "auth");
			return user;
		},
		listWorkspacesForUser: async (userId) => dependencies.identityGateway.listUserWorkspaces(userId),
		resolveWorkspaceForUser: async (userId, workspaceId) => {
			const workspace = dependencies.identityGateway.getUserWorkspace(userId, workspaceId);
			if (!workspace) throw usageError("Login requires a valid mock workspace", `The mock workspace "${workspaceId}" is not available for the selected user.`, "Choose a workspace that the selected user can access.", ["prisma-cli auth login"], "auth");
			return workspace;
		}
	};
}
function workspaceMatchesRef(workspace, ref) {
	return workspace.id === ref || workspace.name.toLowerCase() === ref.toLowerCase();
}
async function resolveCurrentAuthState(dependencies) {
	const session = await dependencies.sessionGateway.readAuthSession();
	if (!session) return {
		authenticated: false,
		provider: null,
		user: null,
		workspace: null,
		credential: null
	};
	const provider = dependencies.identityGateway.getProvider(session.provider);
	const user = dependencies.identityGateway.getUser(session.userId);
	const workspace = dependencies.identityGateway.getWorkspace(session.workspaceId);
	if (!provider || !user || !workspace) return {
		authenticated: false,
		provider: null,
		user: null,
		workspace: null,
		credential: null
	};
	return {
		authenticated: true,
		provider: provider.id,
		user: {
			id: user.id,
			email: user.email,
			name: user.name
		},
		workspace,
		credential: {
			type: "oauth",
			id: null,
			name: null
		}
	};
}
//#endregion
export { createAuthUseCases };
