// @prisma/cli/dist/use-cases/branch.js
//#region src/use-cases/branch.ts
function createBranchUseCases(dependencies) {
	return { list: async () => {
		const projectId = await dependencies.projectStateGateway.readRememberedProjectId();
		if (!projectId) return {
			projectId: "",
			projectName: "not resolved",
			branches: []
		};
		const remoteBranches = await listRemoteBranches(dependencies.branchGateway, projectId);
		return {
			projectId,
			projectName: resolveProjectName(dependencies.projectGateway, projectId) ?? "not resolved",
			branches: buildBranchSummaries(remoteBranches)
		};
	} };
}
function resolveProjectName(projectGateway, projectId) {
	if (!projectId) return null;
	return projectGateway.getProject(projectId)?.name ?? null;
}
async function listRemoteBranches(branchGateway, projectId) {
	if (!projectId) return [];
	return branchGateway.listBranchesForProject(projectId);
}
function buildBranchSummaries(remoteBranches) {
	return sortBranches(remoteBranches.map((branch) => ({
		id: branch.id,
		name: branch.name,
		role: branch.role,
		envMap: branch.role
	})));
}
function sortBranches(branches) {
	return branches.slice().sort((left, right) => {
		const leftRank = branchOrder(left);
		const rightRank = branchOrder(right);
		if (leftRank !== rightRank) return leftRank - rightRank;
		return left.name.localeCompare(right.name);
	});
}
function branchOrder(branch) {
	if (branch.role === "production") return 0;
	return 1;
}
//#endregion
export { createBranchUseCases };
