// @prisma/cli/dist/use-cases/create-cli-gateways.js
//#region src/use-cases/create-cli-gateways.ts
function createCliUseCaseGateways(context) {
	return {
		identityGateway: {
			listProviders: () => context.api.listProviders(),
			getProvider: (providerId) => context.api.getProvider(providerId),
			listUsersForProvider: (providerId) => context.api.listUsersForProvider(providerId).map(toAuthUser),
			getUser: (userId) => {
				const user = context.api.getUser(userId);
				return user ? toAuthUser(user) : void 0;
			},
			getUserForProvider: (providerId, userId) => {
				const user = context.api.getUserForProvider(providerId, userId);
				return user ? toAuthUser(user) : void 0;
			},
			listUserWorkspaces: (userId) => context.api.listUserWorkspaces(userId).map(toAuthWorkspace),
			getWorkspace: (workspaceId) => {
				const workspace = context.api.getWorkspace(workspaceId);
				return workspace ? toAuthWorkspace(workspace) : void 0;
			},
			getUserWorkspace: (userId, workspaceId) => {
				const workspace = context.api.getUserWorkspace(userId, workspaceId);
				return workspace ? toAuthWorkspace(workspace) : void 0;
			}
		},
		projectGateway: {
			listProjectsForWorkspace: (workspaceId) => context.api.listProjectsForWorkspace(workspaceId),
			getProject: (projectId) => context.api.getProject(projectId),
			getProjectForWorkspace: (workspaceId, projectId) => context.api.getProjectForWorkspace(workspaceId, projectId)
		},
		branchGateway: {
			listBranchesForProject: (projectId) => context.api.listBranchesForProject(projectId),
			getBranchForProject: (projectId, name) => {
				return context.api.getBranchForProject(projectId, name);
			},
			getDeployment: (deploymentId) => context.api.getDeployment(deploymentId)
		},
		projectStateGateway: { readRememberedProjectId: async () => {
			return (await context.stateStore.readLastResolvedProject())?.id ?? null;
		} },
		sessionGateway: {
			readAuthSession: async () => {
				return (await context.stateStore.read()).auth;
			},
			writeAuthSession: async (session) => {
				await context.stateStore.setAuthSession(session);
			},
			clearAuthSession: async () => {
				await context.stateStore.clearAuthSession();
			}
		}
	};
}
function toAuthUser(user) {
	return {
		id: user.id,
		name: user.name,
		email: user.email
	};
}
function toAuthWorkspace(workspace) {
	return {
		id: workspace.id,
		name: workspace.name
	};
}
//#endregion
export { createCliUseCaseGateways };
