// @prisma/cli/dist/use-cases/project.js
import { authRequiredError } from "../shell/errors.js";
//#region src/use-cases/project.ts
function createProjectUseCases(dependencies) {
	return {
		list: async (authState) => {
			const workspace = requireWorkspace(authState);
			return {
				workspace,
				projects: listSortedWorkspaceProjects(dependencies.projectGateway, workspace.id).map(toProjectSummary)
			};
		},
		listProjectsForWorkspace: async (workspaceId) => listSortedWorkspaceProjects(dependencies.projectGateway, workspaceId).map(toProjectSummary)
	};
}
function requireWorkspace(authState) {
	if (!authState.authenticated || !authState.workspace) throw authRequiredError();
	return authState.workspace;
}
function listSortedWorkspaceProjects(projectGateway, workspaceId) {
	return projectGateway.listProjectsForWorkspace(workspaceId).slice().sort((left, right) => left.name.localeCompare(right.name) || left.id.localeCompare(right.id));
}
function toProjectSummary(project) {
	return {
		id: project.id,
		name: project.name,
		...project.url ? { url: project.url } : {}
	};
}
//#endregion
export { createProjectUseCases };
